

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Glossary &mdash; MAVIS  documentation</title>
  

  
  
  
  

  

  
  
    

  

  
  
    <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  

  
    <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  

  
        <link rel="author" title="About these documents"
              href="about.html"/>
        <link rel="index" title="Index"
              href="genindex.html"/>
        <link rel="search" title="Search" href="search.html"/>
    <link rel="top" title="MAVIS  documentation" href="index.html"/>
        <link rel="next" title="Literature" href="citations.html"/>
        <link rel="prev" title="util module" href="auto/mavis.util.html"/> 

  
  <script src="_static/js/modernizr.min.js"></script>

</head>

<body class="wy-body-for-nav" role="document">

   
  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search">
          

          
            <a href="index.html" class="icon icon-home"> MAVIS
          

          
          </a>

          
            
            
              <div class="version">
                1.6.4
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="about.html">About</a></li>
<li class="toctree-l1"><a class="reference internal" href="pipeline.html">Running the Pipeline</a></li>
<li class="toctree-l1"><a class="reference internal" href="mavis_input.html">MAVIS standard input file format</a></li>
<li class="toctree-l1"><a class="reference internal" href="reference.html">Reference Input Files</a></li>
<li class="toctree-l1"><a class="reference internal" href="configuration.html">Configuration and Settings</a></li>
<li class="toctree-l1"><a class="reference internal" href="tools_supported.html">Supported SV Callers</a></li>
<li class="toctree-l1"><a class="reference internal" href="illustrations.html">Illustrations</a></li>
<li class="toctree-l1"><a class="reference internal" href="faqs.html">Frequently Asked Questions</a></li>
<li class="toctree-l1"><a class="reference internal" href="theory.html">Theory and Models</a></li>
<li class="toctree-l1"><a class="reference internal" href="development.html">Guidelines for Contributors</a></li>
<li class="toctree-l1"><a class="reference internal" href="package.html">Package Documentation</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Glossary</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#general-terms">General Terms</a></li>
<li class="toctree-l2"><a class="reference internal" href="#configurable-settings">Configurable Settings</a></li>
<li class="toctree-l2"><a class="reference internal" href="#column-names">Column Names</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="citations.html">Literature</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">MAVIS</a>
        
      </nav>


      
      <div class="wy-nav-content">
        <div class="rst-content">
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html">Docs</a> &raquo;</li>
        
      <li>Glossary</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="_sources/glossary.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="glossary">
<h1>Glossary<a class="headerlink" href="#glossary" title="Permalink to this headline">¶</a></h1>
<div class="section" id="general-terms">
<h2>General Terms<a class="headerlink" href="#general-terms" title="Permalink to this headline">¶</a></h2>
<dl class="glossary docutils">
<dt id="term-2bit">2bit</dt>
<dd>File format specification. See <a class="reference external" href="https://genome.ucsc.edu/FAQ/FAQformat#format7">https://genome.ucsc.edu/FAQ/FAQformat#format7</a>.</dd>
<dt id="term-bam">bam</dt>
<dd>File format specification. See <a class="reference external" href="https://genome.ucsc.edu/FAQ/FAQformat#format5.1">https://genome.ucsc.edu/FAQ/FAQformat#format5.1</a>.</dd>
<dt id="term-bed">bed</dt>
<dd>File format specification. See <a class="reference external" href="https://genome.ucsc.edu/FAQ/FAQformat#format1">https://genome.ucsc.edu/FAQ/FAQformat#format1</a>.</dd>
<dt id="term-blat">blat</dt>
<dd>Alignment tool. see <a class="reference external" href="https://genome.ucsc.edu/FAQ/FAQblat.html">https://genome.ucsc.edu/FAQ/FAQblat.html</a>.</dd>
<dt id="term-breakpoint">breakpoint</dt>
<dd>A breakpoint is a genomic position (interval) on some reference/template/chromosome which has a strand and orientation. The orientation describes the portion of the reference that is retained.</dd>
<dt id="term-breakpoint-pair">breakpoint pair</dt>
<dd>Basic definition of a <a class="reference internal" href="#term-structural-variant"><span class="xref std std-term">structural variant</span></a>. Does not automatically imply a classification/type.</dd>
<dt id="term-bwa">BWA</dt>
<dd>BWA is an alignement tool. See <a class="reference external" href="https://github.com/lh3/bwa">https://github.com/lh3/bwa</a></dd>
<dt id="term-event">event</dt>
<dd>Used interchangeably with <a class="reference internal" href="#term-structural-variant"><span class="xref std std-term">structural variant</span></a>.</dd>
<dt id="term-event-type">event type</dt>
<dd>Classification for a structural variant. see <a class="reference internal" href="#term-131"><span class="xref std std-term">event_type</span></a>.</dd>
<dt id="term-fasta">fasta</dt>
<dd>File format specification. See <a class="reference external" href="https://genome.ucsc.edu/FAQ/FAQformat#format18">https://genome.ucsc.edu/FAQ/FAQformat#format18</a>.</dd>
<dt id="term-flanking-read-pair">flanking read pair</dt>
<dd>A pair of reads where one read maps to one side of a set of breakpoints and its mate maps to the other.</dd>
<dt id="term-half-mapped-read">half-mapped read</dt>
<dd>A read whose mate is unaligned. Generally this refers to reads in the evidence stage that are mapped next to a breakpoint.</dd>
<dt id="term-hgvs"><a class="reference internal" href="citations.html#den-dunnen-2016"><span class="std std-ref">HGVS</span></a></dt>
<dd>Community based standard of reccommendations for variant notation. See <a class="reference external" href="http://varnomen.hgvs.org/">http://varnomen.hgvs.org/</a></dd>
<dt id="term-igv">IGV</dt>
<dd>Integrative Genomics Viewer is a visualization tool. see <a class="reference external" href="http://software.broadinstitute.org/software/igv">http://software.broadinstitute.org/software/igv</a>.</dd>
<dt id="term-igv-batch-file">IGV batch file</dt>
<dd>This is a file format type defined by <a class="reference internal" href="#term-igv"><span class="xref std std-term">IGV</span></a> see
<a class="reference external" href="https://software.broadinstitute.org/software/igv/batch">running IGV with a batch file</a>.</dd>
<dt id="term-json">JSON</dt>
<dd>JSON (JavaScript Object Notation) is a data file format. see <a class="reference external" href="https://www.w3schools.com/js/js_json_intro.asp">https://www.w3schools.com/js/js_json_intro.asp</a>.</dd>
<dt id="term-psl">psl</dt>
<dd>File format specification. See <a class="reference external" href="https://genome.ucsc.edu/FAQ/FAQformat#format2">https://genome.ucsc.edu/FAQ/FAQformat#format2</a>.</dd>
<dt id="term-pslx">pslx</dt>
<dd>Extended format of a <a class="reference internal" href="#term-psl"><span class="xref std std-term">psl</span></a>.</dd>
<dt id="term-sge">SGE</dt>
<dd>Sun Grid Engine (SGE) is a job scheduling system for cluster management see <a class="reference external" href="http://star.mit.edu/cluster/docs/0.93.3/guides/sge.html">http://star.mit.edu/cluster/docs/0.93.3/guides/sge.html</a>.</dd>
<dt id="term-slurm">SLURM</dt>
<dd>SLURM is a job scheduling system for cluster management see <a class="reference external" href="https://slurm.schedmd.com/quickstart.html">https://slurm.schedmd.com/quickstart.html</a>.</dd>
<dt id="term-spanning-read">spanning read</dt>
<dd>Applies primarily to small structural variants. Reads which span both breakpoints.</dd>
<dt id="term-split-read">split read</dt>
<dd>A read which aligns next to a breakpoint and is softclipped at one or more sides.</dd>
<dt id="term-structural-variant">structural variant</dt>
<dd>A genomic alteration that can be described by a pair of breakpoints and an <a class="reference internal" href="#term-event-type"><span class="xref std std-term">event type</span></a>. The two breakpoints represent regions in the genome that are broken apart and reattached together.</dd>
<dt id="term-svg">SVG</dt>
<dd>SVG (Scalable vector graph) is an image format. see <a class="reference external" href="https://www.w3schools.com/graphics/svg_intro.asp">https://www.w3schools.com/graphics/svg_intro.asp</a>.</dd>
</dl>
</div>
<div class="section" id="configurable-settings">
<h2>Configurable Settings<a class="headerlink" href="#configurable-settings" title="Permalink to this headline">¶</a></h2>
<dl class="glossary docutils">
<dt id="term-aligner">aligner</dt>
<dd><a class="reference internal" href="auto/mavis.align.html#mavis.align.SUPPORTED_ALIGNER" title="mavis.align.SUPPORTED_ALIGNER"><code class="xref py py-class docutils literal"><span class="pre">SUPPORTED_ALIGNER</span></code></a> - The aligner to use to map the contigs/reads back to the reference e.g blat or bwa. The corresponding environment variable is <code class="docutils literal"><span class="pre">MAVIS_ALIGNER</span></code> and the default value is <code class="docutils literal"><span class="pre">'blat'</span></code>. Accepted values include: <code class="docutils literal"><span class="pre">'bwa</span> <span class="pre">mem'</span></code>, <code class="docutils literal"><span class="pre">'blat'</span></code></dd>
<dt id="term-annotation-filters">annotation_filters</dt>
<dd><a class="reference external" href="https://docs.python.org/3.6/library/stdtypes.html#str" title="(in Python v3.6)"><code class="xref py py-class docutils literal"><span class="pre">str</span></code></a> - A comma separated list of filters to apply to putative annotations. The corresponding environment variable is <code class="docutils literal"><span class="pre">MAVIS_ANNOTATION_FILTERS</span></code> and the default value is <code class="docutils literal"><span class="pre">'choose_more_annotated,choose_transcripts_by_priority'</span></code></dd>
<dt id="term-annotation-memory">annotation_memory</dt>
<dd><a class="reference external" href="https://docs.python.org/3.6/library/functions.html#int" title="(in Python v3.6)"><code class="xref py py-class docutils literal"><span class="pre">int</span></code></a> - Default memory limit (mb) for the annotation stage. The corresponding environment variable is <code class="docutils literal"><span class="pre">MAVIS_ANNOTATION_MEMORY</span></code> and the default value is <code class="docutils literal"><span class="pre">12000</span></code></dd>
<dt id="term-assembly-include-flanking-pairs">assembly_include_flanking_pairs</dt>
<dd><a class="reference external" href="https://docs.python.org/3.6/library/functions.html#bool" title="(in Python v3.6)"><code class="xref py py-class docutils literal"><span class="pre">bool</span></code></a> - If true then when the split reads are assembled, any flanking read pairs will also be added. The corresponding environment variable is <code class="docutils literal"><span class="pre">MAVIS_ASSEMBLY_INCLUDE_FLANKING_PAIRS</span></code> and the default value is <code class="docutils literal"><span class="pre">True</span></code></dd>
<dt id="term-assembly-include-half-mapped-reads">assembly_include_half_mapped_reads</dt>
<dd><a class="reference external" href="https://docs.python.org/3.6/library/functions.html#bool" title="(in Python v3.6)"><code class="xref py py-class docutils literal"><span class="pre">bool</span></code></a> - If true then when the split reads are assembled, any half-mapped read mates will also be added. The corresponding environment variable is <code class="docutils literal"><span class="pre">MAVIS_ASSEMBLY_INCLUDE_HALF_MAPPED_READS</span></code> and the default value is <code class="docutils literal"><span class="pre">True</span></code></dd>
<dt id="term-assembly-max-kmer-size">assembly_max_kmer_size</dt>
<dd><a class="reference external" href="https://docs.python.org/3.6/library/functions.html#int" title="(in Python v3.6)"><code class="xref py py-class docutils literal"><span class="pre">int</span></code></a> - The minimum between this and the smallest length input sequence is used as the kmer size for assembling the debruijn graph. if this is not set (any value less than 0 is considered not set) the default is the 75%% of the minimum length input sequence. The corresponding environment variable is <code class="docutils literal"><span class="pre">MAVIS_ASSEMBLY_MAX_KMER_SIZE</span></code> and the default value is <code class="docutils literal"><span class="pre">-1</span></code></dd>
<dt id="term-assembly-max-kmer-strict">assembly_max_kmer_strict</dt>
<dd><a class="reference external" href="https://docs.python.org/3.6/library/functions.html#bool" title="(in Python v3.6)"><code class="xref py py-class docutils literal"><span class="pre">bool</span></code></a> - If true then any sequences input to the assembly algorithm that cannot create a kmer of this size will be discarded. if false, then the kmer size will be reduced to the minimum input and all input sequences will be used in the assembly algorithm. The corresponding environment variable is <code class="docutils literal"><span class="pre">MAVIS_ASSEMBLY_MAX_KMER_STRICT</span></code> and the default value is <code class="docutils literal"><span class="pre">True</span></code></dd>
<dt id="term-assembly-max-paths">assembly_max_paths</dt>
<dd><a class="reference external" href="https://docs.python.org/3.6/library/functions.html#int" title="(in Python v3.6)"><code class="xref py py-class docutils literal"><span class="pre">int</span></code></a> - The maximum number of paths to resolve. this is used to limit when there is a messy assembly graph to resolve. the assembly will pre-calculate the number of paths (or putative assemblies) and stop if it is greater than the given setting. The corresponding environment variable is <code class="docutils literal"><span class="pre">MAVIS_ASSEMBLY_MAX_PATHS</span></code> and the default value is <code class="docutils literal"><span class="pre">8</span></code></dd>
<dt id="term-assembly-min-edge-weight">assembly_min_edge_weight</dt>
<dd><a class="reference external" href="https://docs.python.org/3.6/library/functions.html#int" title="(in Python v3.6)"><code class="xref py py-class docutils literal"><span class="pre">int</span></code></a> - Discards all edges with a weight/frequency less than this from the debruijn graph. The corresponding environment variable is <code class="docutils literal"><span class="pre">MAVIS_ASSEMBLY_MIN_EDGE_WEIGHT</span></code> and the default value is <code class="docutils literal"><span class="pre">2</span></code></dd>
<dt id="term-assembly-min-exact-match-to-remap">assembly_min_exact_match_to_remap</dt>
<dd><a class="reference external" href="https://docs.python.org/3.6/library/functions.html#int" title="(in Python v3.6)"><code class="xref py py-class docutils literal"><span class="pre">int</span></code></a> - The minimum length of exact matches to initiate remapping a read to a contig. The corresponding environment variable is <code class="docutils literal"><span class="pre">MAVIS_ASSEMBLY_MIN_EXACT_MATCH_TO_REMAP</span></code> and the default value is <code class="docutils literal"><span class="pre">15</span></code></dd>
<dt id="term-assembly-min-nc-edge-weight">assembly_min_nc_edge_weight</dt>
<dd><a class="reference external" href="https://docs.python.org/3.6/library/functions.html#int" title="(in Python v3.6)"><code class="xref py py-class docutils literal"><span class="pre">int</span></code></a> - Discards all non-cutting edges with a weight/frequency less than this from the debruijn graph. The corresponding environment variable is <code class="docutils literal"><span class="pre">MAVIS_ASSEMBLY_MIN_NC_EDGE_WEIGHT</span></code> and the default value is <code class="docutils literal"><span class="pre">3</span></code></dd>
<dt id="term-assembly-min-remap-coverage">assembly_min_remap_coverage</dt>
<dd><a class="reference internal" href="auto/mavis.constants.html#mavis.constants.float_fraction" title="mavis.constants.float_fraction"><code class="xref py py-class docutils literal"><span class="pre">float_fraction</span></code></a> - Minimum fraction of the contig sequence which the remapped sequences must align over. The corresponding environment variable is <code class="docutils literal"><span class="pre">MAVIS_ASSEMBLY_MIN_REMAP_COVERAGE</span></code> and the default value is <code class="docutils literal"><span class="pre">0.9</span></code></dd>
<dt id="term-assembly-min-remapped-seq">assembly_min_remapped_seq</dt>
<dd><a class="reference external" href="https://docs.python.org/3.6/library/functions.html#int" title="(in Python v3.6)"><code class="xref py py-class docutils literal"><span class="pre">int</span></code></a> - The minimum input sequences that must remap for an assembled contig to be used. The corresponding environment variable is <code class="docutils literal"><span class="pre">MAVIS_ASSEMBLY_MIN_REMAPPED_SEQ</span></code> and the default value is <code class="docutils literal"><span class="pre">3</span></code></dd>
<dt id="term-assembly-min-tgt-to-exclude-half-map">assembly_min_tgt_to_exclude_half_map</dt>
<dd><a class="reference external" href="https://docs.python.org/3.6/library/functions.html#int" title="(in Python v3.6)"><code class="xref py py-class docutils literal"><span class="pre">int</span></code></a> - The minimum number of split reads aligning to both breakpoints in order to exclude half-mapped reads from the assembly input. The corresponding environment variable is <code class="docutils literal"><span class="pre">MAVIS_ASSEMBLY_MIN_TGT_TO_EXCLUDE_HALF_MAP</span></code> and the default value is <code class="docutils literal"><span class="pre">7</span></code></dd>
<dt id="term-assembly-min-uniq">assembly_min_uniq</dt>
<dd><a class="reference internal" href="auto/mavis.constants.html#mavis.constants.float_fraction" title="mavis.constants.float_fraction"><code class="xref py py-class docutils literal"><span class="pre">float_fraction</span></code></a> - Minimum percent uniq required to keep separate assembled contigs. if contigs are more similar then the lower scoring, then shorter, contig is dropped. The corresponding environment variable is <code class="docutils literal"><span class="pre">MAVIS_ASSEMBLY_MIN_UNIQ</span></code> and the default value is <code class="docutils literal"><span class="pre">0.1</span></code></dd>
<dt id="term-assembly-strand-concordance">assembly_strand_concordance</dt>
<dd><a class="reference internal" href="auto/mavis.constants.html#mavis.constants.float_fraction" title="mavis.constants.float_fraction"><code class="xref py py-class docutils literal"><span class="pre">float_fraction</span></code></a> - When the number of remapped reads from each strand are compared, the ratio must be above this number to decide on the strand. The corresponding environment variable is <code class="docutils literal"><span class="pre">MAVIS_ASSEMBLY_STRAND_CONCORDANCE</span></code> and the default value is <code class="docutils literal"><span class="pre">0.51</span></code></dd>
<dt id="term-blat-limit-top-aln">blat_limit_top_aln</dt>
<dd><a class="reference external" href="https://docs.python.org/3.6/library/functions.html#int" title="(in Python v3.6)"><code class="xref py py-class docutils literal"><span class="pre">int</span></code></a> - Number of results to return from blat (ranking based on score). The corresponding environment variable is <code class="docutils literal"><span class="pre">MAVIS_BLAT_LIMIT_TOP_ALN</span></code> and the default value is <code class="docutils literal"><span class="pre">10</span></code></dd>
<dt id="term-blat-min-identity">blat_min_identity</dt>
<dd><a class="reference internal" href="auto/mavis.constants.html#mavis.constants.float_fraction" title="mavis.constants.float_fraction"><code class="xref py py-class docutils literal"><span class="pre">float_fraction</span></code></a> - The minimum percent identity match required for blat results when aligning contigs. The corresponding environment variable is <code class="docutils literal"><span class="pre">MAVIS_BLAT_MIN_IDENTITY</span></code> and the default value is <code class="docutils literal"><span class="pre">0.9</span></code></dd>
<dt id="term-breakpoint-color">breakpoint_color</dt>
<dd><a class="reference external" href="https://docs.python.org/3.6/library/stdtypes.html#str" title="(in Python v3.6)"><code class="xref py py-class docutils literal"><span class="pre">str</span></code></a> - Breakpoint outline color. The corresponding environment variable is <code class="docutils literal"><span class="pre">MAVIS_BREAKPOINT_COLOR</span></code> and the default value is <code class="docutils literal"><span class="pre">'#000000'</span></code></dd>
<dt id="term-call-error">call_error</dt>
<dd><a class="reference external" href="https://docs.python.org/3.6/library/functions.html#int" title="(in Python v3.6)"><code class="xref py py-class docutils literal"><span class="pre">int</span></code></a> - Buffer zone for the evidence window. The corresponding environment variable is <code class="docutils literal"><span class="pre">MAVIS_CALL_ERROR</span></code> and the default value is <code class="docutils literal"><span class="pre">10</span></code></dd>
<dt id="term-clean-aligner-files">clean_aligner_files</dt>
<dd><a class="reference external" href="https://docs.python.org/3.6/library/functions.html#bool" title="(in Python v3.6)"><code class="xref py py-class docutils literal"><span class="pre">bool</span></code></a> - Remove the aligner output files after the validation stage is complete. not required for subsequent steps but can be useful in debugging and deep investigation of events. The corresponding environment variable is <code class="docutils literal"><span class="pre">MAVIS_CLEAN_ALIGNER_FILES</span></code> and the default value is <code class="docutils literal"><span class="pre">False</span></code></dd>
<dt id="term-cluster-initial-size-limit">cluster_initial_size_limit</dt>
<dd><a class="reference external" href="https://docs.python.org/3.6/library/functions.html#int" title="(in Python v3.6)"><code class="xref py py-class docutils literal"><span class="pre">int</span></code></a> - The maximum cumulative size of both breakpoints for breakpoint pairs to be used in the initial clustering phase (combining based on overlap). The corresponding environment variable is <code class="docutils literal"><span class="pre">MAVIS_CLUSTER_INITIAL_SIZE_LIMIT</span></code> and the default value is <code class="docutils literal"><span class="pre">25</span></code></dd>
<dt id="term-cluster-radius">cluster_radius</dt>
<dd><a class="reference external" href="https://docs.python.org/3.6/library/functions.html#int" title="(in Python v3.6)"><code class="xref py py-class docutils literal"><span class="pre">int</span></code></a> - Maximum distance allowed between paired breakpoint pairs. The corresponding environment variable is <code class="docutils literal"><span class="pre">MAVIS_CLUSTER_RADIUS</span></code> and the default value is <code class="docutils literal"><span class="pre">100</span></code></dd>
<dt id="term-contig-aln-max-event-size">contig_aln_max_event_size</dt>
<dd><a class="reference external" href="https://docs.python.org/3.6/library/functions.html#int" title="(in Python v3.6)"><code class="xref py py-class docutils literal"><span class="pre">int</span></code></a> - Relates to determining breakpoints when pairing contig alignments. for any given read in a putative pair the soft clipping is extended to include any events of greater than this size. the softclipping is added to the side of the alignment as indicated by the breakpoint we are assigning pairs to. The corresponding environment variable is <code class="docutils literal"><span class="pre">MAVIS_CONTIG_ALN_MAX_EVENT_SIZE</span></code> and the default value is <code class="docutils literal"><span class="pre">50</span></code></dd>
<dt id="term-contig-aln-merge-inner-anchor">contig_aln_merge_inner_anchor</dt>
<dd><a class="reference external" href="https://docs.python.org/3.6/library/functions.html#int" title="(in Python v3.6)"><code class="xref py py-class docutils literal"><span class="pre">int</span></code></a> - The minimum number of consecutive exact match base pairs to not merge events within a contig alignment. The corresponding environment variable is <code class="docutils literal"><span class="pre">MAVIS_CONTIG_ALN_MERGE_INNER_ANCHOR</span></code> and the default value is <code class="docutils literal"><span class="pre">20</span></code></dd>
<dt id="term-contig-aln-merge-outer-anchor">contig_aln_merge_outer_anchor</dt>
<dd><a class="reference external" href="https://docs.python.org/3.6/library/functions.html#int" title="(in Python v3.6)"><code class="xref py py-class docutils literal"><span class="pre">int</span></code></a> - Minimum consecutively aligned exact matches to anchor an end for merging internal events. The corresponding environment variable is <code class="docutils literal"><span class="pre">MAVIS_CONTIG_ALN_MERGE_OUTER_ANCHOR</span></code> and the default value is <code class="docutils literal"><span class="pre">15</span></code></dd>
<dt id="term-contig-aln-min-anchor-size">contig_aln_min_anchor_size</dt>
<dd><a class="reference external" href="https://docs.python.org/3.6/library/functions.html#int" title="(in Python v3.6)"><code class="xref py py-class docutils literal"><span class="pre">int</span></code></a> - The minimum number of aligned bases for a contig (m or =) in order to simplify. do not have to be consecutive. The corresponding environment variable is <code class="docutils literal"><span class="pre">MAVIS_CONTIG_ALN_MIN_ANCHOR_SIZE</span></code> and the default value is <code class="docutils literal"><span class="pre">50</span></code></dd>
<dt id="term-contig-aln-min-extend-overlap">contig_aln_min_extend_overlap</dt>
<dd><a class="reference external" href="https://docs.python.org/3.6/library/functions.html#int" title="(in Python v3.6)"><code class="xref py py-class docutils literal"><span class="pre">int</span></code></a> - Minimum number of bases the query coverage interval must be extended by in order to pair alignments as a single split alignment. The corresponding environment variable is <code class="docutils literal"><span class="pre">MAVIS_CONTIG_ALN_MIN_EXTEND_OVERLAP</span></code> and the default value is <code class="docutils literal"><span class="pre">10</span></code></dd>
<dt id="term-contig-aln-min-query-consumption">contig_aln_min_query_consumption</dt>
<dd><a class="reference internal" href="auto/mavis.constants.html#mavis.constants.float_fraction" title="mavis.constants.float_fraction"><code class="xref py py-class docutils literal"><span class="pre">float_fraction</span></code></a> - Minimum fraction of the original query sequence that must be used by the read(s) of the alignment. The corresponding environment variable is <code class="docutils literal"><span class="pre">MAVIS_CONTIG_ALN_MIN_QUERY_CONSUMPTION</span></code> and the default value is <code class="docutils literal"><span class="pre">0.9</span></code></dd>
<dt id="term-contig-aln-min-score">contig_aln_min_score</dt>
<dd><a class="reference internal" href="auto/mavis.constants.html#mavis.constants.float_fraction" title="mavis.constants.float_fraction"><code class="xref py py-class docutils literal"><span class="pre">float_fraction</span></code></a> - Minimum score for a contig to be used as evidence in a call by contig. The corresponding environment variable is <code class="docutils literal"><span class="pre">MAVIS_CONTIG_ALN_MIN_SCORE</span></code> and the default value is <code class="docutils literal"><span class="pre">0.9</span></code></dd>
<dt id="term-contig-call-distance">contig_call_distance</dt>
<dd><a class="reference external" href="https://docs.python.org/3.6/library/functions.html#int" title="(in Python v3.6)"><code class="xref py py-class docutils literal"><span class="pre">int</span></code></a> - The maximum distance allowed between breakpoint pairs (called by contig) in order for them to pair. The corresponding environment variable is <code class="docutils literal"><span class="pre">MAVIS_CONTIG_CALL_DISTANCE</span></code> and the default value is <code class="docutils literal"><span class="pre">0</span></code></dd>
<dt id="term-domain-color">domain_color</dt>
<dd><a class="reference external" href="https://docs.python.org/3.6/library/stdtypes.html#str" title="(in Python v3.6)"><code class="xref py py-class docutils literal"><span class="pre">str</span></code></a> - Domain fill color. The corresponding environment variable is <code class="docutils literal"><span class="pre">MAVIS_DOMAIN_COLOR</span></code> and the default value is <code class="docutils literal"><span class="pre">'#ccccb3'</span></code></dd>
<dt id="term-domain-mismatch-color">domain_mismatch_color</dt>
<dd><a class="reference external" href="https://docs.python.org/3.6/library/stdtypes.html#str" title="(in Python v3.6)"><code class="xref py py-class docutils literal"><span class="pre">str</span></code></a> - Domain fill color on 0%% match. The corresponding environment variable is <code class="docutils literal"><span class="pre">MAVIS_DOMAIN_MISMATCH_COLOR</span></code> and the default value is <code class="docutils literal"><span class="pre">'#b2182b'</span></code></dd>
<dt id="term-domain-name-regex-filter">domain_name_regex_filter</dt>
<dd><a class="reference external" href="https://docs.python.org/3.6/library/stdtypes.html#str" title="(in Python v3.6)"><code class="xref py py-class docutils literal"><span class="pre">str</span></code></a> - The regular expression used to select domains to be displayed (filtered by name). The corresponding environment variable is <code class="docutils literal"><span class="pre">MAVIS_DOMAIN_NAME_REGEX_FILTER</span></code> and the default value is <code class="docutils literal"><span class="pre">'^PF\\d+$'</span></code></dd>
<dt id="term-domain-scaffold-color">domain_scaffold_color</dt>
<dd><a class="reference external" href="https://docs.python.org/3.6/library/stdtypes.html#str" title="(in Python v3.6)"><code class="xref py py-class docutils literal"><span class="pre">str</span></code></a> - The color of the domain scaffold. The corresponding environment variable is <code class="docutils literal"><span class="pre">MAVIS_DOMAIN_SCAFFOLD_COLOR</span></code> and the default value is <code class="docutils literal"><span class="pre">'#000000'</span></code></dd>
<dt id="term-draw-fusions-only">draw_fusions_only</dt>
<dd><a class="reference external" href="https://docs.python.org/3.6/library/functions.html#bool" title="(in Python v3.6)"><code class="xref py py-class docutils literal"><span class="pre">bool</span></code></a> - Flag to indicate if events which do not produce a fusion transcript should produce illustrations. The corresponding environment variable is <code class="docutils literal"><span class="pre">MAVIS_DRAW_FUSIONS_ONLY</span></code> and the default value is <code class="docutils literal"><span class="pre">True</span></code></dd>
<dt id="term-draw-non-synonymous-cdna-only">draw_non_synonymous_cdna_only</dt>
<dd><a class="reference external" href="https://docs.python.org/3.6/library/functions.html#bool" title="(in Python v3.6)"><code class="xref py py-class docutils literal"><span class="pre">bool</span></code></a> - Flag to indicate if events which are synonymous at the cdna level should produce illustrations. The corresponding environment variable is <code class="docutils literal"><span class="pre">MAVIS_DRAW_NON_SYNONYMOUS_CDNA_ONLY</span></code> and the default value is <code class="docutils literal"><span class="pre">True</span></code></dd>
<dt id="term-drawing-width-iter-increase">drawing_width_iter_increase</dt>
<dd><a class="reference external" href="https://docs.python.org/3.6/library/functions.html#int" title="(in Python v3.6)"><code class="xref py py-class docutils literal"><span class="pre">int</span></code></a> - The amount (in  pixels) by which to increase the drawing width upon failure to fit. The corresponding environment variable is <code class="docutils literal"><span class="pre">MAVIS_DRAWING_WIDTH_ITER_INCREASE</span></code> and the default value is <code class="docutils literal"><span class="pre">500</span></code></dd>
<dt id="term-fetch-min-bin-size">fetch_min_bin_size</dt>
<dd><a class="reference external" href="https://docs.python.org/3.6/library/functions.html#int" title="(in Python v3.6)"><code class="xref py py-class docutils literal"><span class="pre">int</span></code></a> - The minimum size of any bin for reading from a bam file. increasing this number will result in smaller bins being merged or less bins being created (depending on the fetch method). The corresponding environment variable is <code class="docutils literal"><span class="pre">MAVIS_FETCH_MIN_BIN_SIZE</span></code> and the default value is <code class="docutils literal"><span class="pre">50</span></code></dd>
<dt id="term-fetch-reads-bins">fetch_reads_bins</dt>
<dd><a class="reference external" href="https://docs.python.org/3.6/library/functions.html#int" title="(in Python v3.6)"><code class="xref py py-class docutils literal"><span class="pre">int</span></code></a> - Number of bins to split an evidence window into to ensure more even sampling of high coverage regions. The corresponding environment variable is <code class="docutils literal"><span class="pre">MAVIS_FETCH_READS_BINS</span></code> and the default value is <code class="docutils literal"><span class="pre">5</span></code></dd>
<dt id="term-fetch-reads-limit">fetch_reads_limit</dt>
<dd><a class="reference external" href="https://docs.python.org/3.6/library/functions.html#int" title="(in Python v3.6)"><code class="xref py py-class docutils literal"><span class="pre">int</span></code></a> - Maximum number of reads, cap, to loop over for any given evidence window. The corresponding environment variable is <code class="docutils literal"><span class="pre">MAVIS_FETCH_READS_LIMIT</span></code> and the default value is <code class="docutils literal"><span class="pre">3000</span></code></dd>
<dt id="term-filter-cdna-synon">filter_cdna_synon</dt>
<dd><a class="reference external" href="https://docs.python.org/3.6/library/functions.html#bool" title="(in Python v3.6)"><code class="xref py py-class docutils literal"><span class="pre">bool</span></code></a> - Filter all annotations synonymous at the cdna level. The corresponding environment variable is <code class="docutils literal"><span class="pre">MAVIS_FILTER_CDNA_SYNON</span></code> and the default value is <code class="docutils literal"><span class="pre">True</span></code></dd>
<dt id="term-filter-min-flanking-reads">filter_min_flanking_reads</dt>
<dd><a class="reference external" href="https://docs.python.org/3.6/library/functions.html#int" title="(in Python v3.6)"><code class="xref py py-class docutils literal"><span class="pre">int</span></code></a> - Minimum number of flanking pairs for a call by flanking pairs. The corresponding environment variable is <code class="docutils literal"><span class="pre">MAVIS_FILTER_MIN_FLANKING_READS</span></code> and the default value is <code class="docutils literal"><span class="pre">10</span></code></dd>
<dt id="term-filter-min-linking-split-reads">filter_min_linking_split_reads</dt>
<dd><a class="reference external" href="https://docs.python.org/3.6/library/functions.html#int" title="(in Python v3.6)"><code class="xref py py-class docutils literal"><span class="pre">int</span></code></a> - Minimum number of linking split reads for a call by split reads. The corresponding environment variable is <code class="docutils literal"><span class="pre">MAVIS_FILTER_MIN_LINKING_SPLIT_READS</span></code> and the default value is <code class="docutils literal"><span class="pre">1</span></code></dd>
<dt id="term-filter-min-remapped-reads">filter_min_remapped_reads</dt>
<dd><a class="reference external" href="https://docs.python.org/3.6/library/functions.html#int" title="(in Python v3.6)"><code class="xref py py-class docutils literal"><span class="pre">int</span></code></a> - Minimum number of remapped reads for a call by contig. The corresponding environment variable is <code class="docutils literal"><span class="pre">MAVIS_FILTER_MIN_REMAPPED_READS</span></code> and the default value is <code class="docutils literal"><span class="pre">5</span></code></dd>
<dt id="term-filter-min-spanning-reads">filter_min_spanning_reads</dt>
<dd><a class="reference external" href="https://docs.python.org/3.6/library/functions.html#int" title="(in Python v3.6)"><code class="xref py py-class docutils literal"><span class="pre">int</span></code></a> - Minimum number of spanning reads for a call by spanning reads. The corresponding environment variable is <code class="docutils literal"><span class="pre">MAVIS_FILTER_MIN_SPANNING_READS</span></code> and the default value is <code class="docutils literal"><span class="pre">5</span></code></dd>
<dt id="term-filter-min-split-reads">filter_min_split_reads</dt>
<dd><a class="reference external" href="https://docs.python.org/3.6/library/functions.html#int" title="(in Python v3.6)"><code class="xref py py-class docutils literal"><span class="pre">int</span></code></a> - Minimum number of split reads for a call by split reads. The corresponding environment variable is <code class="docutils literal"><span class="pre">MAVIS_FILTER_MIN_SPLIT_READS</span></code> and the default value is <code class="docutils literal"><span class="pre">5</span></code></dd>
<dt id="term-filter-protein-synon">filter_protein_synon</dt>
<dd><a class="reference external" href="https://docs.python.org/3.6/library/functions.html#bool" title="(in Python v3.6)"><code class="xref py py-class docutils literal"><span class="pre">bool</span></code></a> - Filter all annotations synonymous at the protein level. The corresponding environment variable is <code class="docutils literal"><span class="pre">MAVIS_FILTER_PROTEIN_SYNON</span></code> and the default value is <code class="docutils literal"><span class="pre">True</span></code></dd>
<dt id="term-filter-secondary-alignments">filter_secondary_alignments</dt>
<dd><a class="reference external" href="https://docs.python.org/3.6/library/functions.html#bool" title="(in Python v3.6)"><code class="xref py py-class docutils literal"><span class="pre">bool</span></code></a> - Filter secondary alignments when gathering read evidence. The corresponding environment variable is <code class="docutils literal"><span class="pre">MAVIS_FILTER_SECONDARY_ALIGNMENTS</span></code> and the default value is <code class="docutils literal"><span class="pre">True</span></code></dd>
<dt id="term-filter-trans-homopolymers">filter_trans_homopolymers</dt>
<dd><a class="reference external" href="https://docs.python.org/3.6/library/functions.html#bool" title="(in Python v3.6)"><code class="xref py py-class docutils literal"><span class="pre">bool</span></code></a> - Filter all single bp ins/del/dup events that are in a homopolymer region of at least 5 bps and are not paired to a genomic event. The corresponding environment variable is <code class="docutils literal"><span class="pre">MAVIS_FILTER_TRANS_HOMOPOLYMERS</span></code> and the default value is <code class="docutils literal"><span class="pre">True</span></code></dd>
<dt id="term-flanking-call-distance">flanking_call_distance</dt>
<dd><a class="reference external" href="https://docs.python.org/3.6/library/functions.html#int" title="(in Python v3.6)"><code class="xref py py-class docutils literal"><span class="pre">int</span></code></a> - The maximum distance allowed between breakpoint pairs (called by flanking pairs) in order for them to pair. The corresponding environment variable is <code class="docutils literal"><span class="pre">MAVIS_FLANKING_CALL_DISTANCE</span></code> and the default value is <code class="docutils literal"><span class="pre">0</span></code></dd>
<dt id="term-fuzzy-mismatch-number">fuzzy_mismatch_number</dt>
<dd><a class="reference external" href="https://docs.python.org/3.6/library/functions.html#int" title="(in Python v3.6)"><code class="xref py py-class docutils literal"><span class="pre">int</span></code></a> - The number of events/mismatches allowed to be considered a fuzzy match. The corresponding environment variable is <code class="docutils literal"><span class="pre">MAVIS_FUZZY_MISMATCH_NUMBER</span></code> and the default value is <code class="docutils literal"><span class="pre">1</span></code></dd>
<dt id="term-gene1-color">gene1_color</dt>
<dd><a class="reference external" href="https://docs.python.org/3.6/library/stdtypes.html#str" title="(in Python v3.6)"><code class="xref py py-class docutils literal"><span class="pre">str</span></code></a> - The color of genes near the first gene. The corresponding environment variable is <code class="docutils literal"><span class="pre">MAVIS_GENE1_COLOR</span></code> and the default value is <code class="docutils literal"><span class="pre">'#657e91'</span></code></dd>
<dt id="term-gene1-color-selected">gene1_color_selected</dt>
<dd><a class="reference external" href="https://docs.python.org/3.6/library/stdtypes.html#str" title="(in Python v3.6)"><code class="xref py py-class docutils literal"><span class="pre">str</span></code></a> - The color of the first gene. The corresponding environment variable is <code class="docutils literal"><span class="pre">MAVIS_GENE1_COLOR_SELECTED</span></code> and the default value is <code class="docutils literal"><span class="pre">'#518dc5'</span></code></dd>
<dt id="term-gene2-color">gene2_color</dt>
<dd><a class="reference external" href="https://docs.python.org/3.6/library/stdtypes.html#str" title="(in Python v3.6)"><code class="xref py py-class docutils literal"><span class="pre">str</span></code></a> - The color of genes near the second gene. The corresponding environment variable is <code class="docutils literal"><span class="pre">MAVIS_GENE2_COLOR</span></code> and the default value is <code class="docutils literal"><span class="pre">'#325556'</span></code></dd>
<dt id="term-gene2-color-selected">gene2_color_selected</dt>
<dd><a class="reference external" href="https://docs.python.org/3.6/library/stdtypes.html#str" title="(in Python v3.6)"><code class="xref py py-class docutils literal"><span class="pre">str</span></code></a> - The color of the second gene. The corresponding environment variable is <code class="docutils literal"><span class="pre">MAVIS_GENE2_COLOR_SELECTED</span></code> and the default value is <code class="docutils literal"><span class="pre">'#4c9677'</span></code></dd>
<dt id="term-import-env">import_env</dt>
<dd><a class="reference external" href="https://docs.python.org/3.6/library/functions.html#bool" title="(in Python v3.6)"><code class="xref py py-class docutils literal"><span class="pre">bool</span></code></a> - Flag to import environment variables. The corresponding environment variable is <code class="docutils literal"><span class="pre">MAVIS_IMPORT_ENV</span></code> and the default value is <code class="docutils literal"><span class="pre">True</span></code></dd>
<dt id="term-input-call-distance">input_call_distance</dt>
<dd><a class="reference external" href="https://docs.python.org/3.6/library/functions.html#int" title="(in Python v3.6)"><code class="xref py py-class docutils literal"><span class="pre">int</span></code></a> - The maximum distance allowed between breakpoint pairs (called by input tools, not validated) in order for them to pair. The corresponding environment variable is <code class="docutils literal"><span class="pre">MAVIS_INPUT_CALL_DISTANCE</span></code> and the default value is <code class="docutils literal"><span class="pre">5</span></code></dd>
<dt id="term-label-color">label_color</dt>
<dd><a class="reference external" href="https://docs.python.org/3.6/library/stdtypes.html#str" title="(in Python v3.6)"><code class="xref py py-class docutils literal"><span class="pre">str</span></code></a> - The label color. The corresponding environment variable is <code class="docutils literal"><span class="pre">MAVIS_LABEL_COLOR</span></code> and the default value is <code class="docutils literal"><span class="pre">'#000000'</span></code></dd>
<dt id="term-limit-to-chr">limit_to_chr</dt>
<dd><a class="reference internal" href="auto/mavis.util.html#mavis.util.ChrListString" title="mavis.util.ChrListString"><code class="xref py py-class docutils literal"><span class="pre">ChrListString</span></code></a> - A semi-colon delimited list of chromosome names to use. breakpointpairs on other chromosomes will be filteredout. for example ‘1;2;3;4’ would filter out events/breakpoint pairs on any chromosomes but 1, 2, 3, and 4. The corresponding environment variable is <code class="docutils literal"><span class="pre">MAVIS_LIMIT_TO_CHR</span></code> and the default value is <code class="docutils literal"><span class="pre">['1',</span> <span class="pre">'2',</span> <span class="pre">'3',</span> <span class="pre">'4',</span> <span class="pre">'5',</span> <span class="pre">'6',</span> <span class="pre">'7',</span> <span class="pre">'8',</span> <span class="pre">'9',</span> <span class="pre">'10',</span> <span class="pre">'11',</span> <span class="pre">'12',</span> <span class="pre">'13',</span> <span class="pre">'14',</span> <span class="pre">'15',</span> <span class="pre">'16',</span> <span class="pre">'17',</span> <span class="pre">'18',</span> <span class="pre">'19',</span> <span class="pre">'20',</span> <span class="pre">'21',</span> <span class="pre">'22',</span> <span class="pre">'X',</span> <span class="pre">'Y']</span></code></dd>
<dt id="term-mask-fill">mask_fill</dt>
<dd><a class="reference external" href="https://docs.python.org/3.6/library/stdtypes.html#str" title="(in Python v3.6)"><code class="xref py py-class docutils literal"><span class="pre">str</span></code></a> - Color of mask (for deleted region etc.). The corresponding environment variable is <code class="docutils literal"><span class="pre">MAVIS_MASK_FILL</span></code> and the default value is <code class="docutils literal"><span class="pre">'#ffffff'</span></code></dd>
<dt id="term-mask-opacity">mask_opacity</dt>
<dd><a class="reference internal" href="auto/mavis.constants.html#mavis.constants.float_fraction" title="mavis.constants.float_fraction"><code class="xref py py-class docutils literal"><span class="pre">float_fraction</span></code></a> - Opacity of the mask layer. The corresponding environment variable is <code class="docutils literal"><span class="pre">MAVIS_MASK_OPACITY</span></code> and the default value is <code class="docutils literal"><span class="pre">0.7</span></code></dd>
<dt id="term-max-drawing-retries">max_drawing_retries</dt>
<dd><a class="reference external" href="https://docs.python.org/3.6/library/functions.html#int" title="(in Python v3.6)"><code class="xref py py-class docutils literal"><span class="pre">int</span></code></a> - The maximum number of retries for attempting a drawing. each iteration the width is extended. if it is still insufficient after this number a gene-level only drawing will be output. The corresponding environment variable is <code class="docutils literal"><span class="pre">MAVIS_MAX_DRAWING_RETRIES</span></code> and the default value is <code class="docutils literal"><span class="pre">5</span></code></dd>
<dt id="term-max-files">max_files</dt>
<dd><a class="reference external" href="https://docs.python.org/3.6/library/functions.html#int" title="(in Python v3.6)"><code class="xref py py-class docutils literal"><span class="pre">int</span></code></a> - The maximum number of files to output from clustering/splitting. The corresponding environment variable is <code class="docutils literal"><span class="pre">MAVIS_MAX_FILES</span></code> and the default value is <code class="docutils literal"><span class="pre">200</span></code></dd>
<dt id="term-max-orf-cap">max_orf_cap</dt>
<dd><a class="reference external" href="https://docs.python.org/3.6/library/functions.html#int" title="(in Python v3.6)"><code class="xref py py-class docutils literal"><span class="pre">int</span></code></a> - The maximum number of orfs to return (best putative orfs will be retained). The corresponding environment variable is <code class="docutils literal"><span class="pre">MAVIS_MAX_ORF_CAP</span></code> and the default value is <code class="docutils literal"><span class="pre">3</span></code></dd>
<dt id="term-max-proximity">max_proximity</dt>
<dd><a class="reference external" href="https://docs.python.org/3.6/library/functions.html#int" title="(in Python v3.6)"><code class="xref py py-class docutils literal"><span class="pre">int</span></code></a> - The maximum distance away from an annotation before the region in considered to be uninformative. The corresponding environment variable is <code class="docutils literal"><span class="pre">MAVIS_MAX_PROXIMITY</span></code> and the default value is <code class="docutils literal"><span class="pre">5000</span></code></dd>
<dt id="term-max-sc-preceeding-anchor">max_sc_preceeding_anchor</dt>
<dd><a class="reference external" href="https://docs.python.org/3.6/library/functions.html#int" title="(in Python v3.6)"><code class="xref py py-class docutils literal"><span class="pre">int</span></code></a> - When remapping a softclipped read this determines the amount of softclipping allowed on the side opposite of where we expect it. for example for a softclipped read on a breakpoint with a left orientation this limits the amount of softclipping that is allowed on the right. if this is set to none then there is no limit on softclipping. The corresponding environment variable is <code class="docutils literal"><span class="pre">MAVIS_MAX_SC_PRECEEDING_ANCHOR</span></code> and the default value is <code class="docutils literal"><span class="pre">6</span></code></dd>
<dt id="term-memory-limit">memory_limit</dt>
<dd><a class="reference external" href="https://docs.python.org/3.6/library/functions.html#int" title="(in Python v3.6)"><code class="xref py py-class docutils literal"><span class="pre">int</span></code></a> - The maximum number of megabytes (mb) any given job is allowed. The corresponding environment variable is <code class="docutils literal"><span class="pre">MAVIS_MEMORY_LIMIT</span></code> and the default value is <code class="docutils literal"><span class="pre">16000</span></code></dd>
<dt id="term-min-anchor-exact">min_anchor_exact</dt>
<dd><a class="reference external" href="https://docs.python.org/3.6/library/functions.html#int" title="(in Python v3.6)"><code class="xref py py-class docutils literal"><span class="pre">int</span></code></a> - Applies to re-aligning softclipped reads to the opposing breakpoint. the minimum number of consecutive exact matches to anchor a read to initiate targeted realignment. The corresponding environment variable is <code class="docutils literal"><span class="pre">MAVIS_MIN_ANCHOR_EXACT</span></code> and the default value is <code class="docutils literal"><span class="pre">6</span></code></dd>
<dt id="term-min-anchor-fuzzy">min_anchor_fuzzy</dt>
<dd><a class="reference external" href="https://docs.python.org/3.6/library/functions.html#int" title="(in Python v3.6)"><code class="xref py py-class docutils literal"><span class="pre">int</span></code></a> - Applies to re-aligning softclipped reads to the opposing breakpoint. the minimum length of a fuzzy match to anchor a read to initiate targeted realignment. The corresponding environment variable is <code class="docutils literal"><span class="pre">MAVIS_MIN_ANCHOR_FUZZY</span></code> and the default value is <code class="docutils literal"><span class="pre">10</span></code></dd>
<dt id="term-min-anchor-match">min_anchor_match</dt>
<dd><a class="reference internal" href="auto/mavis.constants.html#mavis.constants.float_fraction" title="mavis.constants.float_fraction"><code class="xref py py-class docutils literal"><span class="pre">float_fraction</span></code></a> - Minimum percent match for a read to be kept as evidence. The corresponding environment variable is <code class="docutils literal"><span class="pre">MAVIS_MIN_ANCHOR_MATCH</span></code> and the default value is <code class="docutils literal"><span class="pre">0.9</span></code></dd>
<dt id="term-min-clusters-per-file">min_clusters_per_file</dt>
<dd><a class="reference external" href="https://docs.python.org/3.6/library/functions.html#int" title="(in Python v3.6)"><code class="xref py py-class docutils literal"><span class="pre">int</span></code></a> - The minimum number of breakpoint pairs to output to a file. The corresponding environment variable is <code class="docutils literal"><span class="pre">MAVIS_MIN_CLUSTERS_PER_FILE</span></code> and the default value is <code class="docutils literal"><span class="pre">50</span></code></dd>
<dt id="term-min-domain-mapping-match">min_domain_mapping_match</dt>
<dd><a class="reference internal" href="auto/mavis.constants.html#mavis.constants.float_fraction" title="mavis.constants.float_fraction"><code class="xref py py-class docutils literal"><span class="pre">float_fraction</span></code></a> - A number between 0 and 1 representing the minimum percent match a domain must map to the fusion transcript to be displayed. The corresponding environment variable is <code class="docutils literal"><span class="pre">MAVIS_MIN_DOMAIN_MAPPING_MATCH</span></code> and the default value is <code class="docutils literal"><span class="pre">0.9</span></code></dd>
<dt id="term-min-double-aligned-to-estimate-insertion-size">min_double_aligned_to_estimate_insertion_size</dt>
<dd><a class="reference external" href="https://docs.python.org/3.6/library/functions.html#int" title="(in Python v3.6)"><code class="xref py py-class docutils literal"><span class="pre">int</span></code></a> - The minimum number of reads which map soft-clipped to both breakpoints to assume the size of the untemplated sequence between the breakpoints is at most the read length - 2 * min_softclipping. The corresponding environment variable is <code class="docutils literal"><span class="pre">MAVIS_MIN_DOUBLE_ALIGNED_TO_ESTIMATE_INSERTION_SIZE</span></code> and the default value is <code class="docutils literal"><span class="pre">2</span></code></dd>
<dt id="term-min-flanking-pairs-resolution">min_flanking_pairs_resolution</dt>
<dd><a class="reference external" href="https://docs.python.org/3.6/library/functions.html#int" title="(in Python v3.6)"><code class="xref py py-class docutils literal"><span class="pre">int</span></code></a> - The minimum number of flanking reads required to call a breakpoint by flanking evidence. The corresponding environment variable is <code class="docutils literal"><span class="pre">MAVIS_MIN_FLANKING_PAIRS_RESOLUTION</span></code> and the default value is <code class="docutils literal"><span class="pre">10</span></code></dd>
<dt id="term-min-linking-split-reads">min_linking_split_reads</dt>
<dd><a class="reference external" href="https://docs.python.org/3.6/library/functions.html#int" title="(in Python v3.6)"><code class="xref py py-class docutils literal"><span class="pre">int</span></code></a> - The minimum number of split reads which aligned to both breakpoints. The corresponding environment variable is <code class="docutils literal"><span class="pre">MAVIS_MIN_LINKING_SPLIT_READS</span></code> and the default value is <code class="docutils literal"><span class="pre">2</span></code></dd>
<dt id="term-min-mapping-quality">min_mapping_quality</dt>
<dd><a class="reference external" href="https://docs.python.org/3.6/library/functions.html#int" title="(in Python v3.6)"><code class="xref py py-class docutils literal"><span class="pre">int</span></code></a> - The minimum mapping quality of reads to be used as evidence. The corresponding environment variable is <code class="docutils literal"><span class="pre">MAVIS_MIN_MAPPING_QUALITY</span></code> and the default value is <code class="docutils literal"><span class="pre">5</span></code></dd>
<dt id="term-min-non-target-aligned-split-reads">min_non_target_aligned_split_reads</dt>
<dd><a class="reference external" href="https://docs.python.org/3.6/library/functions.html#int" title="(in Python v3.6)"><code class="xref py py-class docutils literal"><span class="pre">int</span></code></a> - The minimum number of split reads aligned to a breakpoint by the input bam and no forced by local alignment to the target region to call a breakpoint by split read evidence. The corresponding environment variable is <code class="docutils literal"><span class="pre">MAVIS_MIN_NON_TARGET_ALIGNED_SPLIT_READS</span></code> and the default value is <code class="docutils literal"><span class="pre">1</span></code></dd>
<dt id="term-min-orf-size">min_orf_size</dt>
<dd><a class="reference external" href="https://docs.python.org/3.6/library/functions.html#int" title="(in Python v3.6)"><code class="xref py py-class docutils literal"><span class="pre">int</span></code></a> - The minimum length (in base pairs) to retain a putative open reading frame (orf). The corresponding environment variable is <code class="docutils literal"><span class="pre">MAVIS_MIN_ORF_SIZE</span></code> and the default value is <code class="docutils literal"><span class="pre">300</span></code></dd>
<dt id="term-min-sample-size-to-apply-percentage">min_sample_size_to_apply_percentage</dt>
<dd><a class="reference external" href="https://docs.python.org/3.6/library/functions.html#int" title="(in Python v3.6)"><code class="xref py py-class docutils literal"><span class="pre">int</span></code></a> - Minimum number of aligned bases to compute a match percent. if there are less than this number of aligned bases (match or mismatch) the percent comparator is not used. The corresponding environment variable is <code class="docutils literal"><span class="pre">MAVIS_MIN_SAMPLE_SIZE_TO_APPLY_PERCENTAGE</span></code> and the default value is <code class="docutils literal"><span class="pre">10</span></code></dd>
<dt id="term-min-softclipping">min_softclipping</dt>
<dd><a class="reference external" href="https://docs.python.org/3.6/library/functions.html#int" title="(in Python v3.6)"><code class="xref py py-class docutils literal"><span class="pre">int</span></code></a> - Minimum number of soft-clipped bases required for a read to be used as soft-clipped evidence. The corresponding environment variable is <code class="docutils literal"><span class="pre">MAVIS_MIN_SOFTCLIPPING</span></code> and the default value is <code class="docutils literal"><span class="pre">6</span></code></dd>
<dt id="term-min-spanning-reads-resolution">min_spanning_reads_resolution</dt>
<dd><a class="reference external" href="https://docs.python.org/3.6/library/functions.html#int" title="(in Python v3.6)"><code class="xref py py-class docutils literal"><span class="pre">int</span></code></a> - Minimum number of spanning reads required to call an event by spanning evidence. The corresponding environment variable is <code class="docutils literal"><span class="pre">MAVIS_MIN_SPANNING_READS_RESOLUTION</span></code> and the default value is <code class="docutils literal"><span class="pre">5</span></code></dd>
<dt id="term-min-splits-reads-resolution">min_splits_reads_resolution</dt>
<dd><a class="reference external" href="https://docs.python.org/3.6/library/functions.html#int" title="(in Python v3.6)"><code class="xref py py-class docutils literal"><span class="pre">int</span></code></a> - Minimum number of split reads required to call a breakpoint by split reads. The corresponding environment variable is <code class="docutils literal"><span class="pre">MAVIS_MIN_SPLITS_READS_RESOLUTION</span></code> and the default value is <code class="docutils literal"><span class="pre">3</span></code></dd>
<dt id="term-novel-exon-color">novel_exon_color</dt>
<dd><a class="reference external" href="https://docs.python.org/3.6/library/stdtypes.html#str" title="(in Python v3.6)"><code class="xref py py-class docutils literal"><span class="pre">str</span></code></a> - Novel exon fill color. The corresponding environment variable is <code class="docutils literal"><span class="pre">MAVIS_NOVEL_EXON_COLOR</span></code> and the default value is <code class="docutils literal"><span class="pre">'#000000'</span></code></dd>
<dt id="term-outer-window-min-event-size">outer_window_min_event_size</dt>
<dd><a class="reference external" href="https://docs.python.org/3.6/library/functions.html#int" title="(in Python v3.6)"><code class="xref py py-class docutils literal"><span class="pre">int</span></code></a> - The minimum size of an event in order for flanking read evidence to be collected. The corresponding environment variable is <code class="docutils literal"><span class="pre">MAVIS_OUTER_WINDOW_MIN_EVENT_SIZE</span></code> and the default value is <code class="docutils literal"><span class="pre">125</span></code></dd>
<dt id="term-queue">queue</dt>
<dd><a class="reference external" href="https://docs.python.org/3.6/library/stdtypes.html#str" title="(in Python v3.6)"><code class="xref py py-class docutils literal"><span class="pre">str</span></code></a> - The queue jobs are to be submitted to. The corresponding environment variable is <code class="docutils literal"><span class="pre">MAVIS_QUEUE</span></code> and the default value is <code class="docutils literal"><span class="pre">''</span></code></dd>
<dt id="term-scaffold-color">scaffold_color</dt>
<dd><a class="reference external" href="https://docs.python.org/3.6/library/stdtypes.html#str" title="(in Python v3.6)"><code class="xref py py-class docutils literal"><span class="pre">str</span></code></a> - The color used for the gene/transcripts scaffolds. The corresponding environment variable is <code class="docutils literal"><span class="pre">MAVIS_SCAFFOLD_COLOR</span></code> and the default value is <code class="docutils literal"><span class="pre">'#000000'</span></code></dd>
<dt id="term-scheduler">scheduler</dt>
<dd><a class="reference internal" href="auto/mavis.submit.html#mavis.submit.SCHEDULER" title="mavis.submit.SCHEDULER"><code class="xref py py-class docutils literal"><span class="pre">SCHEDULER</span></code></a> - The scheduler being used. The corresponding environment variable is <code class="docutils literal"><span class="pre">MAVIS_SCHEDULER</span></code> and the default value is <code class="docutils literal"><span class="pre">'SLURM'</span></code>. Accepted values include: <code class="docutils literal"><span class="pre">'SGE'</span></code>, <code class="docutils literal"><span class="pre">'SLURM'</span></code></dd>
<dt id="term-spanning-call-distance">spanning_call_distance</dt>
<dd><a class="reference external" href="https://docs.python.org/3.6/library/functions.html#int" title="(in Python v3.6)"><code class="xref py py-class docutils literal"><span class="pre">int</span></code></a> - The maximum distance allowed between breakpoint pairs (called by spanning reads) in order for them to pair. The corresponding environment variable is <code class="docutils literal"><span class="pre">MAVIS_SPANNING_CALL_DISTANCE</span></code> and the default value is <code class="docutils literal"><span class="pre">5</span></code></dd>
<dt id="term-splice-color">splice_color</dt>
<dd><a class="reference external" href="https://docs.python.org/3.6/library/stdtypes.html#str" title="(in Python v3.6)"><code class="xref py py-class docutils literal"><span class="pre">str</span></code></a> - Splicing lines color. The corresponding environment variable is <code class="docutils literal"><span class="pre">MAVIS_SPLICE_COLOR</span></code> and the default value is <code class="docutils literal"><span class="pre">'#000000'</span></code></dd>
<dt id="term-split-call-distance">split_call_distance</dt>
<dd><a class="reference external" href="https://docs.python.org/3.6/library/functions.html#int" title="(in Python v3.6)"><code class="xref py py-class docutils literal"><span class="pre">int</span></code></a> - The maximum distance allowed between breakpoint pairs (called by split reads) in order for them to pair. The corresponding environment variable is <code class="docutils literal"><span class="pre">MAVIS_SPLIT_CALL_DISTANCE</span></code> and the default value is <code class="docutils literal"><span class="pre">10</span></code></dd>
<dt id="term-stdev-count-abnormal">stdev_count_abnormal</dt>
<dd><a class="reference external" href="https://docs.python.org/3.6/library/functions.html#float" title="(in Python v3.6)"><code class="xref py py-class docutils literal"><span class="pre">float</span></code></a> - The number of standard deviations away from the normal considered expected and therefore not qualifying as flanking reads. The corresponding environment variable is <code class="docutils literal"><span class="pre">MAVIS_STDEV_COUNT_ABNORMAL</span></code> and the default value is <code class="docutils literal"><span class="pre">3.0</span></code></dd>
<dt id="term-strand-determining-read">strand_determining_read</dt>
<dd><a class="reference external" href="https://docs.python.org/3.6/library/functions.html#int" title="(in Python v3.6)"><code class="xref py py-class docutils literal"><span class="pre">int</span></code></a> - 1 or 2. the read in the pair which determines if (assuming a stranded protocol) the first or second read in the pair matches the strand sequenced. The corresponding environment variable is <code class="docutils literal"><span class="pre">MAVIS_STRAND_DETERMINING_READ</span></code> and the default value is <code class="docutils literal"><span class="pre">2</span></code></dd>
<dt id="term-time-limit">time_limit</dt>
<dd><a class="reference external" href="https://docs.python.org/3.6/library/functions.html#int" title="(in Python v3.6)"><code class="xref py py-class docutils literal"><span class="pre">int</span></code></a> - The time in seconds any given jobs is allowed. The corresponding environment variable is <code class="docutils literal"><span class="pre">MAVIS_TIME_LIMIT</span></code> and the default value is <code class="docutils literal"><span class="pre">57600</span></code></dd>
<dt id="term-trans-fetch-reads-limit">trans_fetch_reads_limit</dt>
<dd><code class="xref py py-class docutils literal"><span class="pre">nullable_int</span></code> - Related to <a class="reference internal" href="#term-fetch-reads-limit"><span class="xref std std-term">fetch_reads_limit</span></a>. overrides fetch_reads_limit for transcriptome libraries when set. if this has a value of none then fetch_reads_limit will be used for transcriptome libraries instead. The corresponding environment variable is <code class="docutils literal"><span class="pre">MAVIS_TRANS_FETCH_READS_LIMIT</span></code> and the default value is <code class="docutils literal"><span class="pre">12000</span></code></dd>
<dt id="term-trans-min-mapping-quality">trans_min_mapping_quality</dt>
<dd><code class="xref py py-class docutils literal"><span class="pre">nullable_int</span></code> - Related to <a class="reference internal" href="#term-min-mapping-quality"><span class="xref std std-term">min_mapping_quality</span></a>. overrides the min_mapping_quality if the library is a transcriptome and this is set to any number not none. if this value is none, min_mapping_quality is used for transcriptomes aswell as genomes. The corresponding environment variable is <code class="docutils literal"><span class="pre">MAVIS_TRANS_MIN_MAPPING_QUALITY</span></code> and the default value is <code class="docutils literal"><span class="pre">1</span></code></dd>
<dt id="term-trans-validation-memory">trans_validation_memory</dt>
<dd><a class="reference external" href="https://docs.python.org/3.6/library/functions.html#int" title="(in Python v3.6)"><code class="xref py py-class docutils literal"><span class="pre">int</span></code></a> - Default memory limit (mb) for the validation stage (for transcriptomes). The corresponding environment variable is <code class="docutils literal"><span class="pre">MAVIS_TRANS_VALIDATION_MEMORY</span></code> and the default value is <code class="docutils literal"><span class="pre">18000</span></code></dd>
<dt id="term-uninformative-filter">uninformative_filter</dt>
<dd><a class="reference external" href="https://docs.python.org/3.6/library/functions.html#bool" title="(in Python v3.6)"><code class="xref py py-class docutils literal"><span class="pre">bool</span></code></a> - Flag that determines if breakpoint pairs which are not within max_proximity to any annotations are filtered out prior to clustering. The corresponding environment variable is <code class="docutils literal"><span class="pre">MAVIS_UNINFORMATIVE_FILTER</span></code> and the default value is <code class="docutils literal"><span class="pre">False</span></code></dd>
<dt id="term-validation-memory">validation_memory</dt>
<dd><a class="reference external" href="https://docs.python.org/3.6/library/functions.html#int" title="(in Python v3.6)"><code class="xref py py-class docutils literal"><span class="pre">int</span></code></a> - Default memory limit (mb) for the validation stage. The corresponding environment variable is <code class="docutils literal"><span class="pre">MAVIS_VALIDATION_MEMORY</span></code> and the default value is <code class="docutils literal"><span class="pre">16000</span></code></dd>
<dt id="term-width">width</dt>
<dd><a class="reference external" href="https://docs.python.org/3.6/library/functions.html#int" title="(in Python v3.6)"><code class="xref py py-class docutils literal"><span class="pre">int</span></code></a> - The drawing width in pixels. The corresponding environment variable is <code class="docutils literal"><span class="pre">MAVIS_WIDTH</span></code> and the default value is <code class="docutils literal"><span class="pre">1000</span></code></dd>
<dt id="term-write-evidence-files">write_evidence_files</dt>
<dd><a class="reference external" href="https://docs.python.org/3.6/library/functions.html#bool" title="(in Python v3.6)"><code class="xref py py-class docutils literal"><span class="pre">bool</span></code></a> - Write the intermediate bam and bed files containing the raw evidence collected and contigs aligned. not required for subsequent steps but can be useful in debugging and deep investigation of events. The corresponding environment variable is <code class="docutils literal"><span class="pre">MAVIS_WRITE_EVIDENCE_FILES</span></code> and the default value is <code class="docutils literal"><span class="pre">True</span></code></dd>
</dl>
</div>
<div class="section" id="column-names">
<h2>Column Names<a class="headerlink" href="#column-names" title="Permalink to this headline">¶</a></h2>
<p>List of column names and their definitions. The types indicated here are the expected types in a row for a given column name.</p>
<dl class="glossary docutils">
<dt id="term-annotation-figure">annotation_figure</dt>
<dd><code class="docutils literal"><span class="pre">FILEPATH</span></code> - File path to the svg drawing representing the annotation</dd>
<dt id="term-annotation-figure-legend">annotation_figure_legend</dt>
<dd><code class="docutils literal"><span class="pre">JSON</span></code> - <a class="reference internal" href="#term-json"><span class="xref std std-term">JSON</span></a> data for the figure legend</dd>
<dt id="term-annotation-id">annotation_id</dt>
<dd>Identifier for the annotation step</dd>
<dt id="term-break1-chromosome">break1_chromosome</dt>
<dd><a class="reference external" href="https://docs.python.org/3.6/library/stdtypes.html#str" title="(in Python v3.6)"><code class="xref py py-class docutils literal"><span class="pre">str</span></code></a> - The name of the chromosome on which breakpoint 1 is situated</dd>
<dt id="term-break1-ewindow">break1_ewindow</dt>
<dd><code class="docutils literal"><span class="pre">int-int</span></code> - Window where evidence was gathered for the first breakpoint</dd>
<dt id="term-break1-ewindow-count">break1_ewindow_count</dt>
<dd><a class="reference external" href="https://docs.python.org/3.6/library/functions.html#int" title="(in Python v3.6)"><code class="xref py py-class docutils literal"><span class="pre">int</span></code></a> - Number of reads processed/looked-at in the first evidence window</dd>
<dt id="term-break1-ewindow-practical-coverage">break1_ewindow_practical_coverage</dt>
<dd><a class="reference external" href="https://docs.python.org/3.6/library/functions.html#float" title="(in Python v3.6)"><code class="xref py py-class docutils literal"><span class="pre">float</span></code></a> - break2_ewindow_practical_coverage, break1_ewindow_count / len(break1_ewindow). Not the actual
coverage as bins are sampled within and there is a read limit cutoff</dd>
<dt id="term-break1-homologous-seq">break1_homologous_seq</dt>
<dd><a class="reference external" href="https://docs.python.org/3.6/library/stdtypes.html#str" title="(in Python v3.6)"><code class="xref py py-class docutils literal"><span class="pre">str</span></code></a> - Sequence in common at the first breakpoint and other side of the second breakpoint</dd>
<dt id="term-break1-orientation">break1_orientation</dt>
<dd><a class="reference internal" href="auto/mavis.constants.html#mavis.constants.ORIENT" title="mavis.constants.ORIENT"><code class="xref py py-class docutils literal"><span class="pre">ORIENT</span></code></a> - The side of the breakpoint wrt the positive/forward strand that is retained.</dd>
<dt id="term-break1-position-end">break1_position_end</dt>
<dd><a class="reference external" href="https://docs.python.org/3.6/library/functions.html#int" title="(in Python v3.6)"><code class="xref py py-class docutils literal"><span class="pre">int</span></code></a> - End integer inclusive 1-based of the range representing breakpoint 1</dd>
<dt id="term-break1-position-start">break1_position_start</dt>
<dd><a class="reference external" href="https://docs.python.org/3.6/library/functions.html#int" title="(in Python v3.6)"><code class="xref py py-class docutils literal"><span class="pre">int</span></code></a> - Start integer inclusive 1-based of the range representing breakpoint 1</dd>
<dt id="term-break1-seq">break1_seq</dt>
<dd><a class="reference external" href="https://docs.python.org/3.6/library/stdtypes.html#str" title="(in Python v3.6)"><code class="xref py py-class docutils literal"><span class="pre">str</span></code></a> - The sequence up to and including the breakpoint. Always given wrt to the positive/forward strand</dd>
<dt id="term-break1-split-reads">break1_split_reads</dt>
<dd><a class="reference external" href="https://docs.python.org/3.6/library/functions.html#int" title="(in Python v3.6)"><code class="xref py py-class docutils literal"><span class="pre">int</span></code></a> - Number of split reads that call the exact breakpoint given</dd>
<dt id="term-break1-split-reads-forced">break1_split_reads_forced</dt>
<dd><a class="reference external" href="https://docs.python.org/3.6/library/functions.html#int" title="(in Python v3.6)"><code class="xref py py-class docutils literal"><span class="pre">int</span></code></a> - Number of split reads which were aligned to the opposite breakpoint window using a targeted
alignment</dd>
<dt id="term-break1-strand">break1_strand</dt>
<dd><a class="reference internal" href="auto/mavis.constants.html#mavis.constants.STRAND" title="mavis.constants.STRAND"><code class="xref py py-class docutils literal"><span class="pre">STRAND</span></code></a> - The strand wrt to the reference positive/forward strand at this breakpoint.</dd>
<dt id="term-break2-chromosome">break2_chromosome</dt>
<dd>The name of the chromosome on which breakpoint 2 is situated</dd>
<dt id="term-break2-ewindow">break2_ewindow</dt>
<dd><code class="docutils literal"><span class="pre">int-int</span></code> - Window where evidence was gathered for the second breakpoint</dd>
<dt id="term-break2-ewindow-count">break2_ewindow_count</dt>
<dd><a class="reference external" href="https://docs.python.org/3.6/library/functions.html#int" title="(in Python v3.6)"><code class="xref py py-class docutils literal"><span class="pre">int</span></code></a> - Number of reads processed/looked-at in the second evidence window</dd>
<dt id="term-break2-ewindow-practical-coverage">break2_ewindow_practical_coverage</dt>
<dd><a class="reference external" href="https://docs.python.org/3.6/library/functions.html#float" title="(in Python v3.6)"><code class="xref py py-class docutils literal"><span class="pre">float</span></code></a> - break2_ewindow_practical_coverage, break2_ewindow_count / len(break2_ewindow). Not the actual
coverage as bins are sampled within and there is a read limit cutoff</dd>
<dt id="term-break2-homologous-seq">break2_homologous_seq</dt>
<dd><a class="reference external" href="https://docs.python.org/3.6/library/stdtypes.html#str" title="(in Python v3.6)"><code class="xref py py-class docutils literal"><span class="pre">str</span></code></a> - Sequence in common at the second breakpoint and other side of the first breakpoint</dd>
<dt id="term-break2-orientation">break2_orientation</dt>
<dd><a class="reference internal" href="auto/mavis.constants.html#mavis.constants.ORIENT" title="mavis.constants.ORIENT"><code class="xref py py-class docutils literal"><span class="pre">ORIENT</span></code></a> - The side of the breakpoint wrt the positive/forward strand that is retained.</dd>
<dt id="term-break2-position-end">break2_position_end</dt>
<dd><a class="reference external" href="https://docs.python.org/3.6/library/functions.html#int" title="(in Python v3.6)"><code class="xref py py-class docutils literal"><span class="pre">int</span></code></a> - End integer inclusive 1-based of the range representing breakpoint 2</dd>
<dt id="term-break2-position-start">break2_position_start</dt>
<dd><a class="reference external" href="https://docs.python.org/3.6/library/functions.html#int" title="(in Python v3.6)"><code class="xref py py-class docutils literal"><span class="pre">int</span></code></a> - Start integer inclusive 1-based of the range representing breakpoint 2</dd>
<dt id="term-break2-seq">break2_seq</dt>
<dd><a class="reference external" href="https://docs.python.org/3.6/library/stdtypes.html#str" title="(in Python v3.6)"><code class="xref py py-class docutils literal"><span class="pre">str</span></code></a> - The sequence up to and including the breakpoint. Always given wrt to the positive/forward strand</dd>
<dt id="term-break2-split-reads">break2_split_reads</dt>
<dd><a class="reference external" href="https://docs.python.org/3.6/library/functions.html#int" title="(in Python v3.6)"><code class="xref py py-class docutils literal"><span class="pre">int</span></code></a> - Number of split reads that call the exact breakpoint given</dd>
<dt id="term-break2-split-reads-forced">break2_split_reads_forced</dt>
<dd><a class="reference external" href="https://docs.python.org/3.6/library/functions.html#int" title="(in Python v3.6)"><code class="xref py py-class docutils literal"><span class="pre">int</span></code></a> - Number of split reads which were aligned to the opposite breakpoint window using a targeted
alignment</dd>
<dt id="term-break2-strand">break2_strand</dt>
<dd><a class="reference internal" href="auto/mavis.constants.html#mavis.constants.STRAND" title="mavis.constants.STRAND"><code class="xref py py-class docutils literal"><span class="pre">STRAND</span></code></a> - The strand wrt to the reference positive/forward strand at this breakpoint.</dd>
<dt id="term-call-method">call_method</dt>
<dd><a class="reference internal" href="auto/mavis.constants.html#mavis.constants.CALL_METHOD" title="mavis.constants.CALL_METHOD"><code class="xref py py-class docutils literal"><span class="pre">CALL_METHOD</span></code></a> - The method used to call the breakpoints</dd>
<dt id="term-cdna-synon">cdna_synon</dt>
<dd>semi-colon delimited list of transcript ids which have an identical cdna sequence to the cdna sequence of the
current fusion product</dd>
<dt id="term-cluster-id">cluster_id</dt>
<dd>Identifier for the merging/clustering step</dd>
<dt id="term-cluster-size">cluster_size</dt>
<dd><a class="reference external" href="https://docs.python.org/3.6/library/functions.html#int" title="(in Python v3.6)"><code class="xref py py-class docutils literal"><span class="pre">int</span></code></a> - The number of breakpoint pair calls that were grouped in creating the cluster</dd>
<dt id="term-contig-alignment-cigar">contig_alignment_cigar</dt>
<dd>The cigar string(s) representing the contig alignment. Semi-colon delimited</dd>
<dt id="term-contig-alignment-query-name">contig_alignment_query_name</dt>
<dd>The query name for the contig alignment. Should match the ‘read’ name(s) in the .contigs.bam output file</dd>
<dt id="term-contig-alignment-reference-start">contig_alignment_reference_start</dt>
<dd>The reference start(s) &lt;chr&gt;:&lt;position&gt; of the contig alignment. Semi-colon delimited</dd>
<dt id="term-contig-alignment-score">contig_alignment_score</dt>
<dd><a class="reference external" href="https://docs.python.org/3.6/library/functions.html#float" title="(in Python v3.6)"><code class="xref py py-class docutils literal"><span class="pre">float</span></code></a> - A rank based on the alignment tool blat etc. of the alignment being used. An average if
split alignments were used. Lower numbers indicate a better alignment. If it was the best alignment possible
then this would be zero.</dd>
<dt id="term-contig-build-score">contig_build_score</dt>
<dd><a class="reference external" href="https://docs.python.org/3.6/library/functions.html#int" title="(in Python v3.6)"><code class="xref py py-class docutils literal"><span class="pre">int</span></code></a> - Score representing the edge weights of all edges used in building the sequence</dd>
<dt id="term-contig-remap-coverage">contig_remap_coverage</dt>
<dd><a class="reference external" href="https://docs.python.org/3.6/library/functions.html#float" title="(in Python v3.6)"><code class="xref py py-class docutils literal"><span class="pre">float</span></code></a> - Fraction of the contig sequence which is covered by the remapped reads</dd>
<dt id="term-contig-remap-score">contig_remap_score</dt>
<dd><a class="reference external" href="https://docs.python.org/3.6/library/functions.html#float" title="(in Python v3.6)"><code class="xref py py-class docutils literal"><span class="pre">float</span></code></a> - Score representing the number of sequences from the set of sequences given to the assembly
algorithm that were aligned to the resulting contig with an acceptable scoring based on user-set thresholds.
For any sequence its contribution to the score is divided by the number of mappings to give less weight to
multimaps</dd>
<dt id="term-contig-remapped-read-names">contig_remapped_read_names</dt>
<dd>read query names for the reads that were remapped. A -1 or -2 has been appended to the end of the name to
indicate if this is the first or second read in the pair</dd>
<dt id="term-contig-remapped-reads">contig_remapped_reads</dt>
<dd><a class="reference external" href="https://docs.python.org/3.6/library/functions.html#int" title="(in Python v3.6)"><code class="xref py py-class docutils literal"><span class="pre">int</span></code></a> - the number of reads from the input bam that map to the assembled contig</dd>
<dt id="term-contig-seq">contig_seq</dt>
<dd><a class="reference external" href="https://docs.python.org/3.6/library/stdtypes.html#str" title="(in Python v3.6)"><code class="xref py py-class docutils literal"><span class="pre">str</span></code></a> - Sequence of the current contig wrt to the positive forward strand if not strand specific</dd>
<dt id="term-contig-strand-specific">contig_strand_specific</dt>
<dd><a class="reference external" href="https://docs.python.org/3.6/library/functions.html#bool" title="(in Python v3.6)"><code class="xref py py-class docutils literal"><span class="pre">bool</span></code></a> - A flag to indicate if it was possible to resolve the strand for this contig</dd>
<dt id="term-contigs-aligned">contigs_aligned</dt>
<dd><a class="reference external" href="https://docs.python.org/3.6/library/functions.html#int" title="(in Python v3.6)"><code class="xref py py-class docutils literal"><span class="pre">int</span></code></a> - Number of contigs that were able to align</dd>
<dt id="term-contigs-assembled">contigs_assembled</dt>
<dd><a class="reference external" href="https://docs.python.org/3.6/library/functions.html#int" title="(in Python v3.6)"><code class="xref py py-class docutils literal"><span class="pre">int</span></code></a> - Number of contigs that were built from split read sequences</dd>
<dt id="term-131">event_type</dt>
<dd><a class="reference internal" href="auto/mavis.constants.html#mavis.constants.SVTYPE" title="mavis.constants.SVTYPE"><code class="xref py py-class docutils literal"><span class="pre">SVTYPE</span></code></a> - The classification of the event</dd>
<dt id="term-flanking-median-fragment-size">flanking_median_fragment_size</dt>
<dd><a class="reference external" href="https://docs.python.org/3.6/library/functions.html#int" title="(in Python v3.6)"><code class="xref py py-class docutils literal"><span class="pre">int</span></code></a> - The median fragment size of the flanking reads being used as evidence</dd>
<dt id="term-flanking-pairs">flanking_pairs</dt>
<dd><a class="reference external" href="https://docs.python.org/3.6/library/functions.html#int" title="(in Python v3.6)"><code class="xref py py-class docutils literal"><span class="pre">int</span></code></a> - Number of read-pairs where one read aligns to the first breakpoint window and the second read
aligns to the other. The count here is based on the number of unique query names</dd>
<dt id="term-flanking-pairs-compatible">flanking_pairs_compatible</dt>
<dd><a class="reference external" href="https://docs.python.org/3.6/library/functions.html#int" title="(in Python v3.6)"><code class="xref py py-class docutils literal"><span class="pre">int</span></code></a> - Number of flanking pairs of a compatible orientation type. This applies to insertions and
duplications. Flanking pairs supporting an insertion will be compatible to a duplication and flanking pairs
supporting a duplication will be compatible to an insertion (possibly indicating an internal translocation)</dd>
<dt id="term-flanking-stdev-fragment-size">flanking_stdev_fragment_size</dt>
<dd><a class="reference external" href="https://docs.python.org/3.6/library/functions.html#float" title="(in Python v3.6)"><code class="xref py py-class docutils literal"><span class="pre">float</span></code></a> - The standard deviation in fragment size of the flanking reads being used as evidence</dd>
<dt id="term-fusion-cdna-coding-end">fusion_cdna_coding_end</dt>
<dd>Position wrt the 5’ end of the fusion transcript where coding ends last base of the stop codon</dd>
<dt id="term-146">fusion_cdna_coding_end</dt>
<dd><a class="reference external" href="https://docs.python.org/3.6/library/functions.html#int" title="(in Python v3.6)"><code class="xref py py-class docutils literal"><span class="pre">int</span></code></a> - Position wrt the 5’ end of the fusion transcript where coding ends last base of the stop codon</dd>
<dt id="term-fusion-cdna-coding-start">fusion_cdna_coding_start</dt>
<dd><a class="reference external" href="https://docs.python.org/3.6/library/functions.html#int" title="(in Python v3.6)"><code class="xref py py-class docutils literal"><span class="pre">int</span></code></a> - Position wrt the 5’ end of the fusion transcript where coding begins first base of the Met amino acid.</dd>
<dt id="term-fusion-mapped-domains">fusion_mapped_domains</dt>
<dd><code class="docutils literal"><span class="pre">JSON</span></code> - List of domains in <a class="reference internal" href="#term-json"><span class="xref std std-term">JSON</span></a> format where each domain start and end positions are given wrt to the fusion
transcript and the mapping quality is the number of matching amino acid positions over the total
number of amino acids. The sequence is the amino acid sequence of the domain on the reference/original
transcript</dd>
<dt id="term-fusion-protein-hgvs">fusion_protein_hgvs</dt>
<dd><a class="reference external" href="https://docs.python.org/3.6/library/stdtypes.html#str" title="(in Python v3.6)"><code class="xref py py-class docutils literal"><span class="pre">str</span></code></a> - Describes the fusion protein in HGVS notation. Will be None if the change is not an indel or is synonymous</dd>
<dt id="term-fusion-sequence-fasta-file">fusion_sequence_fasta_file</dt>
<dd><code class="docutils literal"><span class="pre">FILEPATH</span></code> - Path to the corresponding fasta output file</dd>
<dt id="term-fusion-sequence-fasta-id">fusion_sequence_fasta_id</dt>
<dd>The sequence identifier for the cdna sequence output fasta file</dd>
<dt id="term-fusion-splicing-pattern">fusion_splicing_pattern</dt>
<dd><code class="xref py py-class docutils literal"><span class="pre">SPLICE_TYPE</span></code> - Type of splicing pattern used to create the fusion cDNA.</dd>
<dt id="term-gene1">gene1</dt>
<dd>Gene for the current annotation at the first breakpoint</dd>
<dt id="term-gene1-aliases">gene1_aliases</dt>
<dd>Other gene names associated with the current annotation at the first breakpoint</dd>
<dt id="term-gene1-direction">gene1_direction</dt>
<dd><a class="reference internal" href="auto/mavis.constants.html#mavis.constants.PRIME" title="mavis.constants.PRIME"><code class="xref py py-class docutils literal"><span class="pre">PRIME</span></code></a> - The direction/prime of the gene</dd>
<dt id="term-gene2">gene2</dt>
<dd>Gene for the current annotation at the second breakpoint</dd>
<dt id="term-gene2-aliases">gene2_aliases</dt>
<dd>Other gene names associated with the current annotation at the second breakpoint</dd>
<dt id="term-gene2-direction">gene2_direction</dt>
<dd><a class="reference internal" href="auto/mavis.constants.html#mavis.constants.PRIME" title="mavis.constants.PRIME"><code class="xref py py-class docutils literal"><span class="pre">PRIME</span></code></a> - The direction/prime of the gene. Has the following possible values</dd>
<dt id="term-gene-product-type">gene_product_type</dt>
<dd><a class="reference internal" href="auto/mavis.constants.html#mavis.constants.GENE_PRODUCT_TYPE" title="mavis.constants.GENE_PRODUCT_TYPE"><code class="xref py py-class docutils literal"><span class="pre">GENE_PRODUCT_TYPE</span></code></a> - Describes if the putative fusion product will be sense or anti-sense</dd>
<dt id="term-genes-encompassed">genes_encompassed</dt>
<dd>Applies to intrachromosomal events only. List of genes which overlap any region that occurs between both
breakpoints. For example in a deletion event these would be deleted genes.</dd>
<dt id="term-genes-overlapping-break1">genes_overlapping_break1</dt>
<dd>list of genes which overlap the first breakpoint</dd>
<dt id="term-genes-overlapping-break2">genes_overlapping_break2</dt>
<dd>list of genes which overlap the second breakpoint</dd>
<dt id="term-genes-proximal-to-break1">genes_proximal_to_break1</dt>
<dd>list of genes near the breakpoint and the distance away from the breakpoint</dd>
<dt id="term-genes-proximal-to-break2">genes_proximal_to_break2</dt>
<dd>list of genes near the breakpoint and the distance away from the breakpoint</dd>
<dt id="term-inferred-pairing">inferred_pairing</dt>
<dd>A semi colon delimited of event identifiers i.e. &lt;annotation_id&gt;_&lt;splicing pattern&gt;_&lt;cds start&gt;_&lt;cds end&gt;
which were paired to the current event based on predicted products</dd>
<dt id="term-library">library</dt>
<dd>Identifier for the library/source</dd>
<dt id="term-linking-split-reads">linking_split_reads</dt>
<dd><a class="reference external" href="https://docs.python.org/3.6/library/functions.html#int" title="(in Python v3.6)"><code class="xref py py-class docutils literal"><span class="pre">int</span></code></a> - Number of split reads that align to both breakpoints</dd>
<dt id="term-net-size">net_size</dt>
<dd><code class="docutils literal"><span class="pre">int-int</span></code> - The net size of an event. For translocations and inversion this will always be 0. For indels it will be
negative for deletions and positive for insertions. It is a range to accommodate non-specific events.</dd>
<dt id="term-opposing-strands">opposing_strands</dt>
<dd><a class="reference external" href="https://docs.python.org/3.6/library/functions.html#bool" title="(in Python v3.6)"><code class="xref py py-class docutils literal"><span class="pre">bool</span></code></a> - Specifies if breakpoints are on opposite strands wrt to the reference. Expects a boolean</dd>
<dt id="term-pairing">pairing</dt>
<dd>A semi colon delimited of event identifiers i.e. &lt;annotation_id&gt;_&lt;splicing pattern&gt;_&lt;cds start&gt;_&lt;cds end&gt;
which were paired to the current event based on breakpoint positions</dd>
<dt id="term-product-id">product_id</dt>
<dd>Unique identifier of the final fusion including splicing and ORF decision from the annotation step</dd>
<dt id="term-protein-synon">protein_synon</dt>
<dd>semi-colon delimited list of transcript ids which produce a translation with an identical amino-acid sequence
to the current fusion product</dd>
<dt id="term-protocol">protocol</dt>
<dd><a class="reference internal" href="auto/mavis.constants.html#mavis.constants.PROTOCOL" title="mavis.constants.PROTOCOL"><code class="xref py py-class docutils literal"><span class="pre">PROTOCOL</span></code></a> - Specifies the type of library</dd>
<dt id="term-raw-break1-split-reads">raw_break1_split_reads</dt>
<dd><a class="reference external" href="https://docs.python.org/3.6/library/functions.html#int" title="(in Python v3.6)"><code class="xref py py-class docutils literal"><span class="pre">int</span></code></a> - Number of split reads before calling the breakpoint</dd>
<dt id="term-raw-break2-split-reads">raw_break2_split_reads</dt>
<dd><a class="reference external" href="https://docs.python.org/3.6/library/functions.html#int" title="(in Python v3.6)"><code class="xref py py-class docutils literal"><span class="pre">int</span></code></a> - Number of split reads before calling the breakpoint</dd>
<dt id="term-raw-flanking-pairs">raw_flanking_pairs</dt>
<dd><a class="reference external" href="https://docs.python.org/3.6/library/functions.html#int" title="(in Python v3.6)"><code class="xref py py-class docutils literal"><span class="pre">int</span></code></a> - Number of flanking reads before calling the breakpoint. The count here is based on the number of
unique query names</dd>
<dt id="term-raw-spanning-reads">raw_spanning_reads</dt>
<dd><a class="reference external" href="https://docs.python.org/3.6/library/functions.html#int" title="(in Python v3.6)"><code class="xref py py-class docutils literal"><span class="pre">int</span></code></a> - Number of spanning reads collected during evidence collection before calling the breakpoint</dd>
<dt id="term-spanning-read-names">spanning_read_names</dt>
<dd>read query names of the spanning reads which support the current event</dd>
<dt id="term-spanning-reads">spanning_reads</dt>
<dd><a class="reference external" href="https://docs.python.org/3.6/library/functions.html#int" title="(in Python v3.6)"><code class="xref py py-class docutils literal"><span class="pre">int</span></code></a> - the number of spanning reads which support the event</dd>
<dt id="term-stranded">stranded</dt>
<dd><a class="reference external" href="https://docs.python.org/3.6/library/functions.html#bool" title="(in Python v3.6)"><code class="xref py py-class docutils literal"><span class="pre">bool</span></code></a> - Specifies if the sequencing protocol was strand specific or not. Expects a boolean</dd>
<dt id="term-supplementary-call">supplementary_call</dt>
<dd><a class="reference external" href="https://docs.python.org/3.6/library/functions.html#bool" title="(in Python v3.6)"><code class="xref py py-class docutils literal"><span class="pre">bool</span></code></a> - Flag to indicate if the current event was a supplementary call, meaning a call that was found as a result of
validating another event.</dd>
<dt id="term-tools">tools</dt>
<dd>The tools that called the event originally from the cluster step. Should be a semi-colon delimited list of
&lt;tool name&gt;_&lt;tool version&gt;</dd>
<dt id="term-tracking-id">tracking_id</dt>
<dd>column used to store input identifiers from the original SV calls. Used to track calls from the input files to
the final outputs.</dd>
<dt id="term-transcript1">transcript1</dt>
<dd>Transcript for the current annotation at the first breakpoint</dd>
<dt id="term-transcript2">transcript2</dt>
<dd>Transcript for the current annotation at the second breakpoint</dd>
<dt id="term-untemplated-seq">untemplated_seq</dt>
<dd><a class="reference external" href="https://docs.python.org/3.6/library/stdtypes.html#str" title="(in Python v3.6)"><code class="xref py py-class docutils literal"><span class="pre">str</span></code></a> - The untemplated/novel sequence between the breakpoints</dd>
<dt id="term-validation-id">validation_id</dt>
<dd>Identifier for the validation step</dd>
</dl>
</div>
</div>


           </div>
           <div class="articleComments">
            
           </div>
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="citations.html" class="btn btn-neutral float-right" title="Literature" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="auto/mavis.util.html" class="btn btn-neutral" title="util module" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2017, creisle.

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  

    <script type="text/javascript">
        var DOCUMENTATION_OPTIONS = {
            URL_ROOT:'./',
            VERSION:'',
            LANGUAGE:'None',
            COLLAPSE_INDEX:false,
            FILE_SUFFIX:'.html',
            HAS_SOURCE:  true,
            SOURCELINK_SUFFIX: '.txt'
        };
    </script>
      <script type="text/javascript" src="_static/jquery.js"></script>
      <script type="text/javascript" src="_static/underscore.js"></script>
      <script type="text/javascript" src="_static/doctools.js"></script>
      <script type="text/javascript" src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>

  

  
  
    <script type="text/javascript" src="_static/js/theme.js"></script>
  

  
  
  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.StickyNav.enable();
      });
  </script>
   

</body>
</html>