

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Reference Input Files &mdash; MAVIS  documentation</title>
  

  
  
  
  

  

  
  
    

  

  
  
    <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  

  
    <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  

  
        <link rel="author" title="About these documents"
              href="about.html"/>
        <link rel="index" title="Index"
              href="genindex.html"/>
        <link rel="search" title="Search" href="search.html"/>
    <link rel="top" title="MAVIS  documentation" href="index.html"/>
        <link rel="next" title="Configuration and Settings" href="configuration.html"/>
        <link rel="prev" title="MAVIS standard input file format" href="mavis_input.html"/> 

  
  <script src="_static/js/modernizr.min.js"></script>

</head>

<body class="wy-body-for-nav" role="document">

   
  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search">
          

          
            <a href="index.html" class="icon icon-home"> MAVIS
          

          
          </a>

          
            
            
              <div class="version">
                1.6.4
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="about.html">About</a></li>
<li class="toctree-l1"><a class="reference internal" href="pipeline.html">Running the Pipeline</a></li>
<li class="toctree-l1"><a class="reference internal" href="mavis_input.html">MAVIS standard input file format</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Reference Input Files</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#reference-genome">Reference Genome</a></li>
<li class="toctree-l2"><a class="reference internal" href="#template-metadata">Template Metadata</a></li>
<li class="toctree-l2"><a class="reference internal" href="#masking-file">Masking File</a></li>
<li class="toctree-l2"><a class="reference internal" href="#annotations">Annotations</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#generating-the-annotations-from-ensembl">Generating the Annotations from <code class="docutils literal"><span class="pre">Ensembl</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#dgv-database-of-genomic-variants-annotations"><code class="docutils literal"><span class="pre">DGV</span> <span class="pre">(Database</span> <span class="pre">of</span> <span class="pre">Genomic</span> <span class="pre">Variants)</span></code> Annotations</a></li>
<li class="toctree-l2"><a class="reference internal" href="#aligner-reference">Aligner Reference</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="configuration.html">Configuration and Settings</a></li>
<li class="toctree-l1"><a class="reference internal" href="tools_supported.html">Supported SV Callers</a></li>
<li class="toctree-l1"><a class="reference internal" href="illustrations.html">Illustrations</a></li>
<li class="toctree-l1"><a class="reference internal" href="faqs.html">Frequently Asked Questions</a></li>
<li class="toctree-l1"><a class="reference internal" href="theory.html">Theory and Models</a></li>
<li class="toctree-l1"><a class="reference internal" href="development.html">Guidelines for Contributors</a></li>
<li class="toctree-l1"><a class="reference internal" href="package.html">Package Documentation</a></li>
<li class="toctree-l1"><a class="reference internal" href="glossary.html">Glossary</a></li>
<li class="toctree-l1"><a class="reference internal" href="citations.html">Literature</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">MAVIS</a>
        
      </nav>


      
      <div class="wy-nav-content">
        <div class="rst-content">
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html">Docs</a> &raquo;</li>
        
      <li>Reference Input Files</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="_sources/reference.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="reference-input-files">
<span id="reference-input"></span><h1>Reference Input Files<a class="headerlink" href="#reference-input-files" title="Permalink to this headline">Â¶</a></h1>
<p>There are several reference files that are required for full functionality of the MAVIS pipeline. If the same
reference file will be reused often then the user may find it helpful to set reasonable defaults. Default values
for any of the reference file arguments can be <a class="reference internal" href="configuration.html#config-environment"><span class="std std-ref">configured through environment variables</span></a>.</p>
<p>To improve the install experience for the users, different configurations of the MAVIS annotations file have been made available. These files can be downloaded below, or if the required configuration is not available, <a class="reference internal" href="#generate-reference-annotations"><span class="std std-ref">instructions on generating the annotations file</span></a> can be found below.</p>
<table border="1" class="docutils">
<colgroup>
<col width="33%" />
<col width="33%" />
<col width="33%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">File Name (Type/Format)</th>
<th class="head">Environment Variable</th>
<th class="head">Download</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td><a class="reference internal" href="#reference-files-reference-genome"><span class="std std-ref">reference genome</span></a> (<a class="reference internal" href="glossary.html#term-fasta"><span class="xref std std-term">fasta</span></a>)</td>
<td><code class="docutils literal"><span class="pre">MAVIS_REFERENCE_GENOME</span></code></td>
<td><div class="first last"><a class='download-button btn btn-neutral' href='http://hgdownload.cse.ucsc.edu/goldenPath/hg19/bigZips/chromFa.tar.gz' download>
    <img src='_static/Ic_cloud_download_48px.svg'>GRCh37/Hg19
</a><br>
<a class='download-button btn btn-neutral' href='http://hgdownload.cse.ucsc.edu/goldenPath/hg38/bigZips/hg38.fa.gz' download>
    <img src='_static/Ic_cloud_download_48px.svg'>GRCh38
</a></div></td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="#reference-files-annotations"><span class="std std-ref">annotations</span></a> (<a class="reference internal" href="glossary.html#term-json"><span class="xref std std-term">JSON</span></a>)</td>
<td><code class="docutils literal"><span class="pre">MAVIS_ANNOTATIONS</span></code></td>
<td><div class="first last"><a class='download-button btn btn-neutral' href='http://www.bcgsc.ca/downloads/mavis/ensembl69_hg19_annotations.json' download>
    <img src='_static/Ic_cloud_download_48px.svg'>GRCh37/Hg19 + Ensembl69
</a><br>
<a class='download-button btn btn-neutral' href='http://www.bcgsc.ca/downloads/mavis/ensembl79_hg38_annotations.json' download>
    <img src='_static/Ic_cloud_download_48px.svg'>GRCh38 + Ensembl79
</a></div></td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="#reference-files-masking"><span class="std std-ref">masking</span></a> (text/tabbed)</td>
<td><code class="docutils literal"><span class="pre">MAVIS_MASKING</span></code></td>
<td><div class="first last"><a class='download-button btn btn-neutral' href='http://www.bcgsc.ca/downloads/mavis/hg19_masking.tab' download>
    <img src='_static/Ic_cloud_download_48px.svg'>GRCh37/Hg19
</a><br>
<a class='download-button btn btn-neutral' href='http://www.bcgsc.ca/downloads/mavis/GRCh38_masking.tab' download>
    <img src='_static/Ic_cloud_download_48px.svg'>GRCh38
</a></div></td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="#reference-files-template-metadata"><span class="std std-ref">template metadata</span></a> (text/tabbed)</td>
<td><code class="docutils literal"><span class="pre">MAVIS_TEMPLATE_METADATA</span></code></td>
<td><div class="first last"><a class='download-button btn btn-neutral' href='http://hgdownload.cse.ucsc.edu/goldenPath/hg19/database/cytoBand.txt.gz' download>
    <img src='_static/Ic_cloud_download_48px.svg'>GRCh37/Hg19
</a><br>
<a class='download-button btn btn-neutral' href='http://hgdownload.cse.ucsc.edu/goldenPath/hg38/database/cytoBand.txt.gz' download>
    <img src='_static/Ic_cloud_download_48px.svg'>GRCh38
</a></div></td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="#reference-files-dgv-annotations"><span class="std std-ref">DGV annotations</span></a> (text/tabbed)</td>
<td><code class="docutils literal"><span class="pre">MAVIS_DGV_ANNOTATION</span></code></td>
<td><div class="first last"><a class='download-button btn btn-neutral' href='http://www.bcgsc.ca/downloads/mavis/dgv_hg19_variants.tab' download>
    <img src='_static/Ic_cloud_download_48px.svg'>GRCh37/Hg19
</a><br>
<a class='download-button btn btn-neutral' href='http://www.bcgsc.ca/downloads/mavis/dgv_hg38_variants.tab' download>
    <img src='_static/Ic_cloud_download_48px.svg'>GRCh38
</a></div></td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="#reference-files-aligner-reference"><span class="std std-ref">aligner reference</span></a></td>
<td><code class="docutils literal"><span class="pre">MAVIS_ALIGNER_REFERENCE</span></code></td>
<td><div class="first last"><a class='download-button btn btn-neutral' href='http://hgdownload.cse.ucsc.edu/goldenPath/hg19/bigZips/hg19.2bit' download>
    <img src='_static/Ic_cloud_download_48px.svg'>GRCh37/Hg19 2bit (blat)
</a><br>
<a class='download-button btn btn-neutral' href='http://hgdownload.cse.ucsc.edu/goldenPath/hg38/bigZips/hg38.2bit' download>
    <img src='_static/Ic_cloud_download_48px.svg'>GRCh38 2bit (blat)
</a></div></td>
</tr>
</tbody>
</table>
<p>If the environment variables above are set they will be used as the default values when any step of the pipeline
script is called (including generating the template config file)</p>
<div class="section" id="reference-genome">
<span id="reference-files-reference-genome"></span><h2>Reference Genome<a class="headerlink" href="#reference-genome" title="Permalink to this headline">Â¶</a></h2>
<p>These are the sequence files in fasta format that are used in aligning and generating the fusion sequences.</p>
</div>
<div class="section" id="template-metadata">
<span id="reference-files-template-metadata"></span><h2>Template Metadata<a class="headerlink" href="#template-metadata" title="Permalink to this headline">Â¶</a></h2>
<p>This is the file which contains the band information for the chromosomes. This is only used during visualization.</p>
<p>The structure of the file should look something like this</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>chr1    0       2300000 p36.33  gneg
chr1    2300000 5400000 p36.32  gpos25
chr1    5400000 7200000 p36.31  gneg
chr1    7200000 9200000 p36.23  gpos25
chr1    9200000 12700000        p36.22  gneg
</pre></div>
</div>
</div>
<div class="section" id="masking-file">
<span id="reference-files-masking"></span><h2>Masking File<a class="headerlink" href="#masking-file" title="Permalink to this headline">Â¶</a></h2>
<p>The masking file is a tab delimited file which contains regions that we should ignore calls in.
This can be used to filter out regions with known false positives, bad mapping, centromeres, telomeres etc.
An example of the expected format is shown below. The file should have four columns: chr, start, end and name.</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>#chr    start   end     name
chr1    0       2300000 centromere
chr1    9200000 12700000        telomere
</pre></div>
</div>
<p>The pre-built masking files in the downloads table above are telomere regions, centromere regions (based on the cytoband file),
and nspan regions (computed with tools/find_repeats.py).</p>
<p>Masking is not required (can provide a header-only file), but is recommended as it will improve performance and specificity.</p>
</div>
<div class="section" id="annotations">
<span id="reference-files-annotations"></span><h2>Annotations<a class="headerlink" href="#annotations" title="Permalink to this headline">Â¶</a></h2>
<p>This is a custom file format. It is a <a class="reference internal" href="glossary.html#term-json"><span class="xref std std-term">JSON</span></a> file which contains the gene, transcript, exon,
translation and protein domain positional information</p>
<p>Pre-built annotation files can be downloaded above. The âbest transcriptâ flag is based on an in-house model, as are the
aliases (custom one-to-one hugo gene name mapping)</p>
<div class="admonition warning">
<p class="first admonition-title">Warning</p>
<p class="last">the <a class="reference internal" href="auto/mavis.annotate.file_io.html#mavis.annotate.file_io.load_reference_genes" title="mavis.annotate.file_io.load_reference_genes"><code class="xref py py-func docutils literal"><span class="pre">load_reference_genes()</span></code></a> will
only load valid translations. If the cds sequence in the annotation is not
a multiple of <a class="reference internal" href="auto/mavis.constants.html#mavis.constants.CODON_SIZE" title="mavis.constants.CODON_SIZE"><code class="xref py py-attr docutils literal"><span class="pre">CODON_SIZE</span></code></a> or if a
reference genome (sequences) is given and the cds start and end are not
M and * amino acids as expected the translation is not loaded</p>
</div>
<p>Example of the <a class="reference internal" href="glossary.html#term-json"><span class="xref std std-term">JSON</span></a> file structure can be seen below</p>
<div class="highlight-javascript"><div class="highlight"><pre><span></span><span class="p">[</span>
    <span class="p">{</span>
        <span class="s2">&quot;name&quot;</span><span class="o">:</span> <span class="nx">string</span><span class="p">,</span>
        <span class="s2">&quot;start&quot;</span><span class="o">:</span> <span class="kr">int</span><span class="p">,</span>
        <span class="s2">&quot;end&quot;</span><span class="o">:</span> <span class="kr">int</span>
        <span class="s2">&quot;aliases&quot;</span><span class="o">:</span> <span class="p">[</span><span class="nx">string</span><span class="p">,</span> <span class="nx">string</span><span class="p">,</span> <span class="p">...],</span>
        <span class="s2">&quot;transcripts&quot;</span><span class="o">:</span> <span class="p">[</span>
            <span class="p">{</span>
                <span class="s2">&quot;name&quot;</span><span class="o">:</span> <span class="nx">string</span><span class="p">,</span>
                <span class="s2">&quot;start&quot;</span><span class="o">:</span> <span class="kr">int</span><span class="p">,</span>
                <span class="s2">&quot;end&quot;</span><span class="o">:</span> <span class="kr">int</span><span class="p">,</span>
                <span class="s2">&quot;exons&quot;</span><span class="o">:</span> <span class="p">[</span>
                    <span class="p">{</span><span class="s2">&quot;start&quot;</span><span class="o">:</span> <span class="kr">int</span><span class="p">,</span> <span class="s2">&quot;end&quot;</span><span class="o">:</span> <span class="kr">int</span><span class="p">,</span> <span class="s2">&quot;name&quot;</span><span class="o">:</span> <span class="nx">string</span><span class="p">},</span>
                    <span class="p">...</span>
                <span class="p">],</span>
                <span class="s2">&quot;cdna_coding_start&quot;</span><span class="o">:</span> <span class="kr">int</span><span class="p">,</span>
                <span class="s2">&quot;cdna_coding_end&quot;</span><span class="o">:</span> <span class="kr">int</span><span class="p">,</span>
                <span class="s2">&quot;domains&quot;</span><span class="o">:</span> <span class="p">[</span>
                    <span class="p">{</span>
                        <span class="s2">&quot;name&quot;</span><span class="o">:</span> <span class="nx">string</span><span class="p">,</span>
                        <span class="s2">&quot;regions&quot;</span><span class="o">:</span> <span class="p">[</span>
                            <span class="p">{</span><span class="s2">&quot;start&quot;</span> <span class="nx">aa_start</span><span class="p">,</span> <span class="s2">&quot;end&quot;</span><span class="o">:</span> <span class="nx">aa_end</span><span class="p">}</span>
                        <span class="p">],</span>
                        <span class="s2">&quot;desc&quot;</span><span class="o">:</span> <span class="nx">string</span>
                    <span class="p">},</span>
                    <span class="p">...</span>
                <span class="p">]</span>
            <span class="p">},</span>
            <span class="p">...</span>
        <span class="p">]</span>
    <span class="p">},</span>
    <span class="p">...</span>
<span class="p">}</span>
</pre></div>
</div>
<p>The provided files were generated with <a class="reference internal" href="citations.html#yates-2016"><span class="std std-ref">Ensembl</span></a>, however it can be generated from any database with the
necessary information so long as the above <a class="reference internal" href="glossary.html#term-json"><span class="xref std std-term">JSON</span></a> structure is respected.</p>
<div class="section" id="generating-the-annotations-from-ensembl">
<span id="generate-reference-annotations"></span><h3>Generating the Annotations from <a class="reference internal" href="citations.html#yates-2016"><span class="std std-ref">Ensembl</span></a><a class="headerlink" href="#generating-the-annotations-from-ensembl" title="Permalink to this headline">Â¶</a></h3>
<p>There is a helper script included with mavis to facilitate generating the custom annotations
file from an instance of the <a class="reference internal" href="citations.html#yates-2016"><span class="std std-ref">Ensembl</span></a> database. This uses the <a class="reference internal" href="citations.html#yates-2016"><span class="std std-ref">Ensembl</span></a> perl api to connect and
pull information from the database. This has been tested with both Ensembl69 and Ensembl79.</p>
<p>Instructions for downloading and installing the perl api can be found on the <a class="reference external" href="http://www.ensembl.org/info/docs/api/api_installation.html">ensembl site</a></p>
<ol class="arabic simple">
<li><strong>Make sure the ensembl perl api modules are added to the PERL5LIB environment variable</strong></li>
</ol>
<div class="highlight-bash"><div class="highlight"><pre><span></span><span class="nv">PERL5LIB</span><span class="o">=</span><span class="si">${</span><span class="nv">PERL5LIB</span><span class="si">}</span>:<span class="nv">$HOME</span>/ensembl_79/bioperl-live
<span class="nv">PERL5LIB</span><span class="o">=</span><span class="si">${</span><span class="nv">PERL5LIB</span><span class="si">}</span>:<span class="nv">$HOME</span>/ensembl_79/ensembl/modules
<span class="nv">PERL5LIB</span><span class="o">=</span><span class="si">${</span><span class="nv">PERL5LIB</span><span class="si">}</span>:<span class="nv">$HOME</span>/ensembl_79/ensembl-compara/modules
<span class="nv">PERL5LIB</span><span class="o">=</span><span class="si">${</span><span class="nv">PERL5LIB</span><span class="si">}</span>:<span class="nv">$HOME</span>/ensembl_79/ensembl-variation/modules
<span class="nv">PERL5LIB</span><span class="o">=</span><span class="si">${</span><span class="nv">PERL5LIB</span><span class="si">}</span>:<span class="nv">$HOME</span>/ensembl_79/ensembl-funcgen/modules
<span class="nb">export</span> PERL5LIB
</pre></div>
</div>
<ol class="arabic simple" start="2">
<li><strong>Configure the environment variables to set defaults for the perl script</strong></li>
</ol>
<div class="highlight-bash"><div class="highlight"><pre><span></span><span class="c1"># required data files</span>
<span class="nb">export</span> <span class="nv">HUGO_ENSEMBL_MAPPING</span><span class="o">=</span>/path/to/mapping/file
<span class="nb">export</span> <span class="nv">BEST_TRANSCRIPTS</span><span class="o">=</span>/path/to/transcripts/file

<span class="c1"># connection information for the ensembl local (or external) server</span>
<span class="nb">export</span> <span class="nv">ENSEMBL_HOST</span><span class="o">=</span>HOSTNAME
<span class="nb">export</span> <span class="nv">ENSEMBL_PASS</span><span class="o">=</span>PASSWORD
<span class="nb">export</span> <span class="nv">ENSEMBL_USER</span><span class="o">=</span>USERNAME
<span class="nb">export</span> <span class="nv">ENSEMBL_PORT</span><span class="o">=</span>PORT_NUMBER
</pre></div>
</div>
<ol class="arabic simple" start="3">
<li><strong>Run the perl script</strong></li>
</ol>
<p>you can view the help menu by running</p>
<div class="highlight-bash"><div class="highlight"><pre><span></span>perl generate_ensembl_json.pl
</pre></div>
</div>
<p>you can override the default input file parameters (configured in the above step) by providing arguments
to the script itself</p>
<div class="highlight-bash"><div class="highlight"><pre><span></span>perl generate_ensembl_json.pl --best_transcript_file /path/to/best/transcripts/file --output /path/to/output/json/file.json
</pre></div>
</div>
<p>or if you have configured the environment variables as given in step 2, then simply provide the output path</p>
<div class="highlight-bash"><div class="highlight"><pre><span></span>perl generate_ensembl_json.pl --output /path/to/output/json/file.json
</pre></div>
</div>
</div>
</div>
<div class="section" id="dgv-database-of-genomic-variants-annotations">
<span id="reference-files-dgv-annotations"></span><h2><a class="reference internal" href="citations.html#macdonald-2014"><span class="std std-ref">DGV (Database of Genomic Variants)</span></a> Annotations<a class="headerlink" href="#dgv-database-of-genomic-variants-annotations" title="Permalink to this headline">Â¶</a></h2>
<p>The DGV annotations file contains regions corresponding to what is found in the database of genomic variants. This is
used to annotate events that are found in healthy control samples and therefore may not be of interest
if looking for somatic events.</p>
<p>The above (downloads table) files were generated from from <a class="reference external" href="http://dgv.tcag.ca/dgv/app/download">DGV</a>
and reformatted to have 4 columns after download. We used awk to convert the raw file</p>
<div class="highlight-bash"><div class="highlight"><pre><span></span>awk &#39;{print $2&quot;\t&quot;$3&quot;\t&quot;$4&quot;\t&quot;$1} GRCh37_hg19_variants_2016-05-15.txt &gt; dgv_hg19_variants.tab
</pre></div>
</div>
<p>Note in hg19 the column is called ânameâ and in hg38 the column is called âvariantaccessionâ.
An example is shown below</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>#chr     start   end     name
1       1       2300000 nsv482937
1       10001   22118   dgv1n82
1       10001   127330  nsv7879
</pre></div>
</div>
</div>
<div class="section" id="aligner-reference">
<span id="reference-files-aligner-reference"></span><h2>Aligner Reference<a class="headerlink" href="#aligner-reference" title="Permalink to this headline">Â¶</a></h2>
<p>The aligner reference file is the reference genome file used by the aligner during the validate stage. For example,
if <a class="reference internal" href="glossary.html#term-blat"><span class="xref std std-term">blat</span></a> is the aligner then this will be a <a class="reference internal" href="glossary.html#term-2bit"><span class="xref std std-term">2bit</span></a> file.</p>
</div>
</div>


           </div>
           <div class="articleComments">
            
           </div>
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="configuration.html" class="btn btn-neutral float-right" title="Configuration and Settings" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="mavis_input.html" class="btn btn-neutral" title="MAVIS standard input file format" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2017, creisle.

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  

    <script type="text/javascript">
        var DOCUMENTATION_OPTIONS = {
            URL_ROOT:'./',
            VERSION:'',
            LANGUAGE:'None',
            COLLAPSE_INDEX:false,
            FILE_SUFFIX:'.html',
            HAS_SOURCE:  true,
            SOURCELINK_SUFFIX: '.txt'
        };
    </script>
      <script type="text/javascript" src="_static/jquery.js"></script>
      <script type="text/javascript" src="_static/underscore.js"></script>
      <script type="text/javascript" src="_static/doctools.js"></script>
      <script type="text/javascript" src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>

  

  
  
    <script type="text/javascript" src="_static/js/theme.js"></script>
  

  
  
  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.StickyNav.enable();
      });
  </script>
   

</body>
</html>