

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Theory and Models &mdash; MAVIS  documentation</title>
  

  
  
  
  

  

  
  
    

  

  
  
    <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  

  
    <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  

  
        <link rel="author" title="About these documents"
              href="about.html"/>
        <link rel="index" title="Index"
              href="genindex.html"/>
        <link rel="search" title="Search" href="search.html"/>
    <link rel="top" title="MAVIS  documentation" href="index.html"/>
        <link rel="next" title="Guidelines for Contributors" href="development.html"/>
        <link rel="prev" title="Frequently Asked Questions" href="faqs.html"/> 

  
  <script src="_static/js/modernizr.min.js"></script>

</head>

<body class="wy-body-for-nav" role="document">

   
  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search">
          

          
            <a href="index.html" class="icon icon-home"> MAVIS
          

          
          </a>

          
            
            
              <div class="version">
                1.6.4
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="about.html">About</a></li>
<li class="toctree-l1"><a class="reference internal" href="pipeline.html">Running the Pipeline</a></li>
<li class="toctree-l1"><a class="reference internal" href="mavis_input.html">MAVIS standard input file format</a></li>
<li class="toctree-l1"><a class="reference internal" href="reference.html">Reference Input Files</a></li>
<li class="toctree-l1"><a class="reference internal" href="configuration.html">Configuration and Settings</a></li>
<li class="toctree-l1"><a class="reference internal" href="tools_supported.html">Supported SV Callers</a></li>
<li class="toctree-l1"><a class="reference internal" href="illustrations.html">Illustrations</a></li>
<li class="toctree-l1"><a class="reference internal" href="faqs.html">Frequently Asked Questions</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Theory and Models</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#introduction">Introduction</a></li>
<li class="toctree-l2"><a class="reference internal" href="#evidence">Evidence</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#types-of-flanking-evidence">Types of Flanking evidence</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#deletion">Deletion</a></li>
<li class="toctree-l4"><a class="reference internal" href="#insertion">Insertion</a></li>
<li class="toctree-l4"><a class="reference internal" href="#duplication">Duplication</a></li>
<li class="toctree-l4"><a class="reference internal" href="#inversion">Inversion</a></li>
<li class="toctree-l4"><a class="reference internal" href="#translocation">Translocation</a></li>
<li class="toctree-l4"><a class="reference internal" href="#inverted-translocation">Inverted Translocation</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#compatible-flanking-pairs">Compatible Flanking Pairs</a></li>
<li class="toctree-l3"><a class="reference internal" href="#calculating-the-evidence-window">Calculating the Evidence Window</a></li>
<li class="toctree-l3"><a class="reference internal" href="#calling-breakpoints-by-flanking-evidence">Calling Breakpoints by Flanking Evidence</a></li>
<li class="toctree-l3"><a class="reference internal" href="#determining-flanking-support">Determining Flanking support</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#classifying-events">Classifying Events</a></li>
<li class="toctree-l2"><a class="reference internal" href="#assembling-contigs">Assembling Contigs</a></li>
<li class="toctree-l2"><a class="reference internal" href="#annotating-events">Annotating Events</a></li>
<li class="toctree-l2"><a class="reference internal" href="#predicting-splicing-patterns">Predicting Splicing Patterns</a></li>
<li class="toctree-l2"><a class="reference internal" href="#pairing-similar-events">Pairing Similar Events</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="development.html">Guidelines for Contributors</a></li>
<li class="toctree-l1"><a class="reference internal" href="package.html">Package Documentation</a></li>
<li class="toctree-l1"><a class="reference internal" href="glossary.html">Glossary</a></li>
<li class="toctree-l1"><a class="reference internal" href="citations.html">Literature</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">MAVIS</a>
        
      </nav>


      
      <div class="wy-nav-content">
        <div class="rst-content">
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html">Docs</a> &raquo;</li>
        
      <li>Theory and Models</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="_sources/theory.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="theory-and-models">
<h1>Theory and Models<a class="headerlink" href="#theory-and-models" title="Permalink to this headline">¶</a></h1>
<div class="section" id="introduction">
<h2>Introduction<a class="headerlink" href="#introduction" title="Permalink to this headline">¶</a></h2>
<p>In MAVIS structural variants (SVs) are defined by a pair of breakpoints</p>
<p>And a breakpoint is defined by</p>
<ol class="arabic simple">
<li>chromosome</li>
<li>base-pair range (start, end). This has a length of 1 for exact calls and more for uncertain/non-specific calls</li>
<li><span class="xref std std-term">orientation</span>. This is Left or Right with respect to the positive/forward strand. This defines which portion of the genome is ‘retained’</li>
<li><span class="xref std std-term">strand</span>. (only applicable to stranded transcriptome libraries)</li>
</ol>
<p>So then a breakpoint pair is any two intervals on the reference genome which are adjacent in the mutant genome</p>
<div class="line-block">
<div class="line"><br /></div>
</div>
<hr class="docutils" />
<div class="line-block">
<div class="line"><br /></div>
</div>
</div>
<div class="section" id="evidence">
<h2>Evidence<a class="headerlink" href="#evidence" title="Permalink to this headline">¶</a></h2>
<p>There are many ways that single reads or paired-end reads can act as support for an SV call.</p>
<div class="figure" id="id1">
<a class="reference internal image-reference" href="_images/read_evidence.svg"><img alt="_images/read_evidence.svg" src="_images/read_evidence.svg" width="100%" /></a>
<p class="caption"><span class="caption-number">Figure 5. </span><span class="caption-text">In the figure above
the red rectangle represents a deletion structural variant. The arrows are types of single or paired-end reads supporting
the event: <a class="reference internal" href="glossary.html#term-flanking-read-pair"><span class="xref std std-term">flanking read pair</span></a>s (F), <a class="reference internal" href="glossary.html#term-split-read"><span class="xref std std-term">split read</span></a>s (S), <a class="reference internal" href="glossary.html#term-half-mapped-read"><span class="xref std std-term">half-mapped read</span></a>s (H),
and <a class="reference internal" href="glossary.html#term-spanning-read"><span class="xref std std-term">spanning read</span></a>s (N).</span></p>
</div>
<div class="section" id="types-of-flanking-evidence">
<span id="theory-types-of-flanking-evidence"></span><h3>Types of Flanking evidence<a class="headerlink" href="#types-of-flanking-evidence" title="Permalink to this headline">¶</a></h3>
<p>One of the most confusing parts about working with contig and paired-end reads is relating them to the
breakpoint so that you can determine which types will support an event. The flanking read types we outline here
are similarly described by <a class="reference external" href="http://software.broadinstitute.org/software/igv/interpreting_insert_size">IGV</a>.
We have used similar coloring for the read pairs in the following diagrams to
facilitate ease of use for those already familiar with viewing bam files in IGV.</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">The major assumptions here are that the ‘normal’ read-pair is a read pair which has one read on the positive/forward
strand and its partner on the negative/reverse strand. It is assumed that partners share a read name, as is the case for Illumina reads.</p>
</div>
<div class="section" id="deletion">
<h4>Deletion<a class="headerlink" href="#deletion" title="Permalink to this headline">¶</a></h4>
<p>For a deletion, we expect the flanking reads to be in the normal orientation but that the
fragment size should be abnormal (for large deletions).</p>
<div class="figure" id="id2">
<a class="reference internal image-reference" href="_images/read_pairs_deletion.svg"><img alt="_images/read_pairs_deletion.svg" src="_images/read_pairs_deletion.svg" width="100%" /></a>
<p class="caption"><span class="caption-number">Figure 6. </span><span class="caption-text">Flanking read pair evidence for a deletion event. the read pairs will have a larger than expected fragment size when mapped to the
reference genome because in the mutant genome they are closer together, owing to the deletion event. (B1) The first breakpoint
which has a left orientation. (B2) The second breakpoint which has a right orientation. Both breakpoints would be on the positive
strand (assuming that the input is stranded) which means that the first read in the pair would be on the positive strand and the
second read in the pair would be on the negative/reverse strand.</span></p>
</div>
</div>
<div class="section" id="insertion">
<h4>Insertion<a class="headerlink" href="#insertion" title="Permalink to this headline">¶</a></h4>
<div class="figure" id="id3">
<a class="reference internal image-reference" href="_images/read_pairs_insertion.svg"><img alt="_images/read_pairs_insertion.svg" src="_images/read_pairs_insertion.svg" width="100%" /></a>
<p class="caption"><span class="caption-number">Figure 7. </span><span class="caption-text">Flanking read pair evidence for an insertion event. The read pairs will have a smaller than expected fragment size when mapped to the
reference genome because in the mutant genome they are father apart, owing to the insertion event. (B1) The first breakpoint
which has a left orientation. (B2) The second breakpoint which has a right orientation. Both breakpoints would be on the positive
strand (assuming that the input is stranded) which means that the first read in the pair would be on the positive strand and the
second read in the pair would be on the negative/reverse strand.</span></p>
</div>
</div>
<div class="section" id="duplication">
<h4>Duplication<a class="headerlink" href="#duplication" title="Permalink to this headline">¶</a></h4>
<div class="figure" id="id4">
<a class="reference internal image-reference" href="_images/read_pairs_duplication.svg"><img alt="_images/read_pairs_duplication.svg" src="_images/read_pairs_duplication.svg" width="100%" /></a>
<p class="caption"><span class="caption-number">Figure 8. </span><span class="caption-text">Flanking read pair evidence for a tandem duplication event. The read pairs will have an abnormal orientation but still the
same strands as the normal read pair. (B1) The first breakpoint will be on the positive strand and have a right orientation.
(B2) The second breakpoint will be on the positive strand and have a left orientation.</span></p>
</div>
</div>
<div class="section" id="inversion">
<h4>Inversion<a class="headerlink" href="#inversion" title="Permalink to this headline">¶</a></h4>
<div class="figure" id="id5">
<a class="reference internal image-reference" href="_images/read_pairs_inversion_LL.svg"><img alt="_images/read_pairs_inversion_LL.svg" src="_images/read_pairs_inversion_LL.svg" width="100%" /></a>
<p class="caption"><span class="caption-number">Figure 9. </span><span class="caption-text">Flanking read pair evidence for an inversion. Both breakpoints have left orientation.</span></p>
</div>
<div class="figure" id="id6">
<a class="reference internal image-reference" href="_images/read_pairs_inversion_RR.svg"><img alt="_images/read_pairs_inversion_RR.svg" src="_images/read_pairs_inversion_RR.svg" width="100%" /></a>
<p class="caption"><span class="caption-number">Figure 10. </span><span class="caption-text">Flanking read pair evidence for an inversion. Both breakpoints have right orientation.</span></p>
</div>
</div>
<div class="section" id="translocation">
<h4>Translocation<a class="headerlink" href="#translocation" title="Permalink to this headline">¶</a></h4>
<div class="figure" id="id7">
<a class="reference internal image-reference" href="_images/read_pairs_translocation_LR.svg"><img alt="_images/read_pairs_translocation_LR.svg" src="_images/read_pairs_translocation_LR.svg" width="100%" /></a>
<p class="caption"><span class="caption-number">Figure 11. </span><span class="caption-text">Flanking read pair evidence for a translocation. (B1) the first breakpoint with a left orientation. (B2) the second breakpoint with a right orientation.</span></p>
</div>
<div class="figure" id="id8">
<a class="reference internal image-reference" href="_images/read_pairs_translocation_RL.svg"><img alt="_images/read_pairs_translocation_RL.svg" src="_images/read_pairs_translocation_RL.svg" width="100%" /></a>
<p class="caption"><span class="caption-number">Figure 12. </span><span class="caption-text">Flanking read pair evidence for a translocation. (B1) the first breakpoint with a right orientation. (B2) the second breakpoint with a left orientation.</span></p>
</div>
</div>
<div class="section" id="inverted-translocation">
<h4>Inverted Translocation<a class="headerlink" href="#inverted-translocation" title="Permalink to this headline">¶</a></h4>
<div class="figure" id="id9">
<a class="reference internal image-reference" href="_images/read_pairs_translocated_inversion_LL.svg"><img alt="_images/read_pairs_translocated_inversion_LL.svg" src="_images/read_pairs_translocated_inversion_LL.svg" width="100%" /></a>
<p class="caption"><span class="caption-number">Figure 13. </span><span class="caption-text">Flanking read pair evidence for an inverted translocation. Both breakpoints have left orientation.</span></p>
</div>
<div class="figure" id="id10">
<a class="reference internal image-reference" href="_images/read_pairs_translocated_inversion_RR.svg"><img alt="_images/read_pairs_translocated_inversion_RR.svg" src="_images/read_pairs_translocated_inversion_RR.svg" width="100%" /></a>
<p class="caption"><span class="caption-number">Figure 14. </span><span class="caption-text">Flanking read pair evidence for an inverted translocation. Both breakpoints have right orientation.</span></p>
</div>
<div class="line-block">
<div class="line"><br /></div>
</div>
<hr class="docutils" />
<div class="line-block">
<div class="line"><br /></div>
</div>
</div>
</div>
<div class="section" id="compatible-flanking-pairs">
<span id="theory-compatible-flanking-pairs"></span><h3>Compatible Flanking Pairs<a class="headerlink" href="#compatible-flanking-pairs" title="Permalink to this headline">¶</a></h3>
<p>For insertion and duplication events compatible flanking pairs are collected. This means that flanking pairs that
support a duplication may be used as compatible flanking evidence for an insertion (in the same region) and similarly
flanking pairs which support an insertion may be compatible flanking evidence for a duplication</p>
<div class="figure" id="id11">
<a class="reference internal image-reference" href="_images/compatible_flanking_pairs.svg"><img alt="_images/compatible_flanking_pairs.svg" src="_images/compatible_flanking_pairs.svg" width="100%" /></a>
<p class="caption"><span class="caption-number">Figure 15. </span><span class="caption-text">The event depicted above may be called as either a duplication or an insertion (depending on the input call). If
the even were called as a duplication the reads in green would be the flanking supoprt and the reads in blue
would be given as compatible flanking support. If the event were called as an insertion the reverse would apply.</span></p>
</div>
<div class="line-block">
<div class="line"><br /></div>
</div>
<hr class="docutils" />
<div class="line-block">
<div class="line"><br /></div>
</div>
</div>
<div class="section" id="calculating-the-evidence-window">
<span id="theory-calculating-the-evidence-window"></span><h3>Calculating the Evidence Window<a class="headerlink" href="#calculating-the-evidence-window" title="Permalink to this headline">¶</a></h3>
<div class="figure" id="id12">
<a class="reference internal image-reference" href="_images/read_pair_definitions.svg"><img alt="_images/read_pair_definitions.svg" src="_images/read_pair_definitions.svg" width="100%" /></a>
<p class="caption"><span class="caption-number">Figure 16. </span><span class="caption-text">Basic Terms used in describing read pairs are shown above: fragment size: the distance between the pair;
read length: the length of the read; fragment size: the combined length of both reads and the fragment size</span></p>
</div>
<p>We make some base assumptions with regards to paired-end read data:</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">the distribution of fragment sizes approximately follows a normal distribution</p>
</div>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">the most common fragment size is the unmutated ‘normal’ fragment</p>
</div>
<p>With the above assumptions we take the median fragment size to be the expected normal.</p>
<p>Given that we expect mutations and therefore abnormal fragment sizes, we use a modified method to calculate the
<strong>median standard deviation</strong> (see code below). We calculate the squared distance away from the
median for each fragment and then take a fraction of this to be ‘normal’ variation. So the most abnormal portion is
ignored, assuming it is supposed to be abnormal. This results in a calculation as follows.</p>
<div class="code highlight-default"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">statistics</span> <span class="k">import</span> <span class="n">median</span>
<span class="kn">import</span> <span class="nn">math</span>

<span class="n">fragments</span> <span class="o">=</span> <span class="p">[</span><span class="nb">abs</span><span class="p">(</span><span class="n">read</span><span class="o">.</span><span class="n">template_length</span><span class="p">)</span> <span class="k">for</span> <span class="n">read</span> <span class="ow">in</span> <span class="n">reads</span><span class="p">]</span>  <span class="c1"># the fragment sizes of the reads</span>
<span class="n">f</span> <span class="o">=</span> <span class="mf">0.95</span> <span class="c1"># fraction</span>
<span class="n">m</span> <span class="o">=</span> <span class="n">median</span><span class="p">(</span><span class="n">fragments</span><span class="p">)</span> <span class="c1"># get the median fragment size value</span>
<span class="n">X</span> <span class="o">=</span> <span class="p">[</span><span class="n">math</span><span class="o">.</span><span class="n">pow</span><span class="p">(</span><span class="n">i</span> <span class="o">-</span> <span class="n">m</span><span class="p">,</span> <span class="mi">2</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">fragments</span><span class="p">]</span>  <span class="c1"># take the square error for each point</span>
<span class="n">end</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="nb">round</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">X</span><span class="p">)</span> <span class="o">*</span> <span class="n">f</span><span class="p">))</span>
<span class="n">X</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">X</span><span class="p">)[</span><span class="mi">0</span><span class="p">:</span><span class="n">end</span><span class="p">]</span>
<span class="n">stdev</span> <span class="o">=</span> <span class="n">math</span><span class="o">.</span><span class="n">sqrt</span><span class="p">(</span><span class="nb">sum</span><span class="p">(</span><span class="n">X</span><span class="p">)</span> <span class="o">/</span> <span class="nb">len</span><span class="p">(</span><span class="n">X</span><span class="p">))</span>
</pre></div>
</div>
<p>This gives us an idea of when to judge an fragment size as abnormal and where we expect our normal read
pairs fragment sizes to fall.</p>
<div class="figure" id="id13">
<a class="reference internal image-reference" href="_images/fragment_sizes_histogram.svg"><img alt="_images/fragment_sizes_histogram.svg" src="_images/fragment_sizes_histogram.svg" width="100%" /></a>
<p class="caption"><span class="caption-number">Figure 17. </span><span class="caption-text">Distribution of fragment sizes (absolute values) of proper read pairs. The black curve representings the fit for
a normal distribution using the standard deviation calculated with all data points. The blue curve is the expected
distribution using a 0.95 fraction of the data. The thick vertical black line is the median and the thin black
lines are standard deviations away from the median.</span></p>
</div>
<p>As we can see from the diagram above, removing the outliers reproduces the observed distribution better than using
all data points</p>
<p>We use this in two ways</p>
<ol class="arabic simple">
<li>to find flanking evidence supporting deletions and insertions</li>
<li>to estimate the window size for where we will need to read from the bam when looking for evidence for a given event</li>
</ol>
<p>The <code class="xref py py-func docutils literal"><span class="pre">_generate_window()</span></code> function uses the above concepts.
The user will define the <code class="docutils literal"><span class="pre">median_fragment_size</span></code> the <code class="docutils literal"><span class="pre">stdev_fragment_size</span></code> , and the <code class="docutils literal"><span class="pre">stdev_count_abnormal</span></code>
parameters defined in the <code class="xref py py-class docutils literal"><span class="pre">VALIDATION_DEFAULTS</span></code> class.</p>
<p>If the library has a transcriptome protocol this becomes a bit more complicated and we must take into account the
possible annotations when calculating the evidence window. see
<code class="xref py py-func docutils literal"><span class="pre">_generate_window()</span></code> for more</p>
<div class="line-block">
<div class="line"><br /></div>
</div>
<hr class="docutils" />
<div class="line-block">
<div class="line"><br /></div>
</div>
</div>
<div class="section" id="calling-breakpoints-by-flanking-evidence">
<span id="theory-calling-breakpoints-by-flanking-evidence"></span><h3>Calling Breakpoints by Flanking Evidence<a class="headerlink" href="#calling-breakpoints-by-flanking-evidence" title="Permalink to this headline">¶</a></h3>
<p>Breakpoints are called by contig, split-read, or flanking pairs evidence. Contigs and split reads are used to call exact
breakpoints, where breakpoints called by flanking reads are generally assigned a probabilistic range.</p>
<p>The metrics used here are similar to those used in calculating the evidence window. We use the
<span class="xref std std-term">max_expected_fragment_size</span> as the outer limit of how large the range can be. This is further refined taking
into account the range spanned by the <a class="reference internal" href="glossary.html#term-flanking-read-pair"><span class="xref std std-term">flanking read pair</span></a> evidence and the position of the opposing breakpoint.</p>
<div class="figure" id="id14">
<a class="reference internal image-reference" href="_images/call_breakpoint_by_flanking_reads.svg"><img alt="_images/call_breakpoint_by_flanking_reads.svg" src="_images/call_breakpoint_by_flanking_reads.svg" width="100%" /></a>
<p class="caption"><span class="caption-number">Figure 18. </span><span class="caption-text">Calculation of the left-oriented breakpoint by flanking reads. Reads mapped to the breakpoint are shown in grey.
The read on the right (black outline, no fill) demonstrates the read length used to narrow the right side bound of
the estimated breakpoint interval.</span></p>
</div>
<div class="line-block">
<div class="line"><br /></div>
</div>
<hr class="docutils" />
<div class="line-block">
<div class="line"><br /></div>
</div>
</div>
<div class="section" id="determining-flanking-support">
<span id="theory-determining-flanking-support"></span><h3>Determining Flanking support<a class="headerlink" href="#determining-flanking-support" title="Permalink to this headline">¶</a></h3>
<div class="figure" id="id15">
<a class="reference internal image-reference" href="_images/flanking_pairs_fragment_sizes_deletion.svg"><img alt="_images/flanking_pairs_fragment_sizes_deletion.svg" src="_images/flanking_pairs_fragment_sizes_deletion.svg" width="100%" /></a>
<p class="caption"><span class="caption-number">Figure 19. </span><span class="caption-text">After a breakpoint has been called we can narrow the interval of expected fragment sizes using the
size of the event. (Left) The colored portion of the graph represents the range in fragment sizes we
expect for a normal/unmutated genome. (Right) For a deletion event we expect the fragment size to be bigger,
outside the normal range. Reads that would flank the breakpoint should follow a similar distribution
to the normal genome but the median will be shifted by the size of the event. The shaded portion of the
graph represents the range in fragment sizes we expect for flanking pairs supporting the deletion event.</span></p>
</div>
<div class="line-block">
<div class="line"><br /></div>
</div>
<hr class="docutils" />
<div class="line-block">
<div class="line"><br /></div>
</div>
</div>
</div>
<div class="section" id="classifying-events">
<span id="theory-classifying-events"></span><h2>Classifying Events<a class="headerlink" href="#classifying-events" title="Permalink to this headline">¶</a></h2>
<p>The following decision tree is used in classifying events based on their breakpoints. Only valid combinations have
been shown. see <a class="reference internal" href="auto/mavis.breakpoint.html#mavis.breakpoint.BreakpointPair.classify" title="mavis.breakpoint.BreakpointPair.classify"><code class="xref py py-func docutils literal"><span class="pre">classify()</span></code></a></p>
<div class="figure" id="id16">
<a class="reference internal image-reference" href="_images/classification_tree.svg"><img alt="_images/classification_tree.svg" src="_images/classification_tree.svg" width="100%" /></a>
<p class="caption"><span class="caption-number">Figure 20. </span><span class="caption-text">Classification Decision Tree. The above  diagram details the decsion logic for classifying events based on the
orientation, strand and chromosomes or their respective breakpoints</span></p>
</div>
<div class="line-block">
<div class="line"><br /></div>
</div>
<hr class="docutils" />
<div class="line-block">
<div class="line"><br /></div>
</div>
</div>
<div class="section" id="assembling-contigs">
<span id="theory-assembling-contigs"></span><h2>Assembling Contigs<a class="headerlink" href="#assembling-contigs" title="Permalink to this headline">¶</a></h2>
<p>During validation, for each breakpoint pair, we attempt to assemble a contig to represent the sequence across
the breakpoints. This is assembled from the supporting reads (<a class="reference internal" href="glossary.html#term-split-read"><span class="xref std std-term">split read</span></a>, <a class="reference internal" href="glossary.html#term-half-mapped-read"><span class="xref std std-term">half-mapped read</span></a>,
<a class="reference internal" href="glossary.html#term-flanking-read-pair"><span class="xref std std-term">flanking read pair</span></a>, and <a class="reference internal" href="glossary.html#term-spanning-read"><span class="xref std std-term">spanning read</span></a>) which have already been collected for the given
event. The sequence from each read and its reverse complement are assembled into contigs using a <span class="xref std std-term">DeBruijn graph</span>.
For strand specific events, we then attempt to resolve the sequence strand of the contig.</p>
<div class="line-block">
<div class="line"><br /></div>
</div>
<hr class="docutils" />
<div class="line-block">
<div class="line"><br /></div>
</div>
</div>
<div class="section" id="annotating-events">
<span id="theory-annotating-events"></span><h2>Annotating Events<a class="headerlink" href="#annotating-events" title="Permalink to this headline">¶</a></h2>
<p>We make the following assumptions when determining the annotations for each event</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">If both breakpoints are in the same gene, they must also be in the same transcript</p>
</div>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">If the breakpoint intervals overlap we do not annotate encompassed genes</p>
</div>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">Encompassed and ‘nearest’ genes are reported without respect to strand</p>
</div>
<p>There are specific questions we want annotation to answer. We collect gene level
annotations which describes things like what gene is near the breakpoint (useful
in the case of a potential promoter swap); what genes (besides the one selected)
also overlap the breakpoint; what genes are encompassed between the breakpoints
(for example in a deletion event the genes that would be deleted).</p>
<div class="figure" id="id17">
<a class="reference internal image-reference" href="_images/annotations_summary.svg"><img alt="_images/annotations_summary.svg" src="_images/annotations_summary.svg" width="100%" /></a>
<p class="caption"><span class="caption-number">Figure 21. </span><span class="caption-text">Gene level annotations at each breakpoint. Note: genes which fall
between a breakpoint pair, encompassed genes, will not be present for
interchromosomal events (translocations)</span></p>
</div>
<p>Next there are the fusion-product level annotations. If the event result in a fusion
transcript, the sequence of the fusion transcript is computed. This is translated
to a putative amino acid sequence from which protein metrics such as the possible
ORFs and domain sequences can be computed.</p>
<div class="line-block">
<div class="line"><br /></div>
</div>
<hr class="docutils" />
<div class="line-block">
<div class="line"><br /></div>
</div>
</div>
<div class="section" id="predicting-splicing-patterns">
<span id="theory-predicting-splicing-patterns"></span><h2>Predicting Splicing Patterns<a class="headerlink" href="#predicting-splicing-patterns" title="Permalink to this headline">¶</a></h2>
<p>After the events have been called and an annotation has been attached, we often want to predict information about the
putative fusion protein, which may be a product. In some cases, when a fusion transcript disrupts a splice-site, it is
not clear what the processed fusion transcript may be. MAVIS will calculate all possibilities according to the
following model.</p>
<div class="figure" id="id18">
<a class="reference internal image-reference" href="_images/splicing_pattern_default.svg"><img alt="_images/splicing_pattern_default.svg" src="_images/splicing_pattern_default.svg" width="100%" /></a>
<p class="caption"><span class="caption-number">Figure 22. </span><span class="caption-text">The default splicing pattern is a list of pairs of donor and acceptor splice sites</span></p>
</div>
<p>For a given list of non-abrogated splice sites (listed 5’ to 3’ on the strand of the transcript) donor splice sites
are paired with all following as seen below</p>
<div class="figure" id="id19">
<a class="reference internal image-reference" href="_images/splicing_pattern_multiple_donors.svg"><img alt="_images/splicing_pattern_multiple_donors.svg" src="_images/splicing_pattern_multiple_donors.svg" width="100%" /></a>
<p class="caption"><span class="caption-number">Figure 23. </span><span class="caption-text">Multiple abrogated acceptors sites. As one can see above this situation will result in 3 different splicing
patterns depending on which donor is paired with the 2nd acceptor site</span></p>
</div>
<div class="figure" id="id20">
<a class="reference internal image-reference" href="_images/splicing_pattern_multiple_acceptors.svg"><img alt="_images/splicing_pattern_multiple_acceptors.svg" src="_images/splicing_pattern_multiple_acceptors.svg" width="100%" /></a>
<p class="caption"><span class="caption-number">Figure 24. </span><span class="caption-text">Multiple abrogated donor sites. As one can see above this situation will result in 3 different splicing
patterns depending on which acceptor is paired with the 2nd donor site</span></p>
</div>
<p>More complex examples are drawn below. There are five classifications (<code class="xref py py-class docutils literal"><span class="pre">SPLICE_TYPE</span></code>) for the different splicing patterns:</p>
<ol class="arabic simple">
<li>Retained intron (<code class="xref py py-class docutils literal"><span class="pre">RETAIN</span></code>)</li>
<li>Skipped exon (<code class="xref py py-attr docutils literal"><span class="pre">SKIP</span></code>)</li>
<li>Multiple retained introns (<code class="xref py py-attr docutils literal"><span class="pre">MULTI_RETAIN</span></code>)</li>
<li>Multiple skipped exons (<code class="xref py py-attr docutils literal"><span class="pre">MULTI_SKIP</span></code>)</li>
<li>Some combination of retained introns and skipped exons (<code class="xref py py-attr docutils literal"><span class="pre">COMPLEX</span></code>)</li>
</ol>
<div class="figure" id="id21">
<a class="reference internal image-reference" href="_images/splicing_model.svg"><img alt="_images/splicing_model.svg" src="_images/splicing_model.svg" width="100%" /></a>
<p class="caption"><span class="caption-number">Figure 25. </span><span class="caption-text">Splicing scenarios</span></p>
</div>
<div class="line-block">
<div class="line"><br /></div>
</div>
<hr class="docutils" />
<div class="line-block">
<div class="line"><br /></div>
</div>
</div>
<div class="section" id="pairing-similar-events">
<span id="theory-pairing-similar-events"></span><h2>Pairing Similar Events<a class="headerlink" href="#pairing-similar-events" title="Permalink to this headline">¶</a></h2>
<p>After breakpoints have been called and annotated we often need to see if the same event was found in different samples. To do this we will need to compare events between genome and transcriptome libraries. For this, the following model is proposed. To compare events between different protocol (genome vs transcriptome) we use the annotation overlying the genome breakpoint and the splicing model we defined above to predict where we would expect to find the transcriptomic breakpoints. This gives rise to the following basic cases.</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">In all cases the predicted breakpoint is either the same as the genomic breakpoint, or it is the same as the nearest retained donor/acceptor to the breakpoint.</p>
</div>
<div class="figure" id="id22">
<a class="reference internal image-reference" href="_images/breakpoint_prediction_exonic.svg"><img alt="_images/breakpoint_prediction_exonic.svg" src="_images/breakpoint_prediction_exonic.svg" width="100%" /></a>
<p class="caption"><span class="caption-number">Figure 26. </span><span class="caption-text">(A-D) The breakpoint lands in an exon and the five prime portion of the transcript is retained. (A) The original
splicing pattern showing the placement of the genomic breakpoint and the retained five prime portion. (B) The first
splice site following the breakpoint is a donor and the second donor is used. (C) The first splice site following the
breakpoint is a donor and the first donor is used. (D) The first splice site following the breakpoint is an acceptor.
(E-H) The breakpoint lands in an exon and the three prime portion of the transcript is retained. (E) The original
splicing pattern showing the placement of the genomic breakpoint and the retained three prime portion. (F) The first
splice site prior to the breakpoint is an acceptor and the first acceptor is used. (G) The first splice site prior to the
breakpoint is an acceptor and the second acceptor is used. (H) The first splice site prior to the breakpoint is a donor</span></p>
</div>
</div>
</div>


           </div>
           <div class="articleComments">
            
           </div>
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="development.html" class="btn btn-neutral float-right" title="Guidelines for Contributors" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="faqs.html" class="btn btn-neutral" title="Frequently Asked Questions" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2017, creisle.

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  

    <script type="text/javascript">
        var DOCUMENTATION_OPTIONS = {
            URL_ROOT:'./',
            VERSION:'',
            LANGUAGE:'None',
            COLLAPSE_INDEX:false,
            FILE_SUFFIX:'.html',
            HAS_SOURCE:  true,
            SOURCELINK_SUFFIX: '.txt'
        };
    </script>
      <script type="text/javascript" src="_static/jquery.js"></script>
      <script type="text/javascript" src="_static/underscore.js"></script>
      <script type="text/javascript" src="_static/doctools.js"></script>
      <script type="text/javascript" src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>

  

  
  
    <script type="text/javascript" src="_static/js/theme.js"></script>
  

  
  
  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.StickyNav.enable();
      });
  </script>
   

</body>
</html>