

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>genomic module &mdash; MAVIS  documentation</title>
  

  
  
  
  

  

  
  
    

  

  
  
    <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
  

  
    <link rel="stylesheet" href="../_static/custom.css" type="text/css" />
  

  
        <link rel="author" title="About these documents"
              href="../about.html"/>
        <link rel="index" title="Index"
              href="../genindex.html"/>
        <link rel="search" title="Search" href="../search.html"/>
    <link rel="top" title="MAVIS  documentation" href="../index.html"/>
        <link rel="up" title="annotate subpackage" href="../mavis.annotate.html"/>
        <link rel="next" title="main module" href="mavis.annotate.main.html"/>
        <link rel="prev" title="fusion module" href="mavis.annotate.fusion.html"/> 

  
  <script src="../_static/js/modernizr.min.js"></script>

</head>

<body class="wy-body-for-nav" role="document">

   
  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search">
          

          
            <a href="../index.html" class="icon icon-home"> MAVIS
          

          
          </a>

          
            
            
              <div class="version">
                1.6.4
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../about.html">About</a></li>
<li class="toctree-l1"><a class="reference internal" href="../pipeline.html">Running the Pipeline</a></li>
<li class="toctree-l1"><a class="reference internal" href="../mavis_input.html">MAVIS standard input file format</a></li>
<li class="toctree-l1"><a class="reference internal" href="../reference.html">Reference Input Files</a></li>
<li class="toctree-l1"><a class="reference internal" href="../configuration.html">Configuration and Settings</a></li>
<li class="toctree-l1"><a class="reference internal" href="../tools_supported.html">Supported SV Callers</a></li>
<li class="toctree-l1"><a class="reference internal" href="../illustrations.html">Illustrations</a></li>
<li class="toctree-l1"><a class="reference internal" href="../faqs.html">Frequently Asked Questions</a></li>
<li class="toctree-l1"><a class="reference internal" href="../theory.html">Theory and Models</a></li>
<li class="toctree-l1"><a class="reference internal" href="../development.html">Guidelines for Contributors</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../package.html">Package Documentation</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="mavis.align.html">align module</a></li>
<li class="toctree-l2 current"><a class="reference internal" href="../mavis.annotate.html">annotate subpackage</a><ul class="current">
<li class="toctree-l3"><a class="reference internal" href="../mavis.annotate.html#sub-package-documentation">Sub-package Documentation</a></li>
<li class="toctree-l3"><a class="reference internal" href="mavis.annotate.base.html">base module</a></li>
<li class="toctree-l3"><a class="reference internal" href="mavis.annotate.constants.html">constants module</a></li>
<li class="toctree-l3"><a class="reference internal" href="mavis.annotate.file_io.html">file_io module</a></li>
<li class="toctree-l3"><a class="reference internal" href="mavis.annotate.fusion.html">fusion module</a></li>
<li class="toctree-l3 current"><a class="current reference internal" href="#">genomic module</a></li>
<li class="toctree-l3"><a class="reference internal" href="mavis.annotate.main.html">main module</a></li>
<li class="toctree-l3"><a class="reference internal" href="mavis.annotate.protein.html">protein module</a></li>
<li class="toctree-l3"><a class="reference internal" href="mavis.annotate.splicing.html">splicing module</a></li>
<li class="toctree-l3"><a class="reference internal" href="mavis.annotate.variant.html">variant module</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="mavis.assemble.html">assemble module</a></li>
<li class="toctree-l2"><a class="reference internal" href="../mavis.bam.html">bam subpackage</a></li>
<li class="toctree-l2"><a class="reference internal" href="mavis.blat.html">blat module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mavis.breakpoint.html">breakpoint module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mavis.checker.html">checker module</a></li>
<li class="toctree-l2"><a class="reference internal" href="../mavis.cluster.html">cluster subpackage</a></li>
<li class="toctree-l2"><a class="reference internal" href="mavis.config.html">config module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mavis.constants.html">constants module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mavis.error.html">error module</a></li>
<li class="toctree-l2"><a class="reference internal" href="../mavis.illustrate.html">illustrate subpackage</a></li>
<li class="toctree-l2"><a class="reference internal" href="mavis.interval.html">interval module</a></li>
<li class="toctree-l2"><a class="reference internal" href="../mavis.pairing.html">pairing subpackage</a></li>
<li class="toctree-l2"><a class="reference internal" href="../mavis.validate.html">validate subpackage</a></li>
<li class="toctree-l2"><a class="reference internal" href="mavis.submit.html">submit module</a></li>
<li class="toctree-l2"><a class="reference internal" href="../mavis.summary.html">summary subpackage</a></li>
<li class="toctree-l2"><a class="reference internal" href="mavis.tools.html">tools module</a></li>
<li class="toctree-l2"><a class="reference internal" href="mavis.util.html">util module</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../glossary.html">Glossary</a></li>
<li class="toctree-l1"><a class="reference internal" href="../citations.html">Literature</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">MAVIS</a>
        
      </nav>


      
      <div class="wy-nav-content">
        <div class="rst-content">
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../index.html">Docs</a> &raquo;</li>
        
          <li><a href="../package.html">Package Documentation</a> &raquo;</li>
        
          <li><a href="../mavis.annotate.html">annotate subpackage</a> &raquo;</li>
        
      <li>genomic module</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="../_sources/auto/mavis.annotate.genomic.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="module-mavis.annotate.genomic">
<span id="genomic-module"></span><h1>genomic module<a class="headerlink" href="#module-mavis.annotate.genomic" title="Permalink to this headline">¶</a></h1>
<dl class="class">
<dt id="mavis.annotate.genomic.Exon">
<em class="property">class </em><code class="descclassname">mavis.annotate.genomic.</code><code class="descname">Exon</code><span class="sig-paren">(</span><em>start</em>, <em>end</em>, <em>transcript=None</em>, <em>name=None</em>, <em>intact_start_splice=True</em>, <em>intact_end_splice=True</em>, <em>seq=None</em>, <em>strand=None</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/mavis/annotate/genomic.html#Exon"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mavis.annotate.genomic.Exon" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="mavis.annotate.base.html#mavis.annotate.base.BioInterval" title="mavis.annotate.base.BioInterval"><code class="xref py py-class docutils literal"><span class="pre">mavis.annotate.base.BioInterval</span></code></a></p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>start</strong> (<a class="reference external" href="https://docs.python.org/3.6/library/functions.html#int" title="(in Python v3.6)"><em>int</em></a>) – the genomic start position</li>
<li><strong>end</strong> (<a class="reference external" href="https://docs.python.org/3.6/library/functions.html#int" title="(in Python v3.6)"><em>int</em></a>) – the genomic end position</li>
<li><strong>name</strong> (<a class="reference external" href="https://docs.python.org/3.6/library/stdtypes.html#str" title="(in Python v3.6)"><em>str</em></a>) – the name of the exon</li>
<li><strong>transcript</strong> (<a class="reference internal" href="#mavis.annotate.genomic.PreTranscript" title="mavis.annotate.genomic.PreTranscript"><em>PreTranscript</em></a>) – the ‘parent’ transcript this exon belongs to</li>
<li><strong>intact_start_splice</strong> (<a class="reference external" href="https://docs.python.org/3.6/library/functions.html#bool" title="(in Python v3.6)"><em>bool</em></a>) – if the starting splice site has been abrogated</li>
<li><strong>intact_end_splice</strong> (<a class="reference external" href="https://docs.python.org/3.6/library/functions.html#bool" title="(in Python v3.6)"><em>bool</em></a>) – if the end splice site has been abrogated</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Raises:</th><td class="field-body"><p class="first last"><a class="reference external" href="https://docs.python.org/3.6/library/exceptions.html#AttributeError" title="(in Python v3.6)"><code class="xref py py-exc docutils literal"><span class="pre">AttributeError</span></code></a> – if the exon start &gt; the exon end</p>
</td>
</tr>
</tbody>
</table>
<p class="rubric">Example</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">Exon</span><span class="p">(</span><span class="mi">15</span><span class="p">,</span> <span class="mi">78</span><span class="p">)</span>
</pre></div>
</div>
<dl class="attribute">
<dt id="mavis.annotate.genomic.Exon.acceptor">
<code class="descname">acceptor</code><a class="headerlink" href="#mavis.annotate.genomic.Exon.acceptor" title="Permalink to this definition">¶</a></dt>
<dd><p><cite>int</cite> – returns the genomic exonic position of the acceptor splice site</p>
</dd></dl>

<dl class="attribute">
<dt id="mavis.annotate.genomic.Exon.acceptor_splice_site">
<code class="descname">acceptor_splice_site</code><a class="headerlink" href="#mavis.annotate.genomic.Exon.acceptor_splice_site" title="Permalink to this definition">¶</a></dt>
<dd><p><a class="reference internal" href="mavis.interval.html#mavis.interval.Interval" title="mavis.interval.Interval"><code class="xref py py-class docutils literal"><span class="pre">Interval</span></code></a> – the genomic range describing the splice site</p>
</dd></dl>

<dl class="attribute">
<dt id="mavis.annotate.genomic.Exon.donor">
<code class="descname">donor</code><a class="headerlink" href="#mavis.annotate.genomic.Exon.donor" title="Permalink to this definition">¶</a></dt>
<dd><p><cite>int</cite> – returns the genomic exonic position of the donor splice site</p>
</dd></dl>

<dl class="attribute">
<dt id="mavis.annotate.genomic.Exon.donor_splice_site">
<code class="descname">donor_splice_site</code><a class="headerlink" href="#mavis.annotate.genomic.Exon.donor_splice_site" title="Permalink to this definition">¶</a></dt>
<dd><p><a class="reference internal" href="mavis.interval.html#mavis.interval.Interval" title="mavis.interval.Interval"><code class="xref py py-class docutils literal"><span class="pre">Interval</span></code></a> – the genomic range describing the splice site</p>
</dd></dl>

<dl class="attribute">
<dt id="mavis.annotate.genomic.Exon.transcript">
<code class="descname">transcript</code><a class="headerlink" href="#mavis.annotate.genomic.Exon.transcript" title="Permalink to this definition">¶</a></dt>
<dd><p><a class="reference internal" href="#mavis.annotate.genomic.PreTranscript" title="mavis.annotate.genomic.PreTranscript"><code class="xref py py-class docutils literal"><span class="pre">PreTranscript</span></code></a> – the transcript this exon belongs to</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="mavis.annotate.genomic.Gene">
<em class="property">class </em><code class="descclassname">mavis.annotate.genomic.</code><code class="descname">Gene</code><span class="sig-paren">(</span><em>chr</em>, <em>start</em>, <em>end</em>, <em>name=None</em>, <em>strand='?'</em>, <em>aliases=None</em>, <em>seq=None</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/mavis/annotate/genomic.html#Gene"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mavis.annotate.genomic.Gene" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="mavis.annotate.base.html#mavis.annotate.base.BioInterval" title="mavis.annotate.base.BioInterval"><code class="xref py py-class docutils literal"><span class="pre">mavis.annotate.base.BioInterval</span></code></a></p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>chr</strong> (<a class="reference external" href="https://docs.python.org/3.6/library/stdtypes.html#str" title="(in Python v3.6)"><em>str</em></a>) – the chromosome</li>
<li><strong>name</strong> (<a class="reference external" href="https://docs.python.org/3.6/library/stdtypes.html#str" title="(in Python v3.6)"><em>str</em></a>) – the gene name/id i.e. ENSG0001</li>
<li><strong>strand</strong> (<a class="reference internal" href="mavis.constants.html#mavis.constants.STRAND" title="mavis.constants.STRAND"><em>STRAND</em></a>) – the genomic strand ‘+’ or ‘-‘</li>
<li><strong>aliases</strong> (<a class="reference external" href="https://docs.python.org/3.6/library/stdtypes.html#list" title="(in Python v3.6)"><code class="xref py py-class docutils literal"><span class="pre">list</span></code></a> of <a class="reference external" href="https://docs.python.org/3.6/library/stdtypes.html#str" title="(in Python v3.6)"><code class="xref py py-class docutils literal"><span class="pre">str</span></code></a>) – a list of aliases. For example the hugo name could go here</li>
<li><strong>seq</strong> (<a class="reference external" href="https://docs.python.org/3.6/library/stdtypes.html#str" title="(in Python v3.6)"><em>str</em></a>) – genomic seq of the gene</li>
</ul>
</td>
</tr>
</tbody>
</table>
<p class="rubric">Example</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">Gene</span><span class="p">(</span><span class="s1">&#39;X&#39;</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">1000</span><span class="p">,</span> <span class="s1">&#39;ENG0001&#39;</span><span class="p">,</span> <span class="s1">&#39;+&#39;</span><span class="p">,</span> <span class="p">[</span><span class="s1">&#39;KRAS&#39;</span><span class="p">])</span>
</pre></div>
</div>
<dl class="attribute">
<dt id="mavis.annotate.genomic.Gene.chr">
<code class="descname">chr</code><a class="headerlink" href="#mavis.annotate.genomic.Gene.chr" title="Permalink to this definition">¶</a></dt>
<dd><p>returns the name of the chromosome that this gene resides on</p>
</dd></dl>

<dl class="method">
<dt id="mavis.annotate.genomic.Gene.get_seq">
<code class="descname">get_seq</code><span class="sig-paren">(</span><em>reference_genome</em>, <em>ignore_cache=False</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/mavis/annotate/genomic.html#Gene.get_seq"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mavis.annotate.genomic.Gene.get_seq" title="Permalink to this definition">¶</a></dt>
<dd><p>gene sequence is always given wrt to the positive forward strand regardless of gene strand</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>reference_genome</strong> (<a class="reference external" href="https://docs.python.org/3.6/library/stdtypes.html#dict" title="(in Python v3.6)"><code class="xref py py-class docutils literal"><span class="pre">dict</span></code></a> of <code class="xref py py-class docutils literal"><span class="pre">Bio.SeqRecord</span></code> by <a class="reference external" href="https://docs.python.org/3.6/library/stdtypes.html#str" title="(in Python v3.6)"><code class="xref py py-class docutils literal"><span class="pre">str</span></code></a>) – dict of reference sequence by
template/chr name</li>
<li><strong>ignore_cache</strong> (<a class="reference external" href="https://docs.python.org/3.6/library/functions.html#bool" title="(in Python v3.6)"><em>bool</em></a>) – if True then stored sequences will be ignored and the function will attempt to retrieve the sequence using the positions and the input reference_genome</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">the sequence of the gene</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last"><a class="reference external" href="https://docs.python.org/3.6/library/stdtypes.html#str" title="(in Python v3.6)">str</a></p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="mavis.annotate.genomic.Gene.key">
<code class="descname">key</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/mavis/annotate/genomic.html#Gene.key"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mavis.annotate.genomic.Gene.key" title="Permalink to this definition">¶</a></dt>
<dd><p>see <code class="xref py py-func docutils literal"><span class="pre">structural_variant.annotate.base.BioInterval.key()</span></code></p>
</dd></dl>

<dl class="attribute">
<dt id="mavis.annotate.genomic.Gene.spliced_transcripts">
<code class="descname">spliced_transcripts</code><a class="headerlink" href="#mavis.annotate.genomic.Gene.spliced_transcripts" title="Permalink to this definition">¶</a></dt>
<dd><p><a class="reference external" href="https://docs.python.org/3.6/library/stdtypes.html#list" title="(in Python v3.6)"><code class="xref any docutils literal"><span class="pre">list</span></code></a> of <a class="reference internal" href="#mavis.annotate.genomic.Transcript" title="mavis.annotate.genomic.Transcript"><code class="xref py py-class docutils literal"><span class="pre">Transcript</span></code></a> – list of transcripts</p>
</dd></dl>

<dl class="method">
<dt id="mavis.annotate.genomic.Gene.to_dict">
<code class="descname">to_dict</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/mavis/annotate/genomic.html#Gene.to_dict"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mavis.annotate.genomic.Gene.to_dict" title="Permalink to this definition">¶</a></dt>
<dd><p>see <code class="xref py py-func docutils literal"><span class="pre">structural_variant.annotate.base.BioInterval.to_dict()</span></code></p>
</dd></dl>

<dl class="method">
<dt id="mavis.annotate.genomic.Gene.transcript_priority">
<code class="descname">transcript_priority</code><span class="sig-paren">(</span><em>transcript</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/mavis/annotate/genomic.html#Gene.transcript_priority"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mavis.annotate.genomic.Gene.transcript_priority" title="Permalink to this definition">¶</a></dt>
<dd><p>prioritizes transcripts from 0 to n-1 based on best transcript flag
and then alphanumeric name sort</p>
<div class="admonition warning">
<p class="first admonition-title">Warning</p>
<p class="last">Lower number means higher priority. This is to make sort work by default</p>
</div>
</dd></dl>

<dl class="attribute">
<dt id="mavis.annotate.genomic.Gene.transcripts">
<code class="descname">transcripts</code><a class="headerlink" href="#mavis.annotate.genomic.Gene.transcripts" title="Permalink to this definition">¶</a></dt>
<dd><p><a class="reference external" href="https://docs.python.org/3.6/library/stdtypes.html#list" title="(in Python v3.6)"><code class="xref any docutils literal"><span class="pre">list</span></code></a> of <a class="reference internal" href="#mavis.annotate.genomic.PreTranscript" title="mavis.annotate.genomic.PreTranscript"><code class="xref py py-class docutils literal"><span class="pre">PreTranscript</span></code></a> – list of unspliced transcripts</p>
</dd></dl>

<dl class="attribute">
<dt id="mavis.annotate.genomic.Gene.translations">
<code class="descname">translations</code><a class="headerlink" href="#mavis.annotate.genomic.Gene.translations" title="Permalink to this definition">¶</a></dt>
<dd><p><a class="reference external" href="https://docs.python.org/3.6/library/stdtypes.html#list" title="(in Python v3.6)"><code class="xref any docutils literal"><span class="pre">list</span></code></a> of <a class="reference internal" href="mavis.annotate.protein.html#mavis.annotate.protein.Translation" title="mavis.annotate.protein.Translation"><code class="xref py py-class docutils literal"><span class="pre">Translation</span></code></a> – list of translations</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="mavis.annotate.genomic.IntergenicRegion">
<em class="property">class </em><code class="descclassname">mavis.annotate.genomic.</code><code class="descname">IntergenicRegion</code><span class="sig-paren">(</span><em>chr</em>, <em>start</em>, <em>end</em>, <em>strand</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/mavis/annotate/genomic.html#IntergenicRegion"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mavis.annotate.genomic.IntergenicRegion" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="mavis.annotate.base.html#mavis.annotate.base.BioInterval" title="mavis.annotate.base.BioInterval"><code class="xref py py-class docutils literal"><span class="pre">mavis.annotate.base.BioInterval</span></code></a></p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>chr</strong> (<a class="reference external" href="https://docs.python.org/3.6/library/stdtypes.html#str" title="(in Python v3.6)"><em>str</em></a>) – the reference object/chromosome for this region</li>
<li><strong>start</strong> (<a class="reference external" href="https://docs.python.org/3.6/library/functions.html#int" title="(in Python v3.6)"><em>int</em></a>) – the start of the IntergenicRegion</li>
<li><strong>end</strong> (<a class="reference external" href="https://docs.python.org/3.6/library/functions.html#int" title="(in Python v3.6)"><em>int</em></a>) – the end of the IntergenicRegion</li>
<li><strong>strand</strong> (<a class="reference internal" href="mavis.constants.html#mavis.constants.STRAND" title="mavis.constants.STRAND"><em>STRAND</em></a>) – the strand the region is defined on</li>
</ul>
</td>
</tr>
</tbody>
</table>
<p class="rubric">Example</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">IntergenicRegion</span><span class="p">(</span><span class="s1">&#39;1&#39;</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">100</span><span class="p">,</span> <span class="s1">&#39;+&#39;</span><span class="p">)</span>
</pre></div>
</div>
<dl class="attribute">
<dt id="mavis.annotate.genomic.IntergenicRegion.chr">
<code class="descname">chr</code><a class="headerlink" href="#mavis.annotate.genomic.IntergenicRegion.chr" title="Permalink to this definition">¶</a></dt>
<dd><p>returns the name of the chromosome that this region resides on</p>
</dd></dl>

<dl class="method">
<dt id="mavis.annotate.genomic.IntergenicRegion.key">
<code class="descname">key</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/mavis/annotate/genomic.html#IntergenicRegion.key"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mavis.annotate.genomic.IntergenicRegion.key" title="Permalink to this definition">¶</a></dt>
<dd><p>see <code class="xref py py-func docutils literal"><span class="pre">structural_variant.annotate.base.BioInterval.key()</span></code></p>
</dd></dl>

<dl class="method">
<dt id="mavis.annotate.genomic.IntergenicRegion.to_dict">
<code class="descname">to_dict</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/mavis/annotate/genomic.html#IntergenicRegion.to_dict"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mavis.annotate.genomic.IntergenicRegion.to_dict" title="Permalink to this definition">¶</a></dt>
<dd><p>see <code class="xref py py-func docutils literal"><span class="pre">structural_variant.annotate.base.BioInterval.to_dict()</span></code></p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="mavis.annotate.genomic.PreTranscript">
<em class="property">class </em><code class="descclassname">mavis.annotate.genomic.</code><code class="descname">PreTranscript</code><span class="sig-paren">(</span><em>exons</em>, <em>gene=None</em>, <em>name=None</em>, <em>strand=None</em>, <em>spliced_transcripts=None</em>, <em>seq=None</em>, <em>is_best_transcript=False</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/mavis/annotate/genomic.html#PreTranscript"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mavis.annotate.genomic.PreTranscript" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="mavis.annotate.base.html#mavis.annotate.base.BioInterval" title="mavis.annotate.base.BioInterval"><code class="xref py py-class docutils literal"><span class="pre">mavis.annotate.base.BioInterval</span></code></a></p>
<p>creates a new transcript object</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>exons</strong> (<a class="reference external" href="https://docs.python.org/3.6/library/stdtypes.html#list" title="(in Python v3.6)"><code class="xref py py-class docutils literal"><span class="pre">list</span></code></a> of <a class="reference internal" href="#mavis.annotate.genomic.Exon" title="mavis.annotate.genomic.Exon"><code class="xref any py py-class docutils literal"><span class="pre">Exon</span></code></a>) – list of Exon that make up the transcript</li>
<li><strong>genomic_start</strong> (<a class="reference external" href="https://docs.python.org/3.6/library/functions.html#int" title="(in Python v3.6)"><em>int</em></a>) – genomic start position of the transcript</li>
<li><strong>genomic_end</strong> (<a class="reference external" href="https://docs.python.org/3.6/library/functions.html#int" title="(in Python v3.6)"><em>int</em></a>) – genomic end position of the transcript</li>
<li><strong>gene</strong> (<a class="reference internal" href="#mavis.annotate.genomic.Gene" title="mavis.annotate.genomic.Gene"><em>Gene</em></a>) – the gene this transcript belongs to</li>
<li><strong>name</strong> (<a class="reference external" href="https://docs.python.org/3.6/library/stdtypes.html#str" title="(in Python v3.6)"><em>str</em></a>) – name of the transcript</li>
<li><strong>strand</strong> (<a class="reference internal" href="mavis.constants.html#mavis.constants.STRAND" title="mavis.constants.STRAND"><em>STRAND</em></a>) – strand the transcript is on, defaults to the strand of the Gene if not specified</li>
<li><strong>seq</strong> (<a class="reference external" href="https://docs.python.org/3.6/library/stdtypes.html#str" title="(in Python v3.6)"><em>str</em></a>) – unspliced cDNA seq</li>
</ul>
</td>
</tr>
</tbody>
</table>
<dl class="method">
<dt id="mavis.annotate.genomic.PreTranscript.convert_cdna_to_genomic">
<code class="descname">convert_cdna_to_genomic</code><span class="sig-paren">(</span><em>pos</em>, <em>splicing_pattern</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/mavis/annotate/genomic.html#PreTranscript.convert_cdna_to_genomic"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mavis.annotate.genomic.PreTranscript.convert_cdna_to_genomic" title="Permalink to this definition">¶</a></dt>
<dd><table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>pos</strong> (<a class="reference external" href="https://docs.python.org/3.6/library/functions.html#int" title="(in Python v3.6)"><em>int</em></a>) – cdna position</li>
<li><strong>splicing_pattern</strong> (<a class="reference internal" href="mavis.annotate.splicing.html#mavis.annotate.splicing.SplicingPattern" title="mavis.annotate.splicing.SplicingPattern"><em>SplicingPattern</em></a>) – list of genomic splice sites 3‘5’ repeating</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">the genomic equivalent</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last"><a class="reference external" href="https://docs.python.org/3.6/library/functions.html#int" title="(in Python v3.6)">int</a></p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="mavis.annotate.genomic.PreTranscript.convert_genomic_to_cdna">
<code class="descname">convert_genomic_to_cdna</code><span class="sig-paren">(</span><em>pos</em>, <em>splicing_pattern</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/mavis/annotate/genomic.html#PreTranscript.convert_genomic_to_cdna"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mavis.annotate.genomic.PreTranscript.convert_genomic_to_cdna" title="Permalink to this definition">¶</a></dt>
<dd><table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>pos</strong> (<a class="reference external" href="https://docs.python.org/3.6/library/functions.html#int" title="(in Python v3.6)"><em>int</em></a>) – the genomic position to be converted</li>
<li><strong>splicing_pattern</strong> (<a class="reference internal" href="mavis.annotate.splicing.html#mavis.annotate.splicing.SplicingPattern" title="mavis.annotate.splicing.SplicingPattern"><em>SplicingPattern</em></a>) – list of genomic splice sites 3‘5’ repeating</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">the cdna equivalent</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first"><a class="reference external" href="https://docs.python.org/3.6/library/functions.html#int" title="(in Python v3.6)">int</a></p>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Raises:</th><td class="field-body"><p class="first last"><code class="xref py py-class docutils literal"><span class="pre">IndexError</span></code> – when a genomic position not present in the
cdna is attempted to be converted</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="mavis.annotate.genomic.PreTranscript.convert_genomic_to_nearest_cdna">
<code class="descname">convert_genomic_to_nearest_cdna</code><span class="sig-paren">(</span><em>pos</em>, <em>splicing_pattern</em>, <em>stick_direction=None</em>, <em>allow_outside=True</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/mavis/annotate/genomic.html#PreTranscript.convert_genomic_to_nearest_cdna"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mavis.annotate.genomic.PreTranscript.convert_genomic_to_nearest_cdna" title="Permalink to this definition">¶</a></dt>
<dd><p>converts a genomic position to its cdna equivalent or (if intronic) the nearest cdna and shift</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>pos</strong> (<a class="reference external" href="https://docs.python.org/3.6/library/functions.html#int" title="(in Python v3.6)"><em>int</em></a>) – the genomic position</li>
<li><strong>splicing_pattern</strong> (<a class="reference internal" href="mavis.annotate.splicing.html#mavis.annotate.splicing.SplicingPattern" title="mavis.annotate.splicing.SplicingPattern"><em>SplicingPattern</em></a>) – the splicing pattern</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first"><ul class="simple">
<li><em>int</em> - the exonic cdna position</li>
<li><em>int</em> - the intronic shift</li>
</ul>
</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last">tuple of int and int</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="mavis.annotate.genomic.PreTranscript.exon_number">
<code class="descname">exon_number</code><span class="sig-paren">(</span><em>exon</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/mavis/annotate/genomic.html#PreTranscript.exon_number"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mavis.annotate.genomic.PreTranscript.exon_number" title="Permalink to this definition">¶</a></dt>
<dd><p>exon numbering is based on the direction of translation</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>exon</strong> (<a class="reference internal" href="#mavis.annotate.genomic.Exon" title="mavis.annotate.genomic.Exon"><em>Exon</em></a>) – the exon to be numbered</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">the exon number (1 based)</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><a class="reference external" href="https://docs.python.org/3.6/library/functions.html#int" title="(in Python v3.6)">int</a></td>
</tr>
<tr class="field-even field"><th class="field-name">Raises:</th><td class="field-body"><a class="reference external" href="https://docs.python.org/3.6/library/exceptions.html#AttributeError" title="(in Python v3.6)"><code class="xref py py-exc docutils literal"><span class="pre">AttributeError</span></code></a> – if the strand is not given or the exon does not belong to the transcript</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="attribute">
<dt id="mavis.annotate.genomic.PreTranscript.gene">
<code class="descname">gene</code><a class="headerlink" href="#mavis.annotate.genomic.PreTranscript.gene" title="Permalink to this definition">¶</a></dt>
<dd><p><a class="reference internal" href="#mavis.annotate.genomic.Gene" title="mavis.annotate.genomic.Gene"><code class="xref any py py-class docutils literal"><span class="pre">Gene</span></code></a> – the gene this transcript belongs to</p>
</dd></dl>

<dl class="method">
<dt id="mavis.annotate.genomic.PreTranscript.generate_splicing_patterns">
<code class="descname">generate_splicing_patterns</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/mavis/annotate/genomic.html#PreTranscript.generate_splicing_patterns"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mavis.annotate.genomic.PreTranscript.generate_splicing_patterns" title="Permalink to this definition">¶</a></dt>
<dd><p>returns a list of splice sites to be connected as a splicing pattern</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">List of positions to be spliced together</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body"><a class="reference external" href="https://docs.python.org/3.6/library/stdtypes.html#list" title="(in Python v3.6)"><code class="xref py py-class docutils literal"><span class="pre">list</span></code></a> of <code class="xref py py-class docutils literal"><span class="pre">SplicingPattern</span></code></td>
</tr>
</tbody>
</table>
<p>see <a class="reference internal" href="../theory.html#theory-predicting-splicing-patterns"><span class="std std-ref">theory - predicting splicing patterns</span></a></p>
</dd></dl>

<dl class="method">
<dt id="mavis.annotate.genomic.PreTranscript.get_cdna_seq">
<code class="descname">get_cdna_seq</code><span class="sig-paren">(</span><em>splicing_pattern</em>, <em>reference_genome=None</em>, <em>ignore_cache=False</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/mavis/annotate/genomic.html#PreTranscript.get_cdna_seq"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mavis.annotate.genomic.PreTranscript.get_cdna_seq" title="Permalink to this definition">¶</a></dt>
<dd><table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>splicing_pattern</strong> (<a class="reference internal" href="mavis.annotate.splicing.html#mavis.annotate.splicing.SplicingPattern" title="mavis.annotate.splicing.SplicingPattern"><em>SplicingPattern</em></a>) – the list of splicing positions</li>
<li><strong>reference_genome</strong> (<a class="reference external" href="https://docs.python.org/3.6/library/stdtypes.html#dict" title="(in Python v3.6)"><code class="xref py py-class docutils literal"><span class="pre">dict</span></code></a> of <code class="xref py py-class docutils literal"><span class="pre">Bio.SeqRecord</span></code> by <a class="reference external" href="https://docs.python.org/3.6/library/stdtypes.html#str" title="(in Python v3.6)"><code class="xref py py-class docutils literal"><span class="pre">str</span></code></a>) – dict of reference sequence
by template/chr name</li>
<li><strong>ignore_cache</strong> (<a class="reference external" href="https://docs.python.org/3.6/library/functions.html#bool" title="(in Python v3.6)"><em>bool</em></a>) – if True then stored sequences will be ignored and the function will attempt to retrieve the sequence using the positions and the input reference_genome</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">the spliced cDNA sequence</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last"><a class="reference external" href="https://docs.python.org/3.6/library/stdtypes.html#str" title="(in Python v3.6)">str</a></p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="mavis.annotate.genomic.PreTranscript.get_seq">
<code class="descname">get_seq</code><span class="sig-paren">(</span><em>reference_genome=None</em>, <em>ignore_cache=False</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/mavis/annotate/genomic.html#PreTranscript.get_seq"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mavis.annotate.genomic.PreTranscript.get_seq" title="Permalink to this definition">¶</a></dt>
<dd><table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>reference_genome</strong> (<a class="reference external" href="https://docs.python.org/3.6/library/stdtypes.html#dict" title="(in Python v3.6)"><code class="xref py py-class docutils literal"><span class="pre">dict</span></code></a> of <code class="xref py py-class docutils literal"><span class="pre">Bio.SeqRecord</span></code> by <a class="reference external" href="https://docs.python.org/3.6/library/stdtypes.html#str" title="(in Python v3.6)"><code class="xref py py-class docutils literal"><span class="pre">str</span></code></a>) – dict of reference sequence
by template/chr name</li>
<li><strong>ignore_cache</strong> (<a class="reference external" href="https://docs.python.org/3.6/library/functions.html#bool" title="(in Python v3.6)"><em>bool</em></a>) – if True then stored sequences will be ignored and the function will attempt to retrieve the sequence using the positions and the input reference_genome</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">the sequence of the transcript including introns (but relative to strand)</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last"><a class="reference external" href="https://docs.python.org/3.6/library/stdtypes.html#str" title="(in Python v3.6)">str</a></p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="attribute">
<dt id="mavis.annotate.genomic.PreTranscript.transcripts">
<code class="descname">transcripts</code><a class="headerlink" href="#mavis.annotate.genomic.PreTranscript.transcripts" title="Permalink to this definition">¶</a></dt>
<dd><p><a class="reference external" href="https://docs.python.org/3.6/library/stdtypes.html#list" title="(in Python v3.6)"><code class="xref py py-class docutils literal"><span class="pre">list</span></code></a> of <a class="reference internal" href="#mavis.annotate.genomic.Transcript" title="mavis.annotate.genomic.Transcript"><code class="xref py py-class docutils literal"><span class="pre">Transcript</span></code></a> – list of spliced transcripts</p>
</dd></dl>

<dl class="attribute">
<dt id="mavis.annotate.genomic.PreTranscript.translations">
<code class="descname">translations</code><a class="headerlink" href="#mavis.annotate.genomic.PreTranscript.translations" title="Permalink to this definition">¶</a></dt>
<dd><p><a class="reference external" href="https://docs.python.org/3.6/library/stdtypes.html#list" title="(in Python v3.6)"><code class="xref py py-class docutils literal"><span class="pre">list</span></code></a> of <a class="reference internal" href="mavis.annotate.protein.html#mavis.annotate.protein.Translation" title="mavis.annotate.protein.Translation"><code class="xref py py-class docutils literal"><span class="pre">Translation</span></code></a> – list of translations associated with this transcript</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="mavis.annotate.genomic.Template">
<em class="property">class </em><code class="descclassname">mavis.annotate.genomic.</code><code class="descname">Template</code><span class="sig-paren">(</span><em>name</em>, <em>start</em>, <em>end</em>, <em>seq=None</em>, <em>bands=None</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/mavis/annotate/genomic.html#Template"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mavis.annotate.genomic.Template" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="mavis.annotate.base.html#mavis.annotate.base.BioInterval" title="mavis.annotate.base.BioInterval"><code class="xref py py-class docutils literal"><span class="pre">mavis.annotate.base.BioInterval</span></code></a></p>
</dd></dl>

<dl class="class">
<dt id="mavis.annotate.genomic.Transcript">
<em class="property">class </em><code class="descclassname">mavis.annotate.genomic.</code><code class="descname">Transcript</code><span class="sig-paren">(</span><em>pre_transcript</em>, <em>splicing_patt</em>, <em>seq=None</em>, <em>translations=None</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/mavis/annotate/genomic.html#Transcript"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mavis.annotate.genomic.Transcript" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="mavis.annotate.base.html#mavis.annotate.base.BioInterval" title="mavis.annotate.base.BioInterval"><code class="xref py py-class docutils literal"><span class="pre">mavis.annotate.base.BioInterval</span></code></a></p>
<p>splicing pattern is given in genomic coordinates</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>pre_transcript</strong> (<a class="reference internal" href="#mavis.annotate.genomic.PreTranscript" title="mavis.annotate.genomic.PreTranscript"><em>PreTranscript</em></a>) – the unspliced transcript</li>
<li><strong>splicing_patt</strong> (<a class="reference external" href="https://docs.python.org/3.6/library/stdtypes.html#list" title="(in Python v3.6)"><code class="xref py py-class docutils literal"><span class="pre">list</span></code></a> of <a class="reference external" href="https://docs.python.org/3.6/library/functions.html#int" title="(in Python v3.6)"><code class="xref py py-class docutils literal"><span class="pre">int</span></code></a>) – the list of splicing positions</li>
<li><strong>seq</strong> (<a class="reference external" href="https://docs.python.org/3.6/library/stdtypes.html#str" title="(in Python v3.6)"><em>str</em></a>) – the cdna sequence</li>
<li><strong>translations</strong> (<a class="reference external" href="https://docs.python.org/3.6/library/stdtypes.html#list" title="(in Python v3.6)"><code class="xref py py-class docutils literal"><span class="pre">list</span></code></a> of <a class="reference internal" href="mavis.annotate.protein.html#mavis.annotate.protein.Translation" title="mavis.annotate.protein.Translation"><code class="xref py py-class docutils literal"><span class="pre">Translation</span></code></a>) – the list of translations of this transcript</li>
</ul>
</td>
</tr>
</tbody>
</table>
<dl class="method">
<dt id="mavis.annotate.genomic.Transcript.convert_cdna_to_genomic">
<code class="descname">convert_cdna_to_genomic</code><span class="sig-paren">(</span><em>pos</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/mavis/annotate/genomic.html#Transcript.convert_cdna_to_genomic"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mavis.annotate.genomic.Transcript.convert_cdna_to_genomic" title="Permalink to this definition">¶</a></dt>
<dd><table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>pos</strong> (<a class="reference external" href="https://docs.python.org/3.6/library/functions.html#int" title="(in Python v3.6)"><em>int</em></a>) – cdna position</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">the genomic equivalent</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><a class="reference external" href="https://docs.python.org/3.6/library/functions.html#int" title="(in Python v3.6)">int</a></td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="mavis.annotate.genomic.Transcript.convert_genomic_to_cdna">
<code class="descname">convert_genomic_to_cdna</code><span class="sig-paren">(</span><em>pos</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/mavis/annotate/genomic.html#Transcript.convert_genomic_to_cdna"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mavis.annotate.genomic.Transcript.convert_genomic_to_cdna" title="Permalink to this definition">¶</a></dt>
<dd><table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>pos</strong> (<a class="reference external" href="https://docs.python.org/3.6/library/functions.html#int" title="(in Python v3.6)"><em>int</em></a>) – the genomic position to be converted</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">the cdna equivalent</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><a class="reference external" href="https://docs.python.org/3.6/library/functions.html#int" title="(in Python v3.6)">int</a></td>
</tr>
<tr class="field-even field"><th class="field-name">Raises:</th><td class="field-body"><a class="reference external" href="https://docs.python.org/3.6/library/exceptions.html#IndexError" title="(in Python v3.6)"><code class="xref py py-exc docutils literal"><span class="pre">IndexError</span></code></a> – when a genomic position not present in the cdna is attempted to be converted</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="mavis.annotate.genomic.Transcript.convert_genomic_to_nearest_cdna">
<code class="descname">convert_genomic_to_nearest_cdna</code><span class="sig-paren">(</span><em>pos</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/mavis/annotate/genomic.html#Transcript.convert_genomic_to_nearest_cdna"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mavis.annotate.genomic.Transcript.convert_genomic_to_nearest_cdna" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="mavis.annotate.genomic.Transcript.get_seq">
<code class="descname">get_seq</code><span class="sig-paren">(</span><em>reference_genome=None</em>, <em>ignore_cache=False</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/mavis/annotate/genomic.html#Transcript.get_seq"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mavis.annotate.genomic.Transcript.get_seq" title="Permalink to this definition">¶</a></dt>
<dd><table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>reference_genome</strong> (<a class="reference external" href="https://docs.python.org/3.6/library/stdtypes.html#dict" title="(in Python v3.6)"><code class="xref py py-class docutils literal"><span class="pre">dict</span></code></a> of <code class="xref py py-class docutils literal"><span class="pre">Bio.SeqRecord</span></code> by <a class="reference external" href="https://docs.python.org/3.6/library/stdtypes.html#str" title="(in Python v3.6)"><code class="xref py py-class docutils literal"><span class="pre">str</span></code></a>) – dict of reference sequence by
template/chr name</li>
<li><strong>ignore_cache</strong> (<a class="reference external" href="https://docs.python.org/3.6/library/functions.html#bool" title="(in Python v3.6)"><em>bool</em></a>) – if True then stored sequences will be ignored and the function will attempt to retrieve the sequence using the positions and the input reference_genome</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">the sequence corresponding to the spliced cdna</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last"><a class="reference external" href="https://docs.python.org/3.6/library/stdtypes.html#str" title="(in Python v3.6)">str</a></p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="attribute">
<dt id="mavis.annotate.genomic.Transcript.unspliced_transcript">
<code class="descname">unspliced_transcript</code><a class="headerlink" href="#mavis.annotate.genomic.Transcript.unspliced_transcript" title="Permalink to this definition">¶</a></dt>
<dd><p><a class="reference internal" href="#mavis.annotate.genomic.PreTranscript" title="mavis.annotate.genomic.PreTranscript"><code class="xref py py-class docutils literal"><span class="pre">PreTranscript</span></code></a> – the unspliced transcript this splice variant belongs to</p>
</dd></dl>

</dd></dl>

</div>


           </div>
           <div class="articleComments">
            
           </div>
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="mavis.annotate.main.html" class="btn btn-neutral float-right" title="main module" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="mavis.annotate.fusion.html" class="btn btn-neutral" title="fusion module" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2017, creisle.

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  

    <script type="text/javascript">
        var DOCUMENTATION_OPTIONS = {
            URL_ROOT:'../',
            VERSION:'',
            LANGUAGE:'None',
            COLLAPSE_INDEX:false,
            FILE_SUFFIX:'.html',
            HAS_SOURCE:  true,
            SOURCELINK_SUFFIX: '.txt'
        };
    </script>
      <script type="text/javascript" src="../_static/jquery.js"></script>
      <script type="text/javascript" src="../_static/underscore.js"></script>
      <script type="text/javascript" src="../_static/doctools.js"></script>
      <script type="text/javascript" src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>

  

  
  
    <script type="text/javascript" src="../_static/js/theme.js"></script>
  

  
  
  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.StickyNav.enable();
      });
  </script>
   

</body>
</html>