

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>mavis.bam.stats &mdash; MAVIS  documentation</title>
  

  
  
  
  

  

  
  
    

  

  
  
    <link rel="stylesheet" href="../../../_static/css/theme.css" type="text/css" />
  

  
    <link rel="stylesheet" href="../../../_static/custom.css" type="text/css" />
  

  
        <link rel="author" title="About these documents"
              href="../../../about.html"/>
        <link rel="index" title="Index"
              href="../../../genindex.html"/>
        <link rel="search" title="Search" href="../../../search.html"/>
    <link rel="top" title="MAVIS  documentation" href="../../../index.html"/>
        <link rel="up" title="Module code" href="../../index.html"/> 

  
  <script src="../../../_static/js/modernizr.min.js"></script>

</head>

<body class="wy-body-for-nav" role="document">

   
  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search">
          

          
            <a href="../../../index.html" class="icon icon-home"> MAVIS
          

          
          </a>

          
            
            
              <div class="version">
                1.6.4
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul>
<li class="toctree-l1"><a class="reference internal" href="../../../about.html">About</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../pipeline.html">Running the Pipeline</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../mavis_input.html">MAVIS standard input file format</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../reference.html">Reference Input Files</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../configuration.html">Configuration and Settings</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../tools_supported.html">Supported SV Callers</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../illustrations.html">Illustrations</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../faqs.html">Frequently Asked Questions</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../theory.html">Theory and Models</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../development.html">Guidelines for Contributors</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../package.html">Package Documentation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../glossary.html">Glossary</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../citations.html">Literature</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../../index.html">MAVIS</a>
        
      </nav>


      
      <div class="wy-nav-content">
        <div class="rst-content">
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../../../index.html">Docs</a> &raquo;</li>
        
          <li><a href="../../index.html">Module code</a> &raquo;</li>
        
      <li>mavis.bam.stats</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <h1>Source code for mavis.bam.stats</h1><div class="highlight"><pre>
<span></span><span class="ch">#!/projects/tumour_char/analysis_scripts/python/centos06/anaconda3_v2.3.0/bin/python</span>
<span class="kn">import</span> <span class="nn">math</span>
<span class="kn">import</span> <span class="nn">statistics</span> <span class="k">as</span> <span class="nn">stats</span>
<span class="kn">import</span> <span class="nn">warnings</span>

<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>

<span class="kn">from</span> <span class="nn">.read</span> <span class="k">import</span> <span class="n">sequenced_strand</span>
<span class="kn">from</span> <span class="nn">..constants</span> <span class="k">import</span> <span class="n">STRAND</span>
<span class="kn">from</span> <span class="nn">..util</span> <span class="k">import</span> <span class="n">devnull</span>


<div class="viewcode-block" id="BamStats"><a class="viewcode-back" href="../../../auto/mavis.bam.stats.html#mavis.bam.stats.BamStats">[docs]</a><span class="k">class</span> <span class="nc">BamStats</span><span class="p">:</span>

    <span class="k">def</span> <span class="nf">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">median_fragment_size</span><span class="p">,</span> <span class="n">stdev_fragment_size</span><span class="p">,</span> <span class="n">read_length</span><span class="p">):</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">median_fragment_size</span> <span class="o">=</span> <span class="n">median_fragment_size</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">stdev_fragment_size</span> <span class="o">=</span> <span class="n">stdev_fragment_size</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">read_length</span> <span class="o">=</span> <span class="n">read_length</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">stranded</span> <span class="o">=</span> <span class="kc">False</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">strand_determining_read</span> <span class="o">=</span> <span class="mi">2</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">sdr_percent_support</span> <span class="o">=</span> <span class="kc">None</span>

<div class="viewcode-block" id="BamStats.add_stranded_information"><a class="viewcode-back" href="../../../auto/mavis.bam.stats.html#mavis.bam.stats.BamStats.add_stranded_information">[docs]</a>    <span class="k">def</span> <span class="nf">add_stranded_information</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">strand_hist</span><span class="p">):</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">stranded</span> <span class="o">=</span> <span class="kc">True</span>
        <span class="k">if</span> <span class="n">strand_hist</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">&gt;</span> <span class="n">strand_hist</span><span class="p">[</span><span class="mi">2</span><span class="p">]:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">strand_determining_read</span> <span class="o">=</span> <span class="mi">1</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">sdr_percent_support</span> <span class="o">=</span> <span class="n">strand_hist</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">*</span> <span class="mi">100</span> <span class="o">/</span> <span class="p">(</span><span class="n">strand_hist</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">+</span> <span class="n">strand_hist</span><span class="p">[</span><span class="mi">2</span><span class="p">])</span>
        <span class="k">elif</span> <span class="n">strand_hist</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">&lt;</span> <span class="n">strand_hist</span><span class="p">[</span><span class="mi">2</span><span class="p">]:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">strand_determining_read</span> <span class="o">=</span> <span class="mi">2</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">sdr_percent_support</span> <span class="o">=</span> <span class="n">strand_hist</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span> <span class="o">*</span> <span class="mi">100</span> <span class="o">/</span> <span class="p">(</span><span class="n">strand_hist</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">+</span> <span class="n">strand_hist</span><span class="p">[</span><span class="mi">2</span><span class="p">])</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">AssertionError</span><span class="p">(</span>
                <span class="s1">&#39;stranded values are equal. either this information was not collected or is not deterministic&#39;</span><span class="p">,</span>
                <span class="n">strand_hist</span><span class="o">.</span><span class="n">items</span><span class="p">())</span></div>

    <span class="k">def</span> <span class="nf">__str__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="n">result</span> <span class="o">=</span> <span class="s1">&#39;BamStats(fragment_size=</span><span class="si">{0.median_fragment_size}</span><span class="s1">+/-&#39;</span> \
                 <span class="s1">&#39;</span><span class="si">{0.stdev_fragment_size:.4}</span><span class="s1">, read_length=</span><span class="si">{0.read_length}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">stranded</span><span class="p">:</span>
            <span class="n">result</span> <span class="o">+=</span> <span class="s1">&#39;, strand_determining_read=</span><span class="si">{0.strand_determining_read}</span><span class="s1">[</span><span class="si">{0.sdr_percent_support:.4}</span><span class="s1">]&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span>
        <span class="n">result</span> <span class="o">+=</span> <span class="s1">&#39;)&#39;</span>
        <span class="k">return</span> <span class="n">result</span></div>


<div class="viewcode-block" id="Histogram"><a class="viewcode-back" href="../../../auto/mavis.bam.stats.html#mavis.bam.stats.Histogram">[docs]</a><span class="k">class</span> <span class="nc">Histogram</span><span class="p">(</span><span class="nb">dict</span><span class="p">):</span>

<div class="viewcode-block" id="Histogram.add"><a class="viewcode-back" href="../../../auto/mavis.bam.stats.html#mavis.bam.stats.Histogram.add">[docs]</a>    <span class="k">def</span> <span class="nf">add</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">item</span><span class="p">,</span> <span class="n">freq</span><span class="o">=</span><span class="mi">1</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        add a key to the histogram with a default frequency of 1</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="p">[</span><span class="n">item</span><span class="p">]</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">item</span><span class="p">,</span> <span class="mi">0</span><span class="p">)</span> <span class="o">+</span> <span class="n">freq</span></div>

<div class="viewcode-block" id="Histogram.median"><a class="viewcode-back" href="../../../auto/mavis.bam.stats.html#mavis.bam.stats.Histogram.median">[docs]</a>    <span class="k">def</span> <span class="nf">median</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        flattens the histogram to compute the median value</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">values</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">val</span><span class="p">,</span> <span class="n">freq</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">freq</span><span class="p">):</span>
                <span class="n">values</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">val</span><span class="p">)</span>
        <span class="n">values</span><span class="o">.</span><span class="n">sort</span><span class="p">()</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">values</span><span class="p">)</span> <span class="o">%</span> <span class="mi">2</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
            <span class="n">low_center</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">values</span><span class="p">)</span> <span class="o">//</span> <span class="mi">2</span>
            <span class="n">high_center</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">values</span><span class="p">)</span> <span class="o">//</span> <span class="mi">2</span> <span class="o">+</span> <span class="mi">1</span>
            <span class="k">return</span> <span class="p">(</span><span class="n">values</span><span class="p">[</span><span class="n">low_center</span> <span class="o">-</span> <span class="mi">1</span><span class="p">]</span> <span class="o">+</span> <span class="n">values</span><span class="p">[</span><span class="n">high_center</span> <span class="o">-</span> <span class="mi">1</span><span class="p">])</span> <span class="o">/</span> <span class="mi">2</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">center</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">values</span><span class="p">)</span> <span class="o">//</span> <span class="mi">2</span> <span class="o">+</span> <span class="mi">1</span>
            <span class="k">return</span> <span class="n">values</span><span class="p">[</span><span class="n">center</span> <span class="o">-</span> <span class="mi">1</span><span class="p">]</span></div>

<div class="viewcode-block" id="Histogram.distribution_stderr"><a class="viewcode-back" href="../../../auto/mavis.bam.stats.html#mavis.bam.stats.Histogram.distribution_stderr">[docs]</a>    <span class="k">def</span> <span class="nf">distribution_stderr</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">median</span><span class="p">,</span> <span class="n">fraction</span><span class="p">,</span> <span class="n">error_function</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">,</span> <span class="n">y</span><span class="p">:</span> <span class="n">math</span><span class="o">.</span><span class="n">pow</span><span class="p">(</span><span class="n">x</span> <span class="o">-</span> <span class="n">y</span><span class="p">,</span> <span class="mi">2</span><span class="p">)):</span>
        <span class="n">values</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">val</span><span class="p">,</span> <span class="n">freq</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="n">err</span> <span class="o">=</span> <span class="n">error_function</span><span class="p">(</span><span class="n">val</span><span class="p">,</span> <span class="n">median</span><span class="p">)</span>
            <span class="n">values</span><span class="o">.</span><span class="n">extend</span><span class="p">([</span><span class="n">err</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">freq</span><span class="p">)])</span>
        <span class="n">values</span><span class="o">.</span><span class="n">sort</span><span class="p">()</span>

        <span class="n">end</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">values</span><span class="p">)</span> <span class="o">*</span> <span class="n">fraction</span><span class="p">)</span>
        <span class="k">return</span> <span class="nb">sum</span><span class="p">(</span><span class="n">values</span><span class="p">[</span><span class="mi">0</span><span class="p">:</span><span class="n">end</span><span class="p">])</span> <span class="o">/</span> <span class="n">end</span></div>

<div class="viewcode-block" id="Histogram.__add__"><a class="viewcode-back" href="../../../auto/mavis.bam.stats.html#mavis.bam.stats.Histogram.__add__">[docs]</a>    <span class="k">def</span> <span class="nf">__add__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">other</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        sum two histograms and return the result as a new histogram</span>

<span class="sd">        Example:</span>
<span class="sd">            &gt;&gt;&gt; x, y = Histogram(), Histogram()</span>
<span class="sd">            &gt;&gt;&gt; x.add(&#39;item&#39;)</span>
<span class="sd">            &gt;&gt;&gt; y.add(&#39;item&#39;)</span>
<span class="sd">            &gt;&gt;&gt; x + y</span>
<span class="sd">            {&#39;item&#39;: 2}</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">result</span> <span class="o">=</span> <span class="n">Histogram</span><span class="p">()</span>
        <span class="n">result</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">val</span><span class="p">,</span> <span class="n">freq</span> <span class="ow">in</span> <span class="n">other</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="n">result</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">val</span><span class="p">,</span> <span class="n">freq</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">result</span></div></div>


<div class="viewcode-block" id="compute_transcriptome_bam_stats"><a class="viewcode-back" href="../../../auto/mavis.bam.stats.html#mavis.bam.stats.compute_transcriptome_bam_stats">[docs]</a><span class="k">def</span> <span class="nf">compute_transcriptome_bam_stats</span><span class="p">(</span>
    <span class="n">bam_cache</span><span class="p">,</span>
    <span class="n">annotations</span><span class="p">,</span>
    <span class="n">sample_size</span><span class="p">,</span>
    <span class="n">min_mapping_quality</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
    <span class="n">stranded</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
    <span class="n">sample_cap</span><span class="o">=</span><span class="mi">10000</span><span class="p">,</span>
    <span class="n">distribution_fraction</span><span class="o">=</span><span class="mf">0.97</span>
<span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    computes various statistical measures relating the input bam file</span>

<span class="sd">    Args:</span>
<span class="sd">        bam_file_handle (BamCache): the input bam file handle</span>
<span class="sd">        annotations (object): see :func:`~mavis.annotate.load_reference_genes`</span>
<span class="sd">        sample_size (int): the number of genes to compute stats over</span>
<span class="sd">        log (callable): outputs logging information</span>
<span class="sd">        min_mapping_quality (int): the minimum mapping quality for a read to be used</span>
<span class="sd">        stranded (bool): if True then reads must match the gene strand</span>
<span class="sd">        sample_cap (int): maximum number of reads to collect for any given sample region</span>
<span class="sd">        distribution_fraction (float): the proportion of the distribution to use in computing stdev</span>

<span class="sd">    Returns:</span>
<span class="sd">        BamStats: the fragment size median, stdev and the read length in a object</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">total_annotations</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="nb">chr</span><span class="p">,</span> <span class="n">anns_list</span> <span class="ow">in</span> <span class="n">annotations</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="k">if</span> <span class="n">bam_cache</span><span class="o">.</span><span class="n">valid_chr</span><span class="p">(</span><span class="nb">chr</span><span class="p">):</span>
            <span class="n">total_annotations</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span><span class="n">anns_list</span><span class="p">)</span>

    <span class="n">genes</span> <span class="o">=</span> <span class="n">total_annotations</span>
    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">total_annotations</span><span class="p">)</span> <span class="o">&gt;</span> <span class="n">sample_size</span><span class="p">:</span>
        <span class="n">randoms</span> <span class="o">=</span> <span class="p">[</span><span class="nb">int</span><span class="p">(</span><span class="n">n</span> <span class="o">*</span> <span class="nb">len</span><span class="p">(</span><span class="n">total_annotations</span><span class="p">))</span> <span class="k">for</span> <span class="n">n</span> <span class="ow">in</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">rand</span><span class="p">(</span><span class="n">sample_size</span><span class="p">)]</span>
        <span class="n">genes</span> <span class="o">=</span> <span class="p">[</span><span class="n">total_annotations</span><span class="p">[</span><span class="n">r</span><span class="p">]</span> <span class="k">for</span> <span class="n">r</span> <span class="ow">in</span> <span class="n">randoms</span><span class="p">]</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">warnings</span><span class="o">.</span><span class="n">warn</span><span class="p">(</span>
            <span class="s1">&#39;insufficient annotations to match requested sample size. requested </span><span class="si">{}</span><span class="s1">, but only </span><span class="si">{}</span><span class="s1"> annotations&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                <span class="n">sample_size</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">total_annotations</span><span class="p">)))</span>

    <span class="n">fragment_hist</span> <span class="o">=</span> <span class="n">Histogram</span><span class="p">()</span>
    <span class="n">read_strand_verification</span> <span class="o">=</span> <span class="n">Histogram</span><span class="p">()</span>
    <span class="n">read_strand_verification</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="n">read_strand_verification</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span> <span class="o">=</span> <span class="mi">0</span>

    <span class="n">read_lengths</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">genes</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">read</span> <span class="ow">in</span> <span class="n">bam_cache</span><span class="o">.</span><span class="n">fetch</span><span class="p">(</span><span class="n">gene</span><span class="o">.</span><span class="n">chr</span><span class="p">,</span> <span class="n">gene</span><span class="o">.</span><span class="n">start</span><span class="p">,</span> <span class="n">gene</span><span class="o">.</span><span class="n">end</span><span class="p">,</span> <span class="n">cache_if</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span> <span class="n">limit</span><span class="o">=</span><span class="n">sample_cap</span><span class="p">):</span>
            <span class="k">if</span> <span class="nb">any</span><span class="p">([</span>
                <span class="n">read</span><span class="o">.</span><span class="n">is_unmapped</span><span class="p">,</span>
                <span class="n">read</span><span class="o">.</span><span class="n">mate_is_unmapped</span><span class="p">,</span>
                <span class="n">read</span><span class="o">.</span><span class="n">mapping_quality</span> <span class="o">&lt;</span> <span class="n">min_mapping_quality</span><span class="p">,</span>
                <span class="n">read</span><span class="o">.</span><span class="n">next_reference_id</span> <span class="o">!=</span> <span class="n">read</span><span class="o">.</span><span class="n">reference_id</span><span class="p">,</span>
                <span class="n">read</span><span class="o">.</span><span class="n">is_secondary</span><span class="p">,</span>
                <span class="ow">not</span> <span class="n">read</span><span class="o">.</span><span class="n">is_proper_pair</span>
            <span class="p">]):</span>
                <span class="k">continue</span>
            <span class="k">if</span> <span class="n">stranded</span><span class="p">:</span>
                <span class="k">try</span><span class="p">:</span>
                    <span class="n">strand</span> <span class="o">=</span> <span class="n">sequenced_strand</span><span class="p">(</span><span class="n">read</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
                    <span class="k">if</span> <span class="n">strand</span> <span class="o">==</span> <span class="n">gene</span><span class="o">.</span><span class="n">get_strand</span><span class="p">():</span>
                        <span class="n">read_strand_verification</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>
                    <span class="k">else</span><span class="p">:</span>
                        <span class="n">read_strand_verification</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="mi">2</span><span class="p">)</span>
                <span class="k">except</span> <span class="ne">ValueError</span><span class="p">:</span>
                    <span class="k">pass</span>
            <span class="k">if</span> <span class="n">stranded</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">read</span><span class="o">.</span><span class="n">is_read1</span><span class="p">:</span>
                    <span class="k">if</span> <span class="n">read</span><span class="o">.</span><span class="n">is_reverse</span><span class="p">:</span>
                        <span class="k">if</span> <span class="n">gene</span><span class="o">.</span><span class="n">get_strand</span><span class="p">()</span> <span class="o">==</span> <span class="n">STRAND</span><span class="o">.</span><span class="n">POS</span><span class="p">:</span>
                            <span class="k">continue</span>
                    <span class="k">elif</span> <span class="n">gene</span><span class="o">.</span><span class="n">get_strand</span><span class="p">()</span> <span class="o">==</span> <span class="n">STRAND</span><span class="o">.</span><span class="n">NEG</span><span class="p">:</span>
                        <span class="k">continue</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="k">if</span> <span class="n">read</span><span class="o">.</span><span class="n">is_reverse</span><span class="p">:</span>
                        <span class="k">if</span> <span class="n">gene</span><span class="o">.</span><span class="n">get_strand</span><span class="p">()</span> <span class="o">==</span> <span class="n">STRAND</span><span class="o">.</span><span class="n">NEG</span><span class="p">:</span>
                            <span class="k">continue</span>
                    <span class="k">elif</span> <span class="n">gene</span><span class="o">.</span><span class="n">get_strand</span><span class="p">()</span> <span class="o">==</span> <span class="n">STRAND</span><span class="o">.</span><span class="n">POS</span><span class="p">:</span>
                        <span class="k">continue</span>
            <span class="n">read_lengths</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">read</span><span class="o">.</span><span class="n">query_sequence</span><span class="p">))</span>

            <span class="k">if</span> <span class="n">read</span><span class="o">.</span><span class="n">reference_end</span> <span class="o">&gt;</span> <span class="n">read</span><span class="o">.</span><span class="n">next_reference_start</span><span class="p">:</span>
                <span class="k">continue</span>
            <span class="n">current_frags</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
            <span class="k">for</span> <span class="n">spl_tx</span> <span class="ow">in</span> <span class="n">gene</span><span class="o">.</span><span class="n">spliced_transcripts</span><span class="p">:</span>
                <span class="k">try</span><span class="p">:</span>
                    <span class="n">cdna_pos1</span> <span class="o">=</span> <span class="n">spl_tx</span><span class="o">.</span><span class="n">convert_genomic_to_cdna</span><span class="p">(</span><span class="n">read</span><span class="o">.</span><span class="n">reference_start</span><span class="p">)</span>
                    <span class="n">cdna_pos2</span> <span class="o">=</span> <span class="n">spl_tx</span><span class="o">.</span><span class="n">convert_genomic_to_cdna</span><span class="p">(</span><span class="n">read</span><span class="o">.</span><span class="n">next_reference_start</span><span class="p">)</span>
                    <span class="n">current_frags</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="nb">abs</span><span class="p">(</span><span class="n">cdna_pos1</span> <span class="o">-</span> <span class="n">cdna_pos2</span><span class="p">)</span> <span class="o">-</span> <span class="mi">2</span><span class="p">)</span>
                <span class="k">except</span> <span class="ne">IndexError</span><span class="p">:</span>
                    <span class="k">pass</span>
            <span class="k">if</span> <span class="n">current_frags</span><span class="p">:</span>
                <span class="n">fragment_hist</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="nb">sum</span><span class="p">(</span><span class="n">current_frags</span><span class="p">)</span> <span class="o">/</span> <span class="nb">len</span><span class="p">(</span><span class="n">current_frags</span><span class="p">))</span>
    <span class="n">read_length</span> <span class="o">=</span> <span class="n">stats</span><span class="o">.</span><span class="n">median</span><span class="p">(</span><span class="n">read_lengths</span><span class="p">)</span>
    <span class="n">result</span> <span class="o">=</span> <span class="n">Histogram</span><span class="p">()</span>
    <span class="k">for</span> <span class="n">val</span><span class="p">,</span> <span class="n">freq</span> <span class="ow">in</span> <span class="n">fragment_hist</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="n">result</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">val</span> <span class="o">+</span> <span class="n">read_length</span><span class="p">,</span> <span class="n">freq</span><span class="p">)</span>
    <span class="n">median</span> <span class="o">=</span> <span class="n">result</span><span class="o">.</span><span class="n">median</span><span class="p">()</span>
    <span class="n">err</span> <span class="o">=</span> <span class="n">result</span><span class="o">.</span><span class="n">distribution_stderr</span><span class="p">(</span><span class="n">median</span><span class="p">,</span> <span class="n">distribution_fraction</span><span class="p">)</span>
    <span class="n">bamstats</span> <span class="o">=</span> <span class="n">BamStats</span><span class="p">(</span><span class="n">median</span><span class="p">,</span> <span class="n">math</span><span class="o">.</span><span class="n">sqrt</span><span class="p">(</span><span class="n">err</span><span class="p">),</span> <span class="n">read_length</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">stranded</span><span class="p">:</span>
        <span class="n">bamstats</span><span class="o">.</span><span class="n">add_stranded_information</span><span class="p">(</span><span class="n">read_strand_verification</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">bamstats</span></div>


<div class="viewcode-block" id="compute_genome_bam_stats"><a class="viewcode-back" href="../../../auto/mavis.bam.stats.html#mavis.bam.stats.compute_genome_bam_stats">[docs]</a><span class="k">def</span> <span class="nf">compute_genome_bam_stats</span><span class="p">(</span>
    <span class="n">bam_file_handle</span><span class="p">,</span>
    <span class="n">sample_bin_size</span><span class="p">,</span>
    <span class="n">sample_size</span><span class="p">,</span>
    <span class="n">min_mapping_quality</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
    <span class="n">sample_cap</span><span class="o">=</span><span class="mi">10000</span><span class="p">,</span>
    <span class="n">distribution_fraction</span><span class="o">=</span><span class="mf">0.99</span>
<span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    computes various statistical measures relating the input bam file</span>

<span class="sd">    Args:</span>
<span class="sd">        bam_file_handle (pysam.AlignmentFile): the input bam file handle</span>
<span class="sd">        sample_bin_size (int): how large to make the sample bin (in bp)</span>
<span class="sd">        sample_size (int): the number of genes to compute stats over</span>
<span class="sd">        log (callable): outputs logging information</span>
<span class="sd">        min_mapping_quality (int): the minimum mapping quality for a read to be used</span>
<span class="sd">        sample_cap (int): maximum number of reads to collect for any given sample region</span>
<span class="sd">        distribution_fraction (float): the proportion of the distribution to use in computing stdev</span>

<span class="sd">    Returns:</span>
<span class="sd">        BamStats: the fragment size median, stdev and the read length in a object</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">total</span> <span class="o">=</span> <span class="nb">sum</span><span class="p">([</span><span class="n">l</span> <span class="o">-</span> <span class="n">sample_bin_size</span> <span class="k">for</span> <span class="n">l</span> <span class="ow">in</span> <span class="n">bam_file_handle</span><span class="o">.</span><span class="n">fh</span><span class="o">.</span><span class="n">lengths</span><span class="p">])</span>
    <span class="n">bins</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">randoms</span> <span class="o">=</span> <span class="p">[</span><span class="nb">int</span><span class="p">(</span><span class="n">n</span> <span class="o">*</span> <span class="p">(</span><span class="n">total</span> <span class="o">-</span> <span class="mi">1</span><span class="p">)</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span> <span class="k">for</span> <span class="n">n</span> <span class="ow">in</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">rand</span><span class="p">(</span><span class="n">sample_size</span><span class="p">)]</span>
    <span class="k">for</span> <span class="n">pos</span> <span class="ow">in</span> <span class="n">randoms</span><span class="p">:</span>
        <span class="n">template_index</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="k">for</span> <span class="n">template_length</span> <span class="ow">in</span> <span class="n">bam_file_handle</span><span class="o">.</span><span class="n">fh</span><span class="o">.</span><span class="n">lengths</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">pos</span> <span class="o">&gt;</span> <span class="n">template_length</span> <span class="o">-</span> <span class="n">sample_bin_size</span><span class="p">:</span>
                <span class="n">pos</span> <span class="o">-=</span> <span class="p">(</span><span class="n">template_length</span> <span class="o">-</span> <span class="n">sample_bin_size</span><span class="p">)</span>
                <span class="n">template_index</span> <span class="o">+=</span> <span class="mi">1</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">break</span>
        <span class="n">bins</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">bam_file_handle</span><span class="o">.</span><span class="n">fh</span><span class="o">.</span><span class="n">references</span><span class="p">[</span><span class="n">template_index</span><span class="p">],</span> <span class="n">pos</span><span class="p">,</span> <span class="n">pos</span> <span class="o">+</span> <span class="n">sample_bin_size</span><span class="p">))</span>

    <span class="n">hist</span> <span class="o">=</span> <span class="n">Histogram</span><span class="p">()</span>
    <span class="n">read_lengths</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">bin_chr</span><span class="p">,</span> <span class="n">bin_start</span><span class="p">,</span> <span class="n">bin_end</span> <span class="ow">in</span> <span class="n">bins</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">read</span> <span class="ow">in</span> <span class="n">bam_file_handle</span><span class="o">.</span><span class="n">fetch</span><span class="p">(</span><span class="n">bin_chr</span><span class="p">,</span> <span class="n">bin_start</span><span class="p">,</span> <span class="n">bin_end</span><span class="p">,</span> <span class="n">limit</span><span class="o">=</span><span class="n">sample_cap</span><span class="p">,</span> <span class="n">cache_if</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="kc">False</span><span class="p">):</span>
            <span class="k">if</span> <span class="nb">any</span><span class="p">([</span>
                <span class="n">read</span><span class="o">.</span><span class="n">is_unmapped</span><span class="p">,</span>
                <span class="n">read</span><span class="o">.</span><span class="n">mate_is_unmapped</span><span class="p">,</span>
                <span class="n">read</span><span class="o">.</span><span class="n">mapping_quality</span> <span class="o">&lt;</span> <span class="n">min_mapping_quality</span><span class="p">,</span>
                <span class="n">read</span><span class="o">.</span><span class="n">next_reference_id</span> <span class="o">!=</span> <span class="n">read</span><span class="o">.</span><span class="n">reference_id</span><span class="p">,</span>
                <span class="n">read</span><span class="o">.</span><span class="n">is_secondary</span><span class="p">,</span>
                <span class="ow">not</span> <span class="n">read</span><span class="o">.</span><span class="n">is_proper_pair</span>
            <span class="p">]):</span>
                <span class="k">continue</span>
            <span class="n">hist</span><span class="p">[</span><span class="nb">abs</span><span class="p">(</span><span class="n">read</span><span class="o">.</span><span class="n">template_length</span><span class="p">)]</span> <span class="o">=</span> <span class="n">hist</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="nb">abs</span><span class="p">(</span><span class="n">read</span><span class="o">.</span><span class="n">template_length</span><span class="p">),</span> <span class="mi">0</span><span class="p">)</span> <span class="o">+</span> <span class="mi">1</span>
            <span class="n">read_lengths</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">read</span><span class="o">.</span><span class="n">query_sequence</span><span class="p">))</span>
    <span class="n">median</span> <span class="o">=</span> <span class="n">hist</span><span class="o">.</span><span class="n">median</span><span class="p">()</span>
    <span class="n">err</span> <span class="o">=</span> <span class="n">hist</span><span class="o">.</span><span class="n">distribution_stderr</span><span class="p">(</span><span class="n">median</span><span class="p">,</span> <span class="n">distribution_fraction</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">BamStats</span><span class="p">(</span><span class="n">median</span><span class="p">,</span> <span class="n">math</span><span class="o">.</span><span class="n">sqrt</span><span class="p">(</span><span class="n">err</span><span class="p">),</span> <span class="n">np</span><span class="o">.</span><span class="n">median</span><span class="p">(</span><span class="n">read_lengths</span><span class="p">))</span></div>
</pre></div>

           </div>
           <div class="articleComments">
            
           </div>
          </div>
          <footer>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2017, creisle.

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  

    <script type="text/javascript">
        var DOCUMENTATION_OPTIONS = {
            URL_ROOT:'../../../',
            VERSION:'',
            LANGUAGE:'None',
            COLLAPSE_INDEX:false,
            FILE_SUFFIX:'.html',
            HAS_SOURCE:  true,
            SOURCELINK_SUFFIX: '.txt'
        };
    </script>
      <script type="text/javascript" src="../../../_static/jquery.js"></script>
      <script type="text/javascript" src="../../../_static/underscore.js"></script>
      <script type="text/javascript" src="../../../_static/doctools.js"></script>
      <script type="text/javascript" src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>

  

  
  
    <script type="text/javascript" src="../../../_static/js/theme.js"></script>
  

  
  
  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.StickyNav.enable();
      });
  </script>
   

</body>
</html>